// rclnodejs/generated/example_interfaces/example_interfaces__srv__AddTwoInts_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const AddTwoInts_RequestRefStruct = StructType({
  a: primitiveTypes.int64,
  b: primitiveTypes.int64,
});
const AddTwoInts_RequestRefArray = ArrayType(AddTwoInts_RequestRefStruct);
const AddTwoInts_RequestRefStructArray = StructType({
  data: AddTwoInts_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AddTwoInts_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AddTwoInts_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AddTwoInts_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new AddTwoInts_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AddTwoInts_RequestArrayWrapper();
  }
  static get ArrayType() {
    return AddTwoInts_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return AddTwoInts_RequestRefStructArray;
  }
  static get refObjectType() {
    return AddTwoInts_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.a = refObject.a;
    this._refObject.b = refObject.b;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    AddTwoInts_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'example_interfaces', subFolder: 'srv', interfaceName: 'AddTwoInts_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get a() {
    return BigInt(this._refObject.a);
  }
  set a(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('a must be type of bigint');
    }
    this._refObject.a = value.toString();
  }
  get b() {
    return BigInt(this._refObject.b);
  }
  set b(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('b must be type of bigint');
    }
    this._refObject.b = value.toString();
  }
  copyRefObject(refObject) {
    this._refObject = new AddTwoInts_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new AddTwoInts_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return AddTwoInts_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"a","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"b","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"example_interfaces","type":"AddTwoInts_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AddTwoInts_Request"};
  }
  hasMember(name) {
    let memberNames = ["a","b"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AddTwoInts_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AddTwoInts_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AddTwoInts_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AddTwoInts_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AddTwoInts_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AddTwoInts_RequestRefArray(size);
    this._refObject = new AddTwoInts_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AddTwoInts_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AddTwoInts_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AddTwoInts_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AddTwoInts_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AddTwoInts_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AddTwoInts_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = AddTwoInts_RequestWrapper;
