// rclnodejs/generated/example_interfaces/example_interfaces__srv__AddTwoInts_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const AddTwoInts_ResponseRefStruct = StructType({
  sum: primitiveTypes.int64,
});
const AddTwoInts_ResponseRefArray = ArrayType(AddTwoInts_ResponseRefStruct);
const AddTwoInts_ResponseRefStructArray = StructType({
  data: AddTwoInts_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AddTwoInts_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AddTwoInts_ResponseRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AddTwoInts_ResponseRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new AddTwoInts_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AddTwoInts_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return AddTwoInts_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return AddTwoInts_ResponseRefStructArray;
  }
  static get refObjectType() {
    return AddTwoInts_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.sum = refObject.sum;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    AddTwoInts_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'example_interfaces', subFolder: 'srv', interfaceName: 'AddTwoInts_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sum() {
    return BigInt(this._refObject.sum);
  }
  set sum(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('sum must be type of bigint');
    }
    this._refObject.sum = value.toString();
  }
  copyRefObject(refObject) {
    this._refObject = new AddTwoInts_ResponseRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new AddTwoInts_ResponseRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return AddTwoInts_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"sum","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"example_interfaces","type":"AddTwoInts_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AddTwoInts_Response"};
  }
  hasMember(name) {
    let memberNames = ["sum"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AddTwoInts_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AddTwoInts_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AddTwoInts_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AddTwoInts_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AddTwoInts_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AddTwoInts_ResponseRefArray(size);
    this._refObject = new AddTwoInts_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AddTwoInts_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AddTwoInts_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AddTwoInts_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AddTwoInts_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AddTwoInts_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AddTwoInts_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = AddTwoInts_ResponseWrapper;
