// rclnodejs/generated/velodyne_msgs/velodyne_msgs__msg__VelodyneScan.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const VelodynePacketWrapper = require('../../generated/velodyne_msgs/velodyne_msgs__msg__VelodynePacket.js');
const VelodyneScanRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  packets: VelodynePacketWrapper.refObjectArrayType,
});
const VelodyneScanRefArray = ArrayType(VelodyneScanRefStruct);
const VelodyneScanRefStructArray = StructType({
  data: VelodyneScanRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class VelodyneScanWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new VelodyneScanRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.packets = VelodynePacketWrapper.createArray();
      this._wrapperFields.packets.copy(other._wrapperFields.packets);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new VelodyneScanRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.packets = VelodynePacketWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new VelodyneScanWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VelodyneScanArrayWrapper();
  }
  static get ArrayType() {
    return VelodyneScanArrayWrapper;
  }
  static get refObjectArrayType() {
    return VelodyneScanRefStructArray;
  }
  static get refObjectType() {
    return VelodyneScanRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.packets.fill([]);
    }
    this._wrapperFields.packets.freeze(own);
    this._refObject.packets = this._wrapperFields.packets.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.packets.copyRefObject(refObject.packets);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.packets.size != 0) {
      VelodynePacketWrapper.ArrayType.freeArray(refObject.packets);
      if (!VelodynePacketWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.packets, VelodynePacketWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    VelodyneScanWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'velodyne_msgs', subFolder: 'msg', interfaceName: 'VelodyneScan'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get packets() {
    return this._wrapperFields.packets;
  }
  set packets(value) {
    this._wrapperFields.packets.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new VelodyneScanRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.packets.copyRefObject(this._refObject.packets);
  }
  copy(other) {
    this._refObject = new VelodyneScanRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.packets.copy(other._wrapperFields.packets);
  }
  static get classType() {
    return VelodyneScanWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"packets","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"velodyne_msgs","type":"VelodynePacket","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"velodyne_msgs","type":"VelodyneScan","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VelodyneScan"};
  }
  hasMember(name) {
    let memberNames = ["header","packets"];
    return memberNames.indexOf(name) !== -1;
  }
}
class VelodyneScanArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VelodyneScanWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VelodyneScanWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VelodyneScanArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VelodyneScanArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new VelodyneScanRefArray(size);
    this._refObject = new VelodyneScanRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VelodyneScanWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VelodyneScanArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VelodyneScanArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VelodyneScanWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VelodyneScanWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VelodyneScanArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = VelodyneScanWrapper;
