// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__NodeDetails_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const NodeDetails_ResponseRefStruct = StructType({
  subscribing: StringWrapper.refObjectArrayType,
  publishing: StringWrapper.refObjectArrayType,
  services: StringWrapper.refObjectArrayType,
});
const NodeDetails_ResponseRefArray = ArrayType(NodeDetails_ResponseRefStruct);
const NodeDetails_ResponseRefStructArray = StructType({
  data: NodeDetails_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class NodeDetails_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new NodeDetails_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.subscribing = StringWrapper.createArray();
      this._wrapperFields.subscribing.copy(other._wrapperFields.subscribing);
      this.subscribing = other.subscribing;
      this._wrapperFields.publishing = StringWrapper.createArray();
      this._wrapperFields.publishing.copy(other._wrapperFields.publishing);
      this.publishing = other.publishing;
      this._wrapperFields.services = StringWrapper.createArray();
      this._wrapperFields.services.copy(other._wrapperFields.services);
      this.services = other.services;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new NodeDetails_ResponseRefStruct();
    this._wrapperFields.subscribing = StringWrapper.createArray();
    this._subscribingArray = [];
    this._wrapperFields.publishing = StringWrapper.createArray();
    this._publishingArray = [];
    this._wrapperFields.services = StringWrapper.createArray();
    this._servicesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new NodeDetails_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NodeDetails_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return NodeDetails_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return NodeDetails_ResponseRefStructArray;
  }
  static get refObjectType() {
    return NodeDetails_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.subscribing.fill([]);
    } else {
      this._wrapperFields.subscribing.fill(this._subscribingArray);
    }
    this._wrapperFields.subscribing.freeze(own);
    this._refObject.subscribing = this._wrapperFields.subscribing.refObject;
    if (own) {
      this._wrapperFields.publishing.fill([]);
    } else {
      this._wrapperFields.publishing.fill(this._publishingArray);
    }
    this._wrapperFields.publishing.freeze(own);
    this._refObject.publishing = this._wrapperFields.publishing.refObject;
    if (own) {
      this._wrapperFields.services.fill([]);
    } else {
      this._wrapperFields.services.fill(this._servicesArray);
    }
    this._wrapperFields.services.freeze(own);
    this._refObject.services = this._wrapperFields.services.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.subscribing.data.length = refObject.subscribing.size;
    for (let index = 0; index < refObject.subscribing.size; index++) {
      this._subscribingArray[index] = refObject.subscribing.data[index].data;
    }
    refObject.publishing.data.length = refObject.publishing.size;
    for (let index = 0; index < refObject.publishing.size; index++) {
      this._publishingArray[index] = refObject.publishing.data[index].data;
    }
    refObject.services.data.length = refObject.services.size;
    for (let index = 0; index < refObject.services.size; index++) {
      this._servicesArray[index] = refObject.services.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.subscribing.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.subscribing);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.subscribing, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.publishing.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.publishing);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.publishing, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.services.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.services);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.services, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    NodeDetails_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'NodeDetails_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get subscribing() {
    return this._subscribingArray;
  }
  set subscribing(value) {
    this._subscribingArray = value;
  }
  get publishing() {
    return this._publishingArray;
  }
  set publishing(value) {
    this._publishingArray = value;
  }
  get services() {
    return this._servicesArray;
  }
  set services(value) {
    this._servicesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NodeDetails_ResponseRefStruct(refObject.toObject());
    refObject.subscribing.data.length = refObject.subscribing.size;
    for (let index = 0; index < refObject.subscribing.size; index++) {
      this._subscribingArray[index] = refObject.subscribing.data[index].data;
    }
    refObject.publishing.data.length = refObject.publishing.size;
    for (let index = 0; index < refObject.publishing.size; index++) {
      this._publishingArray[index] = refObject.publishing.data[index].data;
    }
    refObject.services.data.length = refObject.services.size;
    for (let index = 0; index < refObject.services.size; index++) {
      this._servicesArray[index] = refObject.services.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new NodeDetails_ResponseRefStruct(other._refObject.toObject());
    this._subscribingArray = other._subscribingArray.slice();
    this._publishingArray = other._publishingArray.slice();
    this._servicesArray = other._servicesArray.slice();
  }
  static get classType() {
    return NodeDetails_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"subscribing","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"publishing","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"services","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"NodeDetails_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NodeDetails_Response"};
  }
  hasMember(name) {
    let memberNames = ["subscribing","publishing","services"];
    return memberNames.indexOf(name) !== -1;
  }
}
class NodeDetails_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NodeDetails_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NodeDetails_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeDetails_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeDetails_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new NodeDetails_ResponseRefArray(size);
    this._refObject = new NodeDetails_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NodeDetails_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NodeDetails_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NodeDetails_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NodeDetails_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NodeDetails_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NodeDetails_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = NodeDetails_ResponseWrapper;
