// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__GetROSVersion_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetROSVersion_ResponseRefStruct = StructType({
  version: primitiveTypes.uint8,
  distro: primitiveTypes.string,
});
const GetROSVersion_ResponseRefArray = ArrayType(GetROSVersion_ResponseRefStruct);
const GetROSVersion_ResponseRefStructArray = StructType({
  data: GetROSVersion_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetROSVersion_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetROSVersion_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.distro =  new StringWrapper(other._wrapperFields.distro);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetROSVersion_ResponseRefStruct();
    this._wrapperFields.distro = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetROSVersion_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetROSVersion_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetROSVersion_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetROSVersion_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetROSVersion_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.distro.freeze(own);
    }
    this._refObject.distro = this._wrapperFields.distro.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.version = refObject.version;
    this._wrapperFields.distro.data = refObject.distro.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.distro);
  }
  static destroyRawROS(msg) {
    GetROSVersion_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'GetROSVersion_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get version() {
    return this._refObject.version;
  }
  set version(value) {
    this._refObject.version = value;
  }
  get distro() {
    return this._wrapperFields.distro.data;
  }
  set distro(value) {
    this._wrapperFields.distro.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetROSVersion_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.distro.copyRefObject(this._refObject.distro);
  }
  copy(other) {
    this._refObject = new GetROSVersion_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.distro.copy(other._wrapperFields.distro);
  }
  static get classType() {
    return GetROSVersion_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distro","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"GetROSVersion_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetROSVersion_Response"};
  }
  hasMember(name) {
    let memberNames = ["version","distro"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetROSVersion_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetROSVersion_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetROSVersion_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetROSVersion_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetROSVersion_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetROSVersion_ResponseRefArray(size);
    this._refObject = new GetROSVersion_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetROSVersion_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetROSVersion_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetROSVersion_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetROSVersion_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetROSVersion_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetROSVersion_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetROSVersion_ResponseWrapper;
