// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__Interfaces_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Interfaces_ResponseRefStruct = StructType({
  interfaces: StringWrapper.refObjectArrayType,
});
const Interfaces_ResponseRefArray = ArrayType(Interfaces_ResponseRefStruct);
const Interfaces_ResponseRefStructArray = StructType({
  data: Interfaces_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Interfaces_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Interfaces_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.interfaces = StringWrapper.createArray();
      this._wrapperFields.interfaces.copy(other._wrapperFields.interfaces);
      this.interfaces = other.interfaces;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Interfaces_ResponseRefStruct();
    this._wrapperFields.interfaces = StringWrapper.createArray();
    this._interfacesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new Interfaces_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Interfaces_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return Interfaces_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return Interfaces_ResponseRefStructArray;
  }
  static get refObjectType() {
    return Interfaces_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.interfaces.fill([]);
    } else {
      this._wrapperFields.interfaces.fill(this._interfacesArray);
    }
    this._wrapperFields.interfaces.freeze(own);
    this._refObject.interfaces = this._wrapperFields.interfaces.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.interfaces.data.length = refObject.interfaces.size;
    for (let index = 0; index < refObject.interfaces.size; index++) {
      this._interfacesArray[index] = refObject.interfaces.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.interfaces.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.interfaces);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.interfaces, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    Interfaces_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'Interfaces_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get interfaces() {
    return this._interfacesArray;
  }
  set interfaces(value) {
    this._interfacesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Interfaces_ResponseRefStruct(refObject.toObject());
    refObject.interfaces.data.length = refObject.interfaces.size;
    for (let index = 0; index < refObject.interfaces.size; index++) {
      this._interfacesArray[index] = refObject.interfaces.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new Interfaces_ResponseRefStruct(other._refObject.toObject());
    this._interfacesArray = other._interfacesArray.slice();
  }
  static get classType() {
    return Interfaces_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"interfaces","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"Interfaces_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Interfaces_Response"};
  }
  hasMember(name) {
    let memberNames = ["interfaces"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Interfaces_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Interfaces_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Interfaces_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Interfaces_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Interfaces_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Interfaces_ResponseRefArray(size);
    this._refObject = new Interfaces_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Interfaces_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Interfaces_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Interfaces_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Interfaces_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Interfaces_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Interfaces_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Interfaces_ResponseWrapper;
