// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__ServicesForType_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ServicesForType_ResponseRefStruct = StructType({
  services: StringWrapper.refObjectArrayType,
});
const ServicesForType_ResponseRefArray = ArrayType(ServicesForType_ResponseRefStruct);
const ServicesForType_ResponseRefStructArray = StructType({
  data: ServicesForType_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ServicesForType_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServicesForType_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.services = StringWrapper.createArray();
      this._wrapperFields.services.copy(other._wrapperFields.services);
      this.services = other.services;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ServicesForType_ResponseRefStruct();
    this._wrapperFields.services = StringWrapper.createArray();
    this._servicesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new ServicesForType_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServicesForType_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return ServicesForType_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServicesForType_ResponseRefStructArray;
  }
  static get refObjectType() {
    return ServicesForType_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.services.fill([]);
    } else {
      this._wrapperFields.services.fill(this._servicesArray);
    }
    this._wrapperFields.services.freeze(own);
    this._refObject.services = this._wrapperFields.services.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.services.data.length = refObject.services.size;
    for (let index = 0; index < refObject.services.size; index++) {
      this._servicesArray[index] = refObject.services.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.services.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.services);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.services, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ServicesForType_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'ServicesForType_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get services() {
    return this._servicesArray;
  }
  set services(value) {
    this._servicesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServicesForType_ResponseRefStruct(refObject.toObject());
    refObject.services.data.length = refObject.services.size;
    for (let index = 0; index < refObject.services.size; index++) {
      this._servicesArray[index] = refObject.services.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ServicesForType_ResponseRefStruct(other._refObject.toObject());
    this._servicesArray = other._servicesArray.slice();
  }
  static get classType() {
    return ServicesForType_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"services","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"ServicesForType_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServicesForType_Response"};
  }
  hasMember(name) {
    let memberNames = ["services"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ServicesForType_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServicesForType_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServicesForType_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServicesForType_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServicesForType_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ServicesForType_ResponseRefArray(size);
    this._refObject = new ServicesForType_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServicesForType_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServicesForType_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServicesForType_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServicesForType_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServicesForType_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServicesForType_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ServicesForType_ResponseWrapper;
