// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__ServiceRequestDetails_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TypeDefWrapper = require('../../generated/rosapi_msgs/rosapi_msgs__msg__TypeDef.js');
const ServiceRequestDetails_ResponseRefStruct = StructType({
  typedefs: TypeDefWrapper.refObjectArrayType,
});
const ServiceRequestDetails_ResponseRefArray = ArrayType(ServiceRequestDetails_ResponseRefStruct);
const ServiceRequestDetails_ResponseRefStructArray = StructType({
  data: ServiceRequestDetails_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ServiceRequestDetails_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServiceRequestDetails_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.typedefs = TypeDefWrapper.createArray();
      this._wrapperFields.typedefs.copy(other._wrapperFields.typedefs);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ServiceRequestDetails_ResponseRefStruct();
    this._wrapperFields.typedefs = TypeDefWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ServiceRequestDetails_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServiceRequestDetails_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return ServiceRequestDetails_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServiceRequestDetails_ResponseRefStructArray;
  }
  static get refObjectType() {
    return ServiceRequestDetails_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.typedefs.fill([]);
    }
    this._wrapperFields.typedefs.freeze(own);
    this._refObject.typedefs = this._wrapperFields.typedefs.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.typedefs.copyRefObject(refObject.typedefs);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.typedefs.size != 0) {
      TypeDefWrapper.ArrayType.freeArray(refObject.typedefs);
      if (!TypeDefWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.typedefs, TypeDefWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ServiceRequestDetails_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'ServiceRequestDetails_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get typedefs() {
    return this._wrapperFields.typedefs;
  }
  set typedefs(value) {
    this._wrapperFields.typedefs.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ServiceRequestDetails_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.typedefs.copyRefObject(this._refObject.typedefs);
  }
  copy(other) {
    this._refObject = new ServiceRequestDetails_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.typedefs.copy(other._wrapperFields.typedefs);
  }
  static get classType() {
    return ServiceRequestDetails_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"typedefs","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rosapi_msgs","type":"TypeDef","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"ServiceRequestDetails_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServiceRequestDetails_Response"};
  }
  hasMember(name) {
    let memberNames = ["typedefs"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ServiceRequestDetails_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServiceRequestDetails_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServiceRequestDetails_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceRequestDetails_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceRequestDetails_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ServiceRequestDetails_ResponseRefArray(size);
    this._refObject = new ServiceRequestDetails_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServiceRequestDetails_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServiceRequestDetails_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServiceRequestDetails_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServiceRequestDetails_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServiceRequestDetails_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServiceRequestDetails_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ServiceRequestDetails_ResponseWrapper;
