// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__Topics_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Topics_ResponseRefStruct = StructType({
  topics: StringWrapper.refObjectArrayType,
  types: StringWrapper.refObjectArrayType,
});
const Topics_ResponseRefArray = ArrayType(Topics_ResponseRefStruct);
const Topics_ResponseRefStructArray = StructType({
  data: Topics_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Topics_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Topics_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.topics = StringWrapper.createArray();
      this._wrapperFields.topics.copy(other._wrapperFields.topics);
      this.topics = other.topics;
      this._wrapperFields.types = StringWrapper.createArray();
      this._wrapperFields.types.copy(other._wrapperFields.types);
      this.types = other.types;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Topics_ResponseRefStruct();
    this._wrapperFields.topics = StringWrapper.createArray();
    this._topicsArray = [];
    this._wrapperFields.types = StringWrapper.createArray();
    this._typesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new Topics_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Topics_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return Topics_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return Topics_ResponseRefStructArray;
  }
  static get refObjectType() {
    return Topics_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.topics.fill([]);
    } else {
      this._wrapperFields.topics.fill(this._topicsArray);
    }
    this._wrapperFields.topics.freeze(own);
    this._refObject.topics = this._wrapperFields.topics.refObject;
    if (own) {
      this._wrapperFields.types.fill([]);
    } else {
      this._wrapperFields.types.fill(this._typesArray);
    }
    this._wrapperFields.types.freeze(own);
    this._refObject.types = this._wrapperFields.types.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.topics.data.length = refObject.topics.size;
    for (let index = 0; index < refObject.topics.size; index++) {
      this._topicsArray[index] = refObject.topics.data[index].data;
    }
    refObject.types.data.length = refObject.types.size;
    for (let index = 0; index < refObject.types.size; index++) {
      this._typesArray[index] = refObject.types.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.topics.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.topics);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.topics, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.types.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.types);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.types, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    Topics_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'Topics_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get topics() {
    return this._topicsArray;
  }
  set topics(value) {
    this._topicsArray = value;
  }
  get types() {
    return this._typesArray;
  }
  set types(value) {
    this._typesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Topics_ResponseRefStruct(refObject.toObject());
    refObject.topics.data.length = refObject.topics.size;
    for (let index = 0; index < refObject.topics.size; index++) {
      this._topicsArray[index] = refObject.topics.data[index].data;
    }
    refObject.types.data.length = refObject.types.size;
    for (let index = 0; index < refObject.types.size; index++) {
      this._typesArray[index] = refObject.types.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new Topics_ResponseRefStruct(other._refObject.toObject());
    this._topicsArray = other._topicsArray.slice();
    this._typesArray = other._typesArray.slice();
  }
  static get classType() {
    return Topics_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"topics","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"types","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"Topics_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Topics_Response"};
  }
  hasMember(name) {
    let memberNames = ["topics","types"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Topics_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Topics_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Topics_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Topics_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Topics_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Topics_ResponseRefArray(size);
    this._refObject = new Topics_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Topics_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Topics_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Topics_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Topics_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Topics_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Topics_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Topics_ResponseWrapper;
