// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__ServiceNode_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ServiceNode_ResponseRefStruct = StructType({
  node: primitiveTypes.string,
});
const ServiceNode_ResponseRefArray = ArrayType(ServiceNode_ResponseRefStruct);
const ServiceNode_ResponseRefStructArray = StructType({
  data: ServiceNode_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ServiceNode_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServiceNode_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.node =  new StringWrapper(other._wrapperFields.node);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ServiceNode_ResponseRefStruct();
    this._wrapperFields.node = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ServiceNode_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServiceNode_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return ServiceNode_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServiceNode_ResponseRefStructArray;
  }
  static get refObjectType() {
    return ServiceNode_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.node.freeze(own);
    }
    this._refObject.node = this._wrapperFields.node.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.node.data = refObject.node.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.node);
  }
  static destroyRawROS(msg) {
    ServiceNode_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'ServiceNode_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node() {
    return this._wrapperFields.node.data;
  }
  set node(value) {
    this._wrapperFields.node.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServiceNode_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.node.copyRefObject(this._refObject.node);
  }
  copy(other) {
    this._refObject = new ServiceNode_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.node.copy(other._wrapperFields.node);
  }
  static get classType() {
    return ServiceNode_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"ServiceNode_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServiceNode_Response"};
  }
  hasMember(name) {
    let memberNames = ["node"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ServiceNode_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServiceNode_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServiceNode_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceNode_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceNode_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ServiceNode_ResponseRefArray(size);
    this._refObject = new ServiceNode_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServiceNode_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServiceNode_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServiceNode_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServiceNode_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServiceNode_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServiceNode_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ServiceNode_ResponseWrapper;
