// rclnodejs/generated/pcl_msgs/pcl_msgs__msg__ModelCoefficients.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const ModelCoefficientsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  values: Float32Wrapper.refObjectArrayType,
});
const ModelCoefficientsRefArray = ArrayType(ModelCoefficientsRefStruct);
const ModelCoefficientsRefStructArray = StructType({
  data: ModelCoefficientsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ModelCoefficientsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ModelCoefficientsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.values = Float32Wrapper.createArray();
      this._wrapperFields.values.copy(other._wrapperFields.values);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ModelCoefficientsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.values = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ModelCoefficientsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ModelCoefficientsArrayWrapper();
  }
  static get ArrayType() {
    return ModelCoefficientsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ModelCoefficientsRefStructArray;
  }
  static get refObjectType() {
    return ModelCoefficientsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.values.fill(Float32Array.from([]));
    }
    this._wrapperFields.values.freeze(own);
    this._refObject.values = this._wrapperFields.values.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.values.copyRefObject(refObject.values);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.values.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.values);
      if (!Float32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.values, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ModelCoefficientsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pcl_msgs', subFolder: 'msg', interfaceName: 'ModelCoefficients'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get values() {
    return this._wrapperFields['values'].data;
  }
  set values(value) {
    this._wrapperFields['values'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ModelCoefficientsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.values.copyRefObject(this._refObject.values);
  }
  copy(other) {
    this._refObject = new ModelCoefficientsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.values.copy(other._wrapperFields.values);
  }
  static get classType() {
    return ModelCoefficientsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pcl_msgs","type":"ModelCoefficients","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ModelCoefficients"};
  }
  hasMember(name) {
    let memberNames = ["header","values"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ModelCoefficientsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ModelCoefficientsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ModelCoefficientsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ModelCoefficientsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ModelCoefficientsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ModelCoefficientsRefArray(size);
    this._refObject = new ModelCoefficientsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ModelCoefficientsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ModelCoefficientsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ModelCoefficientsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ModelCoefficientsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ModelCoefficientsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ModelCoefficientsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ModelCoefficientsWrapper;
