// rclnodejs/generated/nav2_msgs/nav2_msgs__action__Spin_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const Spin_ResultRefStruct = StructType({
  total_elapsed_time: DurationWrapper.refObjectType,
});
const Spin_ResultRefArray = ArrayType(Spin_ResultRefStruct);
const Spin_ResultRefStructArray = StructType({
  data: Spin_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Spin_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Spin_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.total_elapsed_time =  new DurationWrapper(other._wrapperFields.total_elapsed_time);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Spin_ResultRefStruct();
    this._wrapperFields.total_elapsed_time = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Spin_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Spin_ResultArrayWrapper();
  }
  static get ArrayType() {
    return Spin_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return Spin_ResultRefStructArray;
  }
  static get refObjectType() {
    return Spin_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.total_elapsed_time.freeze(own);
    this._refObject.total_elapsed_time = this._wrapperFields.total_elapsed_time.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.total_elapsed_time.copyRefObject(refObject.total_elapsed_time);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.total_elapsed_time);
  }
  static destroyRawROS(msg) {
    Spin_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'Spin_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get total_elapsed_time() {
    return this._wrapperFields.total_elapsed_time;
  }
  set total_elapsed_time(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.total_elapsed_time.copy(value);
    } else {
      this._wrapperFields.total_elapsed_time.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new Spin_ResultRefStruct(refObject.toObject());
    this._wrapperFields.total_elapsed_time.copyRefObject(this._refObject.total_elapsed_time);
  }
  copy(other) {
    this._refObject = new Spin_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.total_elapsed_time.copy(other._wrapperFields.total_elapsed_time);
  }
  static get classType() {
    return Spin_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"total_elapsed_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"Spin_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Spin_Result"};
  }
  hasMember(name) {
    let memberNames = ["total_elapsed_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Spin_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Spin_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Spin_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Spin_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Spin_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Spin_ResultRefArray(size);
    this._refObject = new Spin_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Spin_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Spin_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Spin_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Spin_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Spin_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Spin_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Spin_ResultWrapper;
