// rclnodejs/generated/nav2_msgs/nav2_msgs__action__ComputeAndTrackRoute_Goal.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const ComputeAndTrackRoute_GoalRefStruct = StructType({
  start_id: primitiveTypes.uint16,
  start: PoseStampedWrapper.refObjectType,
  goal_id: primitiveTypes.uint16,
  goal: PoseStampedWrapper.refObjectType,
  use_start: primitiveTypes.bool,
  use_poses: primitiveTypes.bool,
});
const ComputeAndTrackRoute_GoalRefArray = ArrayType(ComputeAndTrackRoute_GoalRefStruct);
const ComputeAndTrackRoute_GoalRefStructArray = StructType({
  data: ComputeAndTrackRoute_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ComputeAndTrackRoute_GoalWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ComputeAndTrackRoute_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.start =  new PoseStampedWrapper(other._wrapperFields.start);
      this._wrapperFields.goal =  new PoseStampedWrapper(other._wrapperFields.goal);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ComputeAndTrackRoute_GoalRefStruct();
    this._wrapperFields.start = new PoseStampedWrapper();
    this._wrapperFields.goal = new PoseStampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ComputeAndTrackRoute_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ComputeAndTrackRoute_GoalArrayWrapper();
  }
  static get ArrayType() {
    return ComputeAndTrackRoute_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return ComputeAndTrackRoute_GoalRefStructArray;
  }
  static get refObjectType() {
    return ComputeAndTrackRoute_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.start.freeze(own);
    this._refObject.start = this._wrapperFields.start.refObject;
    this._wrapperFields.goal.freeze(own);
    this._refObject.goal = this._wrapperFields.goal.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.start_id = refObject.start_id;
    this._wrapperFields.start.copyRefObject(refObject.start);
    this._refObject.goal_id = refObject.goal_id;
    this._wrapperFields.goal.copyRefObject(refObject.goal);
    this._refObject.use_start = refObject.use_start;
    this._refObject.use_poses = refObject.use_poses;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.start);
    PoseStampedWrapper.freeStruct(refObject.goal);
  }
  static destroyRawROS(msg) {
    ComputeAndTrackRoute_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'ComputeAndTrackRoute_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start_id() {
    return this._refObject.start_id;
  }
  set start_id(value) {
    this._refObject.start_id = value;
  }
  get start() {
    return this._wrapperFields.start;
  }
  set start(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.start.copy(value);
    } else {
      this._wrapperFields.start.copy(new PoseStampedWrapper(value));
    }
  }
  get goal_id() {
    return this._refObject.goal_id;
  }
  set goal_id(value) {
    this._refObject.goal_id = value;
  }
  get goal() {
    return this._wrapperFields.goal;
  }
  set goal(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.goal.copy(value);
    } else {
      this._wrapperFields.goal.copy(new PoseStampedWrapper(value));
    }
  }
  get use_start() {
    return this._refObject.use_start;
  }
  set use_start(value) {
    this._refObject.use_start = value;
  }
  get use_poses() {
    return this._refObject.use_poses;
  }
  set use_poses(value) {
    this._refObject.use_poses = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ComputeAndTrackRoute_GoalRefStruct(refObject.toObject());
    this._wrapperFields.start.copyRefObject(this._refObject.start);
    this._wrapperFields.goal.copyRefObject(this._refObject.goal);
  }
  copy(other) {
    this._refObject = new ComputeAndTrackRoute_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.start.copy(other._wrapperFields.start);
    this._wrapperFields.goal.copy(other._wrapperFields.goal);
  }
  static get classType() {
    return ComputeAndTrackRoute_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"use_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"use_poses","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ComputeAndTrackRoute_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ComputeAndTrackRoute_Goal"};
  }
  hasMember(name) {
    let memberNames = ["start_id","start","goal_id","goal","use_start","use_poses"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ComputeAndTrackRoute_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ComputeAndTrackRoute_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ComputeAndTrackRoute_GoalWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeAndTrackRoute_GoalArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeAndTrackRoute_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ComputeAndTrackRoute_GoalRefArray(size);
    this._refObject = new ComputeAndTrackRoute_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ComputeAndTrackRoute_GoalWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ComputeAndTrackRoute_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ComputeAndTrackRoute_GoalArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ComputeAndTrackRoute_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ComputeAndTrackRoute_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ComputeAndTrackRoute_GoalArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ComputeAndTrackRoute_GoalWrapper;
