// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__Route.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RouteNodeWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__RouteNode.js');
const RouteEdgeWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__RouteEdge.js');
const RouteRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  route_cost: primitiveTypes.float32,
  nodes: RouteNodeWrapper.refObjectArrayType,
  edges: RouteEdgeWrapper.refObjectArrayType,
});
const RouteRefArray = ArrayType(RouteRefStruct);
const RouteRefStructArray = StructType({
  data: RouteRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RouteWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RouteRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.nodes = RouteNodeWrapper.createArray();
      this._wrapperFields.nodes.copy(other._wrapperFields.nodes);
      this._wrapperFields.edges = RouteEdgeWrapper.createArray();
      this._wrapperFields.edges.copy(other._wrapperFields.edges);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RouteRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.nodes = RouteNodeWrapper.createArray();
    this._wrapperFields.edges = RouteEdgeWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new RouteWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RouteArrayWrapper();
  }
  static get ArrayType() {
    return RouteArrayWrapper;
  }
  static get refObjectArrayType() {
    return RouteRefStructArray;
  }
  static get refObjectType() {
    return RouteRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.nodes.fill([]);
    }
    this._wrapperFields.nodes.freeze(own);
    this._refObject.nodes = this._wrapperFields.nodes.refObject;
    if (own) {
      this._wrapperFields.edges.fill([]);
    }
    this._wrapperFields.edges.freeze(own);
    this._refObject.edges = this._wrapperFields.edges.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.route_cost = refObject.route_cost;
    this._wrapperFields.nodes.copyRefObject(refObject.nodes);
    this._wrapperFields.edges.copyRefObject(refObject.edges);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.nodes.size != 0) {
      RouteNodeWrapper.ArrayType.freeArray(refObject.nodes);
      if (!RouteNodeWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.nodes, RouteNodeWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.edges.size != 0) {
      RouteEdgeWrapper.ArrayType.freeArray(refObject.edges);
      if (!RouteEdgeWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.edges, RouteEdgeWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    RouteWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'Route'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get route_cost() {
    return this._refObject.route_cost;
  }
  set route_cost(value) {
    this._refObject.route_cost = value;
  }
  get nodes() {
    return this._wrapperFields.nodes;
  }
  set nodes(value) {
    this._wrapperFields.nodes.fill(value);
  }
  get edges() {
    return this._wrapperFields.edges;
  }
  set edges(value) {
    this._wrapperFields.edges.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new RouteRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.nodes.copyRefObject(this._refObject.nodes);
    this._wrapperFields.edges.copyRefObject(this._refObject.edges);
  }
  copy(other) {
    this._refObject = new RouteRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.nodes.copy(other._wrapperFields.nodes);
    this._wrapperFields.edges.copy(other._wrapperFields.edges);
  }
  static get classType() {
    return RouteWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"route_cost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nodes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"nav2_msgs","type":"RouteNode","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"edges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"nav2_msgs","type":"RouteEdge","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"Route","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Route"};
  }
  hasMember(name) {
    let memberNames = ["header","route_cost","nodes","edges"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RouteArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RouteWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RouteWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RouteRefArray(size);
    this._refObject = new RouteRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RouteWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RouteArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RouteArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RouteWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RouteWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RouteArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RouteWrapper;
