// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__RouteNode.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const RouteNodeRefStruct = StructType({
  nodeid: primitiveTypes.uint16,
  position: PointWrapper.refObjectType,
});
const RouteNodeRefArray = ArrayType(RouteNodeRefStruct);
const RouteNodeRefStructArray = StructType({
  data: RouteNodeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RouteNodeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RouteNodeRefStruct(other._refObject.toObject());
      this._wrapperFields.position =  new PointWrapper(other._wrapperFields.position);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RouteNodeRefStruct();
    this._wrapperFields.position = new PointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RouteNodeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RouteNodeArrayWrapper();
  }
  static get ArrayType() {
    return RouteNodeArrayWrapper;
  }
  static get refObjectArrayType() {
    return RouteNodeRefStructArray;
  }
  static get refObjectType() {
    return RouteNodeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.position.freeze(own);
    this._refObject.position = this._wrapperFields.position.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.nodeid = refObject.nodeid;
    this._wrapperFields.position.copyRefObject(refObject.position);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointWrapper.freeStruct(refObject.position);
  }
  static destroyRawROS(msg) {
    RouteNodeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'RouteNode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get nodeid() {
    return this._refObject.nodeid;
  }
  set nodeid(value) {
    this._refObject.nodeid = value;
  }
  get position() {
    return this._wrapperFields.position;
  }
  set position(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.position.copy(value);
    } else {
      this._wrapperFields.position.copy(new PointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new RouteNodeRefStruct(refObject.toObject());
    this._wrapperFields.position.copyRefObject(this._refObject.position);
  }
  copy(other) {
    this._refObject = new RouteNodeRefStruct(other._refObject.toObject());
    this._wrapperFields.position.copy(other._wrapperFields.position);
  }
  static get classType() {
    return RouteNodeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"nodeid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"RouteNode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RouteNode"};
  }
  hasMember(name) {
    let memberNames = ["nodeid","position"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RouteNodeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RouteNodeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RouteNodeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteNodeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteNodeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RouteNodeRefArray(size);
    this._refObject = new RouteNodeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RouteNodeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RouteNodeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RouteNodeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RouteNodeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RouteNodeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RouteNodeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RouteNodeWrapper;
