// rclnodejs/generated/nav2_msgs/nav2_msgs__action__ComputeRoute_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const RouteWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__Route.js');
const ComputeRoute_ResultRefStruct = StructType({
  planning_time: DurationWrapper.refObjectType,
  path: PathWrapper.refObjectType,
  route: RouteWrapper.refObjectType,
});
const ComputeRoute_ResultRefArray = ArrayType(ComputeRoute_ResultRefStruct);
const ComputeRoute_ResultRefStructArray = StructType({
  data: ComputeRoute_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ComputeRoute_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ComputeRoute_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.planning_time =  new DurationWrapper(other._wrapperFields.planning_time);
      this._wrapperFields.path =  new PathWrapper(other._wrapperFields.path);
      this._wrapperFields.route =  new RouteWrapper(other._wrapperFields.route);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ComputeRoute_ResultRefStruct();
    this._wrapperFields.planning_time = new DurationWrapper();
    this._wrapperFields.path = new PathWrapper();
    this._wrapperFields.route = new RouteWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ComputeRoute_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ComputeRoute_ResultArrayWrapper();
  }
  static get ArrayType() {
    return ComputeRoute_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return ComputeRoute_ResultRefStructArray;
  }
  static get refObjectType() {
    return ComputeRoute_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.planning_time.freeze(own);
    this._refObject.planning_time = this._wrapperFields.planning_time.refObject;
    this._wrapperFields.path.freeze(own);
    this._refObject.path = this._wrapperFields.path.refObject;
    this._wrapperFields.route.freeze(own);
    this._refObject.route = this._wrapperFields.route.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.planning_time.copyRefObject(refObject.planning_time);
    this._wrapperFields.path.copyRefObject(refObject.path);
    this._wrapperFields.route.copyRefObject(refObject.route);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.planning_time);
    PathWrapper.freeStruct(refObject.path);
    RouteWrapper.freeStruct(refObject.route);
  }
  static destroyRawROS(msg) {
    ComputeRoute_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'ComputeRoute_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get planning_time() {
    return this._wrapperFields.planning_time;
  }
  set planning_time(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.planning_time.copy(value);
    } else {
      this._wrapperFields.planning_time.copy(new DurationWrapper(value));
    }
  }
  get path() {
    return this._wrapperFields.path;
  }
  set path(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.path.copy(value);
    } else {
      this._wrapperFields.path.copy(new PathWrapper(value));
    }
  }
  get route() {
    return this._wrapperFields.route;
  }
  set route(value) {
    if (value instanceof RouteWrapper) {
      this._wrapperFields.route.copy(value);
    } else {
      this._wrapperFields.route.copy(new RouteWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ComputeRoute_ResultRefStruct(refObject.toObject());
    this._wrapperFields.planning_time.copyRefObject(this._refObject.planning_time);
    this._wrapperFields.path.copyRefObject(this._refObject.path);
    this._wrapperFields.route.copyRefObject(this._refObject.route);
  }
  copy(other) {
    this._refObject = new ComputeRoute_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.planning_time.copy(other._wrapperFields.planning_time);
    this._wrapperFields.path.copy(other._wrapperFields.path);
    this._wrapperFields.route.copy(other._wrapperFields.route);
  }
  static get classType() {
    return ComputeRoute_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"NONE","value":0},{"type":"uint16","name":"UNKNOWN","value":400},{"type":"uint16","name":"TF_ERROR","value":401},{"type":"uint16","name":"NO_VALID_GRAPH","value":402},{"type":"uint16","name":"INDETERMINANT_NODES_ON_GRAPH","value":403},{"type":"uint16","name":"TIMEOUT","value":404},{"type":"uint16","name":"NO_VALID_ROUTE","value":405},{"type":"uint16","name":"INVALID_EDGE_SCORER_USE","value":407}],"fields":[{"name":"planning_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"route","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav2_msgs","type":"Route","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ComputeRoute_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ComputeRoute_Result"};
  }
  hasMember(name) {
    let memberNames = ["planning_time","path","route"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ComputeRoute_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ComputeRoute_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ComputeRoute_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeRoute_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeRoute_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ComputeRoute_ResultRefArray(size);
    this._refObject = new ComputeRoute_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ComputeRoute_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ComputeRoute_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ComputeRoute_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ComputeRoute_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ComputeRoute_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ComputeRoute_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ComputeRoute_ResultWrapper, "NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "UNKNOWN", {value: 400, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "TF_ERROR", {value: 401, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "NO_VALID_GRAPH", {value: 402, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "INDETERMINANT_NODES_ON_GRAPH", {value: 403, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "TIMEOUT", {value: 404, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "NO_VALID_ROUTE", {value: 405, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeRoute_ResultWrapper, "INVALID_EDGE_SCORER_USE", {value: 407, writable: false, enumerable: true, configurable: true});
module.exports = ComputeRoute_ResultWrapper;
