// rclnodejs/generated/nav2_msgs/nav2_msgs__action__ComputeAndTrackRoute_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const ComputeAndTrackRoute_ResultRefStruct = StructType({
  execution_duration: DurationWrapper.refObjectType,
});
const ComputeAndTrackRoute_ResultRefArray = ArrayType(ComputeAndTrackRoute_ResultRefStruct);
const ComputeAndTrackRoute_ResultRefStructArray = StructType({
  data: ComputeAndTrackRoute_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ComputeAndTrackRoute_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ComputeAndTrackRoute_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.execution_duration =  new DurationWrapper(other._wrapperFields.execution_duration);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ComputeAndTrackRoute_ResultRefStruct();
    this._wrapperFields.execution_duration = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ComputeAndTrackRoute_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ComputeAndTrackRoute_ResultArrayWrapper();
  }
  static get ArrayType() {
    return ComputeAndTrackRoute_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return ComputeAndTrackRoute_ResultRefStructArray;
  }
  static get refObjectType() {
    return ComputeAndTrackRoute_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.execution_duration.freeze(own);
    this._refObject.execution_duration = this._wrapperFields.execution_duration.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.execution_duration.copyRefObject(refObject.execution_duration);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.execution_duration);
  }
  static destroyRawROS(msg) {
    ComputeAndTrackRoute_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'ComputeAndTrackRoute_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get execution_duration() {
    return this._wrapperFields.execution_duration;
  }
  set execution_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.execution_duration.copy(value);
    } else {
      this._wrapperFields.execution_duration.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ComputeAndTrackRoute_ResultRefStruct(refObject.toObject());
    this._wrapperFields.execution_duration.copyRefObject(this._refObject.execution_duration);
  }
  copy(other) {
    this._refObject = new ComputeAndTrackRoute_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.execution_duration.copy(other._wrapperFields.execution_duration);
  }
  static get classType() {
    return ComputeAndTrackRoute_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"NONE","value":0},{"type":"uint16","name":"UNKNOWN","value":400},{"type":"uint16","name":"TF_ERROR","value":401},{"type":"uint16","name":"NO_VALID_GRAPH","value":402},{"type":"uint16","name":"INDETERMINANT_NODES_ON_GRAPH","value":403},{"type":"uint16","name":"TIMEOUT","value":404},{"type":"uint16","name":"NO_VALID_ROUTE","value":405},{"type":"uint16","name":"OPERATION_FAILED","value":406},{"type":"uint16","name":"INVALID_EDGE_SCORER_USE","value":407}],"fields":[{"name":"execution_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ComputeAndTrackRoute_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ComputeAndTrackRoute_Result"};
  }
  hasMember(name) {
    let memberNames = ["execution_duration"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ComputeAndTrackRoute_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ComputeAndTrackRoute_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ComputeAndTrackRoute_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeAndTrackRoute_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeAndTrackRoute_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ComputeAndTrackRoute_ResultRefArray(size);
    this._refObject = new ComputeAndTrackRoute_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ComputeAndTrackRoute_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ComputeAndTrackRoute_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ComputeAndTrackRoute_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ComputeAndTrackRoute_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ComputeAndTrackRoute_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ComputeAndTrackRoute_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "UNKNOWN", {value: 400, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "TF_ERROR", {value: 401, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "NO_VALID_GRAPH", {value: 402, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "INDETERMINANT_NODES_ON_GRAPH", {value: 403, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "TIMEOUT", {value: 404, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "NO_VALID_ROUTE", {value: 405, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "OPERATION_FAILED", {value: 406, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ComputeAndTrackRoute_ResultWrapper, "INVALID_EDGE_SCORER_USE", {value: 407, writable: false, enumerable: true, configurable: true});
module.exports = ComputeAndTrackRoute_ResultWrapper;
