// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__DynamicEdges_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const EdgeCostWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__EdgeCost.js');
const DynamicEdges_RequestRefStruct = StructType({
  closed_edges: UInt16Wrapper.refObjectArrayType,
  opened_edges: UInt16Wrapper.refObjectArrayType,
  adjust_edges: EdgeCostWrapper.refObjectArrayType,
});
const DynamicEdges_RequestRefArray = ArrayType(DynamicEdges_RequestRefStruct);
const DynamicEdges_RequestRefStructArray = StructType({
  data: DynamicEdges_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class DynamicEdges_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new DynamicEdges_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.closed_edges = UInt16Wrapper.createArray();
      this._wrapperFields.closed_edges.copy(other._wrapperFields.closed_edges);
      this._wrapperFields.opened_edges = UInt16Wrapper.createArray();
      this._wrapperFields.opened_edges.copy(other._wrapperFields.opened_edges);
      this._wrapperFields.adjust_edges = EdgeCostWrapper.createArray();
      this._wrapperFields.adjust_edges.copy(other._wrapperFields.adjust_edges);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new DynamicEdges_RequestRefStruct();
    this._wrapperFields.closed_edges = UInt16Wrapper.createArray();
    this._wrapperFields.opened_edges = UInt16Wrapper.createArray();
    this._wrapperFields.adjust_edges = EdgeCostWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new DynamicEdges_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DynamicEdges_RequestArrayWrapper();
  }
  static get ArrayType() {
    return DynamicEdges_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return DynamicEdges_RequestRefStructArray;
  }
  static get refObjectType() {
    return DynamicEdges_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.closed_edges.fill(Uint16Array.from([]));
    }
    this._wrapperFields.closed_edges.freeze(own);
    this._refObject.closed_edges = this._wrapperFields.closed_edges.refObject;
    if (own) {
      this._wrapperFields.opened_edges.fill(Uint16Array.from([]));
    }
    this._wrapperFields.opened_edges.freeze(own);
    this._refObject.opened_edges = this._wrapperFields.opened_edges.refObject;
    if (own) {
      this._wrapperFields.adjust_edges.fill([]);
    }
    this._wrapperFields.adjust_edges.freeze(own);
    this._refObject.adjust_edges = this._wrapperFields.adjust_edges.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.closed_edges.copyRefObject(refObject.closed_edges);
    this._wrapperFields.opened_edges.copyRefObject(refObject.opened_edges);
    this._wrapperFields.adjust_edges.copyRefObject(refObject.adjust_edges);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.closed_edges.size != 0) {
      UInt16Wrapper.ArrayType.freeArray(refObject.closed_edges);
      if (!UInt16Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.closed_edges, UInt16Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.opened_edges.size != 0) {
      UInt16Wrapper.ArrayType.freeArray(refObject.opened_edges);
      if (!UInt16Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.opened_edges, UInt16Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.adjust_edges.size != 0) {
      EdgeCostWrapper.ArrayType.freeArray(refObject.adjust_edges);
      if (!EdgeCostWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.adjust_edges, EdgeCostWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    DynamicEdges_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'DynamicEdges_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get closed_edges() {
    return this._wrapperFields['closed_edges'].data;
  }
  set closed_edges(value) {
    this._wrapperFields['closed_edges'].fill(value);
  }
  get opened_edges() {
    return this._wrapperFields['opened_edges'].data;
  }
  set opened_edges(value) {
    this._wrapperFields['opened_edges'].fill(value);
  }
  get adjust_edges() {
    return this._wrapperFields.adjust_edges;
  }
  set adjust_edges(value) {
    this._wrapperFields.adjust_edges.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new DynamicEdges_RequestRefStruct(refObject.toObject());
    this._wrapperFields.closed_edges.copyRefObject(this._refObject.closed_edges);
    this._wrapperFields.opened_edges.copyRefObject(this._refObject.opened_edges);
    this._wrapperFields.adjust_edges.copyRefObject(this._refObject.adjust_edges);
  }
  copy(other) {
    this._refObject = new DynamicEdges_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.closed_edges.copy(other._wrapperFields.closed_edges);
    this._wrapperFields.opened_edges.copy(other._wrapperFields.opened_edges);
    this._wrapperFields.adjust_edges.copy(other._wrapperFields.adjust_edges);
  }
  static get classType() {
    return DynamicEdges_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"closed_edges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"opened_edges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"adjust_edges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"nav2_msgs","type":"EdgeCost","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"DynamicEdges_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DynamicEdges_Request"};
  }
  hasMember(name) {
    let memberNames = ["closed_edges","opened_edges","adjust_edges"];
    return memberNames.indexOf(name) !== -1;
  }
}
class DynamicEdges_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DynamicEdges_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DynamicEdges_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DynamicEdges_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DynamicEdges_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new DynamicEdges_RequestRefArray(size);
    this._refObject = new DynamicEdges_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DynamicEdges_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DynamicEdges_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DynamicEdges_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DynamicEdges_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DynamicEdges_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DynamicEdges_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = DynamicEdges_RequestWrapper;
