// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__CollisionMonitorState.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CollisionMonitorStateRefStruct = StructType({
  action_type: primitiveTypes.uint8,
  polygon_name: primitiveTypes.string,
});
const CollisionMonitorStateRefArray = ArrayType(CollisionMonitorStateRefStruct);
const CollisionMonitorStateRefStructArray = StructType({
  data: CollisionMonitorStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class CollisionMonitorStateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new CollisionMonitorStateRefStruct(other._refObject.toObject());
      this._wrapperFields.polygon_name =  new StringWrapper(other._wrapperFields.polygon_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new CollisionMonitorStateRefStruct();
    this._wrapperFields.polygon_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CollisionMonitorStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CollisionMonitorStateArrayWrapper();
  }
  static get ArrayType() {
    return CollisionMonitorStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return CollisionMonitorStateRefStructArray;
  }
  static get refObjectType() {
    return CollisionMonitorStateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.polygon_name.freeze(own);
    }
    this._refObject.polygon_name = this._wrapperFields.polygon_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.action_type = refObject.action_type;
    this._wrapperFields.polygon_name.data = refObject.polygon_name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.polygon_name);
  }
  static destroyRawROS(msg) {
    CollisionMonitorStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'CollisionMonitorState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get action_type() {
    return this._refObject.action_type;
  }
  set action_type(value) {
    this._refObject.action_type = value;
  }
  get polygon_name() {
    return this._wrapperFields.polygon_name.data;
  }
  set polygon_name(value) {
    this._wrapperFields.polygon_name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CollisionMonitorStateRefStruct(refObject.toObject());
    this._wrapperFields.polygon_name.copyRefObject(this._refObject.polygon_name);
  }
  copy(other) {
    this._refObject = new CollisionMonitorStateRefStruct(other._refObject.toObject());
    this._wrapperFields.polygon_name.copy(other._wrapperFields.polygon_name);
  }
  static get classType() {
    return CollisionMonitorStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"DO_NOTHING","value":0},{"type":"uint8","name":"STOP","value":1},{"type":"uint8","name":"SLOWDOWN","value":2},{"type":"uint8","name":"APPROACH","value":3}],"fields":[{"name":"action_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"polygon_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"CollisionMonitorState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CollisionMonitorState"};
  }
  hasMember(name) {
    let memberNames = ["action_type","polygon_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
class CollisionMonitorStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CollisionMonitorStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CollisionMonitorStateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CollisionMonitorStateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CollisionMonitorStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new CollisionMonitorStateRefArray(size);
    this._refObject = new CollisionMonitorStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CollisionMonitorStateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CollisionMonitorStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CollisionMonitorStateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CollisionMonitorStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CollisionMonitorStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CollisionMonitorStateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(CollisionMonitorStateWrapper, "DO_NOTHING", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CollisionMonitorStateWrapper, "STOP", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CollisionMonitorStateWrapper, "SLOWDOWN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CollisionMonitorStateWrapper, "APPROACH", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = CollisionMonitorStateWrapper;
