// rclnodejs/generated/nav2_msgs/nav2_msgs__action__ComputeAndTrackRoute_Feedback.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RouteWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__Route.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ComputeAndTrackRoute_FeedbackRefStruct = StructType({
  last_node_id: primitiveTypes.uint16,
  next_node_id: primitiveTypes.uint16,
  current_edge_id: primitiveTypes.uint16,
  route: RouteWrapper.refObjectType,
  path: PathWrapper.refObjectType,
  operations_triggered: StringWrapper.refObjectArrayType,
  rerouted: primitiveTypes.bool,
});
const ComputeAndTrackRoute_FeedbackRefArray = ArrayType(ComputeAndTrackRoute_FeedbackRefStruct);
const ComputeAndTrackRoute_FeedbackRefStructArray = StructType({
  data: ComputeAndTrackRoute_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ComputeAndTrackRoute_FeedbackWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ComputeAndTrackRoute_FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.route =  new RouteWrapper(other._wrapperFields.route);
      this._wrapperFields.path =  new PathWrapper(other._wrapperFields.path);
      this._wrapperFields.operations_triggered = StringWrapper.createArray();
      this._wrapperFields.operations_triggered.copy(other._wrapperFields.operations_triggered);
      this.operations_triggered = other.operations_triggered;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ComputeAndTrackRoute_FeedbackRefStruct();
    this._wrapperFields.route = new RouteWrapper();
    this._wrapperFields.path = new PathWrapper();
    this._wrapperFields.operations_triggered = StringWrapper.createArray();
    this._operations_triggeredArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new ComputeAndTrackRoute_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ComputeAndTrackRoute_FeedbackArrayWrapper();
  }
  static get ArrayType() {
    return ComputeAndTrackRoute_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return ComputeAndTrackRoute_FeedbackRefStructArray;
  }
  static get refObjectType() {
    return ComputeAndTrackRoute_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.route.freeze(own);
    this._refObject.route = this._wrapperFields.route.refObject;
    this._wrapperFields.path.freeze(own);
    this._refObject.path = this._wrapperFields.path.refObject;
    if (own) {
      this._wrapperFields.operations_triggered.fill([]);
    } else {
      this._wrapperFields.operations_triggered.fill(this._operations_triggeredArray);
    }
    this._wrapperFields.operations_triggered.freeze(own);
    this._refObject.operations_triggered = this._wrapperFields.operations_triggered.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.last_node_id = refObject.last_node_id;
    this._refObject.next_node_id = refObject.next_node_id;
    this._refObject.current_edge_id = refObject.current_edge_id;
    this._wrapperFields.route.copyRefObject(refObject.route);
    this._wrapperFields.path.copyRefObject(refObject.path);
    refObject.operations_triggered.data.length = refObject.operations_triggered.size;
    for (let index = 0; index < refObject.operations_triggered.size; index++) {
      this._operations_triggeredArray[index] = refObject.operations_triggered.data[index].data;
    }
    this._refObject.rerouted = refObject.rerouted;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RouteWrapper.freeStruct(refObject.route);
    PathWrapper.freeStruct(refObject.path);
    if (refObject.operations_triggered.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.operations_triggered);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.operations_triggered, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ComputeAndTrackRoute_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'ComputeAndTrackRoute_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get last_node_id() {
    return this._refObject.last_node_id;
  }
  set last_node_id(value) {
    this._refObject.last_node_id = value;
  }
  get next_node_id() {
    return this._refObject.next_node_id;
  }
  set next_node_id(value) {
    this._refObject.next_node_id = value;
  }
  get current_edge_id() {
    return this._refObject.current_edge_id;
  }
  set current_edge_id(value) {
    this._refObject.current_edge_id = value;
  }
  get route() {
    return this._wrapperFields.route;
  }
  set route(value) {
    if (value instanceof RouteWrapper) {
      this._wrapperFields.route.copy(value);
    } else {
      this._wrapperFields.route.copy(new RouteWrapper(value));
    }
  }
  get path() {
    return this._wrapperFields.path;
  }
  set path(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.path.copy(value);
    } else {
      this._wrapperFields.path.copy(new PathWrapper(value));
    }
  }
  get operations_triggered() {
    return this._operations_triggeredArray;
  }
  set operations_triggered(value) {
    this._operations_triggeredArray = value;
  }
  get rerouted() {
    return this._refObject.rerouted;
  }
  set rerouted(value) {
    this._refObject.rerouted = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ComputeAndTrackRoute_FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.route.copyRefObject(this._refObject.route);
    this._wrapperFields.path.copyRefObject(this._refObject.path);
    refObject.operations_triggered.data.length = refObject.operations_triggered.size;
    for (let index = 0; index < refObject.operations_triggered.size; index++) {
      this._operations_triggeredArray[index] = refObject.operations_triggered.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ComputeAndTrackRoute_FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.route.copy(other._wrapperFields.route);
    this._wrapperFields.path.copy(other._wrapperFields.path);
    this._operations_triggeredArray = other._operations_triggeredArray.slice();
  }
  static get classType() {
    return ComputeAndTrackRoute_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"last_node_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"next_node_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_edge_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"route","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav2_msgs","type":"Route","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"operations_triggered","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rerouted","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ComputeAndTrackRoute_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ComputeAndTrackRoute_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["last_node_id","next_node_id","current_edge_id","route","path","operations_triggered","rerouted"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ComputeAndTrackRoute_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ComputeAndTrackRoute_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ComputeAndTrackRoute_FeedbackWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeAndTrackRoute_FeedbackArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeAndTrackRoute_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ComputeAndTrackRoute_FeedbackRefArray(size);
    this._refObject = new ComputeAndTrackRoute_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ComputeAndTrackRoute_FeedbackWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ComputeAndTrackRoute_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ComputeAndTrackRoute_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ComputeAndTrackRoute_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ComputeAndTrackRoute_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ComputeAndTrackRoute_FeedbackArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ComputeAndTrackRoute_FeedbackWrapper;
