// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__IsPathValid_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const IsPathValid_ResponseRefStruct = StructType({
  is_valid: primitiveTypes.bool,
  invalid_pose_indices: Int32Wrapper.refObjectArrayType,
});
const IsPathValid_ResponseRefArray = ArrayType(IsPathValid_ResponseRefStruct);
const IsPathValid_ResponseRefStructArray = StructType({
  data: IsPathValid_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class IsPathValid_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new IsPathValid_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.invalid_pose_indices = Int32Wrapper.createArray();
      this._wrapperFields.invalid_pose_indices.copy(other._wrapperFields.invalid_pose_indices);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new IsPathValid_ResponseRefStruct();
    this._wrapperFields.invalid_pose_indices = Int32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new IsPathValid_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IsPathValid_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return IsPathValid_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return IsPathValid_ResponseRefStructArray;
  }
  static get refObjectType() {
    return IsPathValid_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.invalid_pose_indices.fill(Int32Array.from([]));
    }
    this._wrapperFields.invalid_pose_indices.freeze(own);
    this._refObject.invalid_pose_indices = this._wrapperFields.invalid_pose_indices.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.is_valid = refObject.is_valid;
    this._wrapperFields.invalid_pose_indices.copyRefObject(refObject.invalid_pose_indices);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.invalid_pose_indices.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.invalid_pose_indices);
      if (!Int32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.invalid_pose_indices, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    IsPathValid_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'IsPathValid_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get is_valid() {
    return this._refObject.is_valid;
  }
  set is_valid(value) {
    this._refObject.is_valid = value;
  }
  get invalid_pose_indices() {
    return this._wrapperFields['invalid_pose_indices'].data;
  }
  set invalid_pose_indices(value) {
    this._wrapperFields['invalid_pose_indices'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new IsPathValid_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.invalid_pose_indices.copyRefObject(this._refObject.invalid_pose_indices);
  }
  copy(other) {
    this._refObject = new IsPathValid_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.invalid_pose_indices.copy(other._wrapperFields.invalid_pose_indices);
  }
  static get classType() {
    return IsPathValid_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"is_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"invalid_pose_indices","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"IsPathValid_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IsPathValid_Response"};
  }
  hasMember(name) {
    let memberNames = ["is_valid","invalid_pose_indices"];
    return memberNames.indexOf(name) !== -1;
  }
}
class IsPathValid_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IsPathValid_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IsPathValid_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IsPathValid_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IsPathValid_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new IsPathValid_ResponseRefArray(size);
    this._refObject = new IsPathValid_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IsPathValid_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IsPathValid_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IsPathValid_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IsPathValid_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IsPathValid_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IsPathValid_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = IsPathValid_ResponseWrapper;
