// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__ClearEntireCostmap_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EmptyWrapper = require('../../generated/std_msgs/std_msgs__msg__Empty.js');
const ClearEntireCostmap_RequestRefStruct = StructType({
  request: EmptyWrapper.refObjectType,
});
const ClearEntireCostmap_RequestRefArray = ArrayType(ClearEntireCostmap_RequestRefStruct);
const ClearEntireCostmap_RequestRefStructArray = StructType({
  data: ClearEntireCostmap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ClearEntireCostmap_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ClearEntireCostmap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.request =  new EmptyWrapper(other._wrapperFields.request);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ClearEntireCostmap_RequestRefStruct();
    this._wrapperFields.request = new EmptyWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ClearEntireCostmap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ClearEntireCostmap_RequestArrayWrapper();
  }
  static get ArrayType() {
    return ClearEntireCostmap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ClearEntireCostmap_RequestRefStructArray;
  }
  static get refObjectType() {
    return ClearEntireCostmap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.request.freeze(own);
    this._refObject.request = this._wrapperFields.request.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.request.copyRefObject(refObject.request);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    EmptyWrapper.freeStruct(refObject.request);
  }
  static destroyRawROS(msg) {
    ClearEntireCostmap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'ClearEntireCostmap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get request() {
    return this._wrapperFields.request;
  }
  set request(value) {
    if (value instanceof EmptyWrapper) {
      this._wrapperFields.request.copy(value);
    } else {
      this._wrapperFields.request.copy(new EmptyWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ClearEntireCostmap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.request.copyRefObject(this._refObject.request);
  }
  copy(other) {
    this._refObject = new ClearEntireCostmap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.request.copy(other._wrapperFields.request);
  }
  static get classType() {
    return ClearEntireCostmap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"request","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Empty","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ClearEntireCostmap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ClearEntireCostmap_Request"};
  }
  hasMember(name) {
    let memberNames = ["request"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ClearEntireCostmap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ClearEntireCostmap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ClearEntireCostmap_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ClearEntireCostmap_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ClearEntireCostmap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ClearEntireCostmap_RequestRefArray(size);
    this._refObject = new ClearEntireCostmap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ClearEntireCostmap_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ClearEntireCostmap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ClearEntireCostmap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ClearEntireCostmap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ClearEntireCostmap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ClearEntireCostmap_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ClearEntireCostmap_RequestWrapper;
