// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__SaveMap_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SaveMap_RequestRefStruct = StructType({
  map_topic: primitiveTypes.string,
  map_url: primitiveTypes.string,
  image_format: primitiveTypes.string,
  map_mode: primitiveTypes.string,
  free_thresh: primitiveTypes.float32,
  occupied_thresh: primitiveTypes.float32,
});
const SaveMap_RequestRefArray = ArrayType(SaveMap_RequestRefStruct);
const SaveMap_RequestRefStructArray = StructType({
  data: SaveMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SaveMap_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SaveMap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.map_topic =  new StringWrapper(other._wrapperFields.map_topic);
      this._wrapperFields.map_url =  new StringWrapper(other._wrapperFields.map_url);
      this._wrapperFields.image_format =  new StringWrapper(other._wrapperFields.image_format);
      this._wrapperFields.map_mode =  new StringWrapper(other._wrapperFields.map_mode);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SaveMap_RequestRefStruct();
    this._wrapperFields.map_topic = new StringWrapper();
    this._wrapperFields.map_url = new StringWrapper();
    this._wrapperFields.image_format = new StringWrapper();
    this._wrapperFields.map_mode = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SaveMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SaveMap_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SaveMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SaveMap_RequestRefStructArray;
  }
  static get refObjectType() {
    return SaveMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.map_topic.freeze(own);
    }
    this._refObject.map_topic = this._wrapperFields.map_topic.refObject;
    if (own) {
      this._wrapperFields.map_url.freeze(own);
    }
    this._refObject.map_url = this._wrapperFields.map_url.refObject;
    if (own) {
      this._wrapperFields.image_format.freeze(own);
    }
    this._refObject.image_format = this._wrapperFields.image_format.refObject;
    if (own) {
      this._wrapperFields.map_mode.freeze(own);
    }
    this._refObject.map_mode = this._wrapperFields.map_mode.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map_topic.data = refObject.map_topic.data;
    this._wrapperFields.map_url.data = refObject.map_url.data;
    this._wrapperFields.image_format.data = refObject.image_format.data;
    this._wrapperFields.map_mode.data = refObject.map_mode.data;
    this._refObject.free_thresh = refObject.free_thresh;
    this._refObject.occupied_thresh = refObject.occupied_thresh;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.map_topic);
    StringWrapper.freeStruct(refObject.map_url);
    StringWrapper.freeStruct(refObject.image_format);
    StringWrapper.freeStruct(refObject.map_mode);
  }
  static destroyRawROS(msg) {
    SaveMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'SaveMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map_topic() {
    return this._wrapperFields.map_topic.data;
  }
  set map_topic(value) {
    this._wrapperFields.map_topic.data = value;
  }
  get map_url() {
    return this._wrapperFields.map_url.data;
  }
  set map_url(value) {
    this._wrapperFields.map_url.data = value;
  }
  get image_format() {
    return this._wrapperFields.image_format.data;
  }
  set image_format(value) {
    this._wrapperFields.image_format.data = value;
  }
  get map_mode() {
    return this._wrapperFields.map_mode.data;
  }
  set map_mode(value) {
    this._wrapperFields.map_mode.data = value;
  }
  get free_thresh() {
    return this._refObject.free_thresh;
  }
  set free_thresh(value) {
    this._refObject.free_thresh = value;
  }
  get occupied_thresh() {
    return this._refObject.occupied_thresh;
  }
  set occupied_thresh(value) {
    this._refObject.occupied_thresh = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SaveMap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.map_topic.copyRefObject(this._refObject.map_topic);
    this._wrapperFields.map_url.copyRefObject(this._refObject.map_url);
    this._wrapperFields.image_format.copyRefObject(this._refObject.image_format);
    this._wrapperFields.map_mode.copyRefObject(this._refObject.map_mode);
  }
  copy(other) {
    this._refObject = new SaveMap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.map_topic.copy(other._wrapperFields.map_topic);
    this._wrapperFields.map_url.copy(other._wrapperFields.map_url);
    this._wrapperFields.image_format.copy(other._wrapperFields.image_format);
    this._wrapperFields.map_mode.copy(other._wrapperFields.map_mode);
  }
  static get classType() {
    return SaveMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map_topic","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"map_url","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"image_format","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"map_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"free_thresh","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"occupied_thresh","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"SaveMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SaveMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["map_topic","map_url","image_format","map_mode","free_thresh","occupied_thresh"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SaveMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SaveMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SaveMap_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SaveMap_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SaveMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SaveMap_RequestRefArray(size);
    this._refObject = new SaveMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SaveMap_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SaveMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SaveMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SaveMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SaveMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SaveMap_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SaveMap_RequestWrapper;
