// rclnodejs/generated/bond/bond__msg__Constants.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ConstantsRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const ConstantsRefArray = ArrayType(ConstantsRefStruct);
const ConstantsRefStructArray = StructType({
  data: ConstantsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ConstantsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ConstantsRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ConstantsRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new ConstantsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ConstantsArrayWrapper();
  }
  static get ArrayType() {
    return ConstantsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ConstantsRefStructArray;
  }
  static get refObjectType() {
    return ConstantsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ConstantsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'bond', subFolder: 'msg', interfaceName: 'Constants'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ConstantsRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new ConstantsRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return ConstantsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"float32","name":"DEAD_PUBLISH_PERIOD","value":0.05},{"type":"float32","name":"DEFAULT_CONNECT_TIMEOUT","value":10},{"type":"float32","name":"DEFAULT_HEARTBEAT_TIMEOUT","value":4},{"type":"float32","name":"DEFAULT_DISCONNECT_TIMEOUT","value":2},{"type":"float32","name":"DEFAULT_HEARTBEAT_PERIOD","value":1},{"type":"string","name":"DISABLE_HEARTBEAT_TIMEOUT_PARAM","value":"/bond_disable_heartbeat_timeout"}],"fields":[],"baseType":{"pkgName":"bond","type":"Constants","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Constants"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ConstantsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ConstantsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ConstantsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConstantsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConstantsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ConstantsRefArray(size);
    this._refObject = new ConstantsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ConstantsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ConstantsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ConstantsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ConstantsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ConstantsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ConstantsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ConstantsWrapper, "DEAD_PUBLISH_PERIOD", {value: 0.05, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ConstantsWrapper, "DEFAULT_CONNECT_TIMEOUT", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ConstantsWrapper, "DEFAULT_HEARTBEAT_TIMEOUT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ConstantsWrapper, "DEFAULT_DISCONNECT_TIMEOUT", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ConstantsWrapper, "DEFAULT_HEARTBEAT_PERIOD", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ConstantsWrapper, "DISABLE_HEARTBEAT_TIMEOUT_PARAM", {value: "/bond_disable_heartbeat_timeout", writable: false, enumerable: true, configurable: true});
module.exports = ConstantsWrapper;
