// rclnodejs/generated/grid_map_msgs/grid_map_msgs__srv__GetGridMap_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetGridMap_RequestRefStruct = StructType({
  frame_id: primitiveTypes.string,
  position_x: primitiveTypes.float64,
  position_y: primitiveTypes.float64,
  length_x: primitiveTypes.float64,
  length_y: primitiveTypes.float64,
  layers: StringWrapper.refObjectArrayType,
});
const GetGridMap_RequestRefArray = ArrayType(GetGridMap_RequestRefStruct);
const GetGridMap_RequestRefStructArray = StructType({
  data: GetGridMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetGridMap_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetGridMap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
      this._wrapperFields.layers = StringWrapper.createArray();
      this._wrapperFields.layers.copy(other._wrapperFields.layers);
      this.layers = other.layers;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetGridMap_RequestRefStruct();
    this._wrapperFields.frame_id = new StringWrapper();
    this._wrapperFields.layers = StringWrapper.createArray();
    this._layersArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new GetGridMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetGridMap_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetGridMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetGridMap_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetGridMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.frame_id.freeze(own);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
    if (own) {
      this._wrapperFields.layers.fill([]);
    } else {
      this._wrapperFields.layers.fill(this._layersArray);
    }
    this._wrapperFields.layers.freeze(own);
    this._refObject.layers = this._wrapperFields.layers.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
    this._refObject.position_x = refObject.position_x;
    this._refObject.position_y = refObject.position_y;
    this._refObject.length_x = refObject.length_x;
    this._refObject.length_y = refObject.length_y;
    refObject.layers.data.length = refObject.layers.size;
    for (let index = 0; index < refObject.layers.size; index++) {
      this._layersArray[index] = refObject.layers.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.frame_id);
    if (refObject.layers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.layers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.layers, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GetGridMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'grid_map_msgs', subFolder: 'srv', interfaceName: 'GetGridMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frame_id() {
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._wrapperFields.frame_id.data = value;
  }
  get position_x() {
    return this._refObject.position_x;
  }
  set position_x(value) {
    this._refObject.position_x = value;
  }
  get position_y() {
    return this._refObject.position_y;
  }
  set position_y(value) {
    this._refObject.position_y = value;
  }
  get length_x() {
    return this._refObject.length_x;
  }
  set length_x(value) {
    this._refObject.length_x = value;
  }
  get length_y() {
    return this._refObject.length_y;
  }
  set length_y(value) {
    this._refObject.length_y = value;
  }
  get layers() {
    return this._layersArray;
  }
  set layers(value) {
    this._layersArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetGridMap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
    refObject.layers.data.length = refObject.layers.size;
    for (let index = 0; index < refObject.layers.size; index++) {
      this._layersArray[index] = refObject.layers.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new GetGridMap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
    this._layersArray = other._layersArray.slice();
  }
  static get classType() {
    return GetGridMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"length_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"length_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"layers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"grid_map_msgs","type":"GetGridMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetGridMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["frame_id","position_x","position_y","length_x","length_y","layers"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetGridMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetGridMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetGridMap_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGridMap_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGridMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetGridMap_RequestRefArray(size);
    this._refObject = new GetGridMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetGridMap_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetGridMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetGridMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetGridMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetGridMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetGridMap_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetGridMap_RequestWrapper;
