// rclnodejs/generated/grid_map_msgs/grid_map_msgs__srv__SetGridMap_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GridMapWrapper = require('../../generated/grid_map_msgs/grid_map_msgs__msg__GridMap.js');
const SetGridMap_RequestRefStruct = StructType({
  map: GridMapWrapper.refObjectType,
});
const SetGridMap_RequestRefArray = ArrayType(SetGridMap_RequestRefStruct);
const SetGridMap_RequestRefStructArray = StructType({
  data: SetGridMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetGridMap_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetGridMap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.map =  new GridMapWrapper(other._wrapperFields.map);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetGridMap_RequestRefStruct();
    this._wrapperFields.map = new GridMapWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetGridMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetGridMap_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetGridMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetGridMap_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetGridMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.map.freeze(own);
    this._refObject.map = this._wrapperFields.map.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map.copyRefObject(refObject.map);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GridMapWrapper.freeStruct(refObject.map);
  }
  static destroyRawROS(msg) {
    SetGridMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'grid_map_msgs', subFolder: 'srv', interfaceName: 'SetGridMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map() {
    return this._wrapperFields.map;
  }
  set map(value) {
    if (value instanceof GridMapWrapper) {
      this._wrapperFields.map.copy(value);
    } else {
      this._wrapperFields.map.copy(new GridMapWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetGridMap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.map.copyRefObject(this._refObject.map);
  }
  copy(other) {
    this._refObject = new SetGridMap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.map.copy(other._wrapperFields.map);
  }
  static get classType() {
    return SetGridMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"grid_map_msgs","type":"GridMap","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"grid_map_msgs","type":"SetGridMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetGridMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["map"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetGridMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetGridMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetGridMap_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGridMap_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGridMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetGridMap_RequestRefArray(size);
    this._refObject = new SetGridMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetGridMap_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetGridMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetGridMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetGridMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetGridMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetGridMap_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetGridMap_RequestWrapper;
