// rclnodejs/generated/grid_map_msgs/grid_map_msgs__msg__GridMap.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GridMapInfoWrapper = require('../../generated/grid_map_msgs/grid_map_msgs__msg__GridMapInfo.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float32MultiArrayWrapper = require('../../generated/std_msgs/std_msgs__msg__Float32MultiArray.js');
const GridMapRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  info: GridMapInfoWrapper.refObjectType,
  layers: StringWrapper.refObjectArrayType,
  basic_layers: StringWrapper.refObjectArrayType,
  data: Float32MultiArrayWrapper.refObjectArrayType,
  outer_start_index: primitiveTypes.uint16,
  inner_start_index: primitiveTypes.uint16,
});
const GridMapRefArray = ArrayType(GridMapRefStruct);
const GridMapRefStructArray = StructType({
  data: GridMapRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GridMapWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GridMapRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.info =  new GridMapInfoWrapper(other._wrapperFields.info);
      this._wrapperFields.layers = StringWrapper.createArray();
      this._wrapperFields.layers.copy(other._wrapperFields.layers);
      this.layers = other.layers;
      this._wrapperFields.basic_layers = StringWrapper.createArray();
      this._wrapperFields.basic_layers.copy(other._wrapperFields.basic_layers);
      this.basic_layers = other.basic_layers;
      this._wrapperFields.data = Float32MultiArrayWrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GridMapRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.info = new GridMapInfoWrapper();
    this._wrapperFields.layers = StringWrapper.createArray();
    this._layersArray = [];
    this._wrapperFields.basic_layers = StringWrapper.createArray();
    this._basic_layersArray = [];
    this._wrapperFields.data = Float32MultiArrayWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GridMapWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GridMapArrayWrapper();
  }
  static get ArrayType() {
    return GridMapArrayWrapper;
  }
  static get refObjectArrayType() {
    return GridMapRefStructArray;
  }
  static get refObjectType() {
    return GridMapRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.info.freeze(own);
    this._refObject.info = this._wrapperFields.info.refObject;
    if (own) {
      this._wrapperFields.layers.fill([]);
    } else {
      this._wrapperFields.layers.fill(this._layersArray);
    }
    this._wrapperFields.layers.freeze(own);
    this._refObject.layers = this._wrapperFields.layers.refObject;
    if (own) {
      this._wrapperFields.basic_layers.fill([]);
    } else {
      this._wrapperFields.basic_layers.fill(this._basic_layersArray);
    }
    this._wrapperFields.basic_layers.freeze(own);
    this._refObject.basic_layers = this._wrapperFields.basic_layers.refObject;
    if (own) {
      this._wrapperFields.data.fill([]);
    }
    this._wrapperFields.data.freeze(own);
    this._refObject.data = this._wrapperFields.data.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.info.copyRefObject(refObject.info);
    refObject.layers.data.length = refObject.layers.size;
    for (let index = 0; index < refObject.layers.size; index++) {
      this._layersArray[index] = refObject.layers.data[index].data;
    }
    refObject.basic_layers.data.length = refObject.basic_layers.size;
    for (let index = 0; index < refObject.basic_layers.size; index++) {
      this._basic_layersArray[index] = refObject.basic_layers.data[index].data;
    }
    this._wrapperFields.data.copyRefObject(refObject.data);
    this._refObject.outer_start_index = refObject.outer_start_index;
    this._refObject.inner_start_index = refObject.inner_start_index;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    GridMapInfoWrapper.freeStruct(refObject.info);
    if (refObject.layers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.layers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.layers, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.basic_layers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.basic_layers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.basic_layers, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.data.size != 0) {
      Float32MultiArrayWrapper.ArrayType.freeArray(refObject.data);
      if (!Float32MultiArrayWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.data, Float32MultiArrayWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GridMapWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'grid_map_msgs', subFolder: 'msg', interfaceName: 'GridMap'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get info() {
    return this._wrapperFields.info;
  }
  set info(value) {
    if (value instanceof GridMapInfoWrapper) {
      this._wrapperFields.info.copy(value);
    } else {
      this._wrapperFields.info.copy(new GridMapInfoWrapper(value));
    }
  }
  get layers() {
    return this._layersArray;
  }
  set layers(value) {
    this._layersArray = value;
  }
  get basic_layers() {
    return this._basic_layersArray;
  }
  set basic_layers(value) {
    this._basic_layersArray = value;
  }
  get data() {
    return this._wrapperFields.data;
  }
  set data(value) {
    this._wrapperFields.data.fill(value);
  }
  get outer_start_index() {
    return this._refObject.outer_start_index;
  }
  set outer_start_index(value) {
    this._refObject.outer_start_index = value;
  }
  get inner_start_index() {
    return this._refObject.inner_start_index;
  }
  set inner_start_index(value) {
    this._refObject.inner_start_index = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GridMapRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.info.copyRefObject(this._refObject.info);
    refObject.layers.data.length = refObject.layers.size;
    for (let index = 0; index < refObject.layers.size; index++) {
      this._layersArray[index] = refObject.layers.data[index].data;
    }
    refObject.basic_layers.data.length = refObject.basic_layers.size;
    for (let index = 0; index < refObject.basic_layers.size; index++) {
      this._basic_layersArray[index] = refObject.basic_layers.data[index].data;
    }
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new GridMapRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.info.copy(other._wrapperFields.info);
    this._layersArray = other._layersArray.slice();
    this._basic_layersArray = other._basic_layersArray.slice();
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return GridMapWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"grid_map_msgs","type":"GridMapInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"layers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"basic_layers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"std_msgs","type":"Float32MultiArray","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"outer_start_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"inner_start_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"grid_map_msgs","type":"GridMap","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GridMap"};
  }
  hasMember(name) {
    let memberNames = ["header","info","layers","basic_layers","data","outer_start_index","inner_start_index"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GridMapArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GridMapWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GridMapWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GridMapArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GridMapArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GridMapRefArray(size);
    this._refObject = new GridMapRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GridMapWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GridMapArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GridMapArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GridMapWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GridMapWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GridMapArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GridMapWrapper;
