// rclnodejs/generated/grid_map_msgs/grid_map_msgs__srv__ProcessFile_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ProcessFile_RequestRefStruct = StructType({
  file_path: primitiveTypes.string,
  topic_name: primitiveTypes.string,
});
const ProcessFile_RequestRefArray = ArrayType(ProcessFile_RequestRefStruct);
const ProcessFile_RequestRefStructArray = StructType({
  data: ProcessFile_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ProcessFile_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ProcessFile_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.file_path =  new StringWrapper(other._wrapperFields.file_path);
      this._wrapperFields.topic_name =  new StringWrapper(other._wrapperFields.topic_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ProcessFile_RequestRefStruct();
    this._wrapperFields.file_path = new StringWrapper();
    this._wrapperFields.topic_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ProcessFile_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ProcessFile_RequestArrayWrapper();
  }
  static get ArrayType() {
    return ProcessFile_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ProcessFile_RequestRefStructArray;
  }
  static get refObjectType() {
    return ProcessFile_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.file_path.freeze(own);
    }
    this._refObject.file_path = this._wrapperFields.file_path.refObject;
    if (own) {
      this._wrapperFields.topic_name.freeze(own);
    }
    this._refObject.topic_name = this._wrapperFields.topic_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.file_path.data = refObject.file_path.data;
    this._wrapperFields.topic_name.data = refObject.topic_name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.file_path);
    StringWrapper.freeStruct(refObject.topic_name);
  }
  static destroyRawROS(msg) {
    ProcessFile_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'grid_map_msgs', subFolder: 'srv', interfaceName: 'ProcessFile_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get file_path() {
    return this._wrapperFields.file_path.data;
  }
  set file_path(value) {
    this._wrapperFields.file_path.data = value;
  }
  get topic_name() {
    return this._wrapperFields.topic_name.data;
  }
  set topic_name(value) {
    this._wrapperFields.topic_name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ProcessFile_RequestRefStruct(refObject.toObject());
    this._wrapperFields.file_path.copyRefObject(this._refObject.file_path);
    this._wrapperFields.topic_name.copyRefObject(this._refObject.topic_name);
  }
  copy(other) {
    this._refObject = new ProcessFile_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.file_path.copy(other._wrapperFields.file_path);
    this._wrapperFields.topic_name.copy(other._wrapperFields.topic_name);
  }
  static get classType() {
    return ProcessFile_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"file_path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"topic_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"grid_map_msgs","type":"ProcessFile_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ProcessFile_Request"};
  }
  hasMember(name) {
    let memberNames = ["file_path","topic_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ProcessFile_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ProcessFile_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ProcessFile_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProcessFile_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProcessFile_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ProcessFile_RequestRefArray(size);
    this._refObject = new ProcessFile_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ProcessFile_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ProcessFile_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ProcessFile_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ProcessFile_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ProcessFile_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ProcessFile_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ProcessFile_RequestWrapper;
