// rclnodejs/generated/grid_map_msgs/grid_map_msgs__srv__GetGridMap_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GridMapWrapper = require('../../generated/grid_map_msgs/grid_map_msgs__msg__GridMap.js');
const GetGridMap_ResponseRefStruct = StructType({
  map: GridMapWrapper.refObjectType,
});
const GetGridMap_ResponseRefArray = ArrayType(GetGridMap_ResponseRefStruct);
const GetGridMap_ResponseRefStructArray = StructType({
  data: GetGridMap_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetGridMap_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetGridMap_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.map =  new GridMapWrapper(other._wrapperFields.map);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetGridMap_ResponseRefStruct();
    this._wrapperFields.map = new GridMapWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetGridMap_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetGridMap_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetGridMap_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetGridMap_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetGridMap_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.map.freeze(own);
    this._refObject.map = this._wrapperFields.map.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map.copyRefObject(refObject.map);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GridMapWrapper.freeStruct(refObject.map);
  }
  static destroyRawROS(msg) {
    GetGridMap_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'grid_map_msgs', subFolder: 'srv', interfaceName: 'GetGridMap_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map() {
    return this._wrapperFields.map;
  }
  set map(value) {
    if (value instanceof GridMapWrapper) {
      this._wrapperFields.map.copy(value);
    } else {
      this._wrapperFields.map.copy(new GridMapWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetGridMap_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.map.copyRefObject(this._refObject.map);
  }
  copy(other) {
    this._refObject = new GetGridMap_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.map.copy(other._wrapperFields.map);
  }
  static get classType() {
    return GetGridMap_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"grid_map_msgs","type":"GridMap","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"grid_map_msgs","type":"GetGridMap_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetGridMap_Response"};
  }
  hasMember(name) {
    let memberNames = ["map"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetGridMap_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetGridMap_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetGridMap_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGridMap_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGridMap_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetGridMap_ResponseRefArray(size);
    this._refObject = new GetGridMap_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetGridMap_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetGridMap_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetGridMap_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetGridMap_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetGridMap_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetGridMap_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetGridMap_ResponseWrapper;
