// rclnodejs/generated/aruco_msgs/aruco_msgs__msg__Marker.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PoseWithCovarianceWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseWithCovariance.js');
const MarkerRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id: primitiveTypes.uint32,
  pose: PoseWithCovarianceWrapper.refObjectType,
  confidence: primitiveTypes.float64,
});
const MarkerRefArray = ArrayType(MarkerRefStruct);
const MarkerRefStructArray = StructType({
  data: MarkerRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MarkerWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MarkerRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.pose =  new PoseWithCovarianceWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MarkerRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.pose = new PoseWithCovarianceWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MarkerWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MarkerArrayWrapper();
  }
  static get ArrayType() {
    return MarkerArrayWrapper;
  }
  static get refObjectArrayType() {
    return MarkerRefStructArray;
  }
  static get refObjectType() {
    return MarkerRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.pose.freeze(own);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.id = refObject.id;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._refObject.confidence = refObject.confidence;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PoseWithCovarianceWrapper.freeStruct(refObject.pose);
  }
  static destroyRawROS(msg) {
    MarkerWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'aruco_msgs', subFolder: 'msg', interfaceName: 'Marker'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWithCovarianceWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWithCovarianceWrapper(value));
    }
  }
  get confidence() {
    return this._refObject.confidence;
  }
  set confidence(value) {
    this._refObject.confidence = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MarkerRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new MarkerRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return MarkerWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseWithCovariance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"confidence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"aruco_msgs","type":"Marker","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Marker"};
  }
  hasMember(name) {
    let memberNames = ["header","id","pose","confidence"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MarkerArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MarkerWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MarkerWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MarkerArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MarkerArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MarkerRefArray(size);
    this._refObject = new MarkerRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MarkerWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MarkerArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MarkerArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MarkerWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MarkerWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MarkerArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MarkerWrapper;
