// rclnodejs/generated/visualization_msgs/visualization_msgs__msg__InteractiveMarkerInit.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const InteractiveMarkerWrapper = require('../../generated/visualization_msgs/visualization_msgs__msg__InteractiveMarker.js');
const InteractiveMarkerInitRefStruct = StructType({
  server_id: primitiveTypes.string,
  seq_num: primitiveTypes.uint64,
  markers: InteractiveMarkerWrapper.refObjectArrayType,
});
const InteractiveMarkerInitRefArray = ArrayType(InteractiveMarkerInitRefStruct);
const InteractiveMarkerInitRefStructArray = StructType({
  data: InteractiveMarkerInitRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class InteractiveMarkerInitWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new InteractiveMarkerInitRefStruct(other._refObject.toObject());
      this._wrapperFields.server_id =  new StringWrapper(other._wrapperFields.server_id);
      this._wrapperFields.markers = InteractiveMarkerWrapper.createArray();
      this._wrapperFields.markers.copy(other._wrapperFields.markers);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new InteractiveMarkerInitRefStruct();
    this._wrapperFields.server_id = new StringWrapper();
    this._wrapperFields.markers = InteractiveMarkerWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new InteractiveMarkerInitWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new InteractiveMarkerInitArrayWrapper();
  }
  static get ArrayType() {
    return InteractiveMarkerInitArrayWrapper;
  }
  static get refObjectArrayType() {
    return InteractiveMarkerInitRefStructArray;
  }
  static get refObjectType() {
    return InteractiveMarkerInitRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.server_id.freeze(own);
    }
    this._refObject.server_id = this._wrapperFields.server_id.refObject;
    if (own) {
      this._wrapperFields.markers.fill([]);
    }
    this._wrapperFields.markers.freeze(own);
    this._refObject.markers = this._wrapperFields.markers.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.server_id.data = refObject.server_id.data;
    this._refObject.seq_num = refObject.seq_num;
    this._wrapperFields.markers.copyRefObject(refObject.markers);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.server_id);
    if (refObject.markers.size != 0) {
      InteractiveMarkerWrapper.ArrayType.freeArray(refObject.markers);
      if (!InteractiveMarkerWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.markers, InteractiveMarkerWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    InteractiveMarkerInitWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'visualization_msgs', subFolder: 'msg', interfaceName: 'InteractiveMarkerInit'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get server_id() {
    return this._wrapperFields.server_id.data;
  }
  set server_id(value) {
    this._wrapperFields.server_id.data = value;
  }
  get seq_num() {
    return BigInt(this._refObject.seq_num);
  }
  set seq_num(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('seq_num must be type of bigint');
    }
    this._refObject.seq_num = value.toString();
  }
  get markers() {
    return this._wrapperFields.markers;
  }
  set markers(value) {
    this._wrapperFields.markers.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new InteractiveMarkerInitRefStruct(refObject.toObject());
    this._wrapperFields.server_id.copyRefObject(this._refObject.server_id);
    this._wrapperFields.markers.copyRefObject(this._refObject.markers);
  }
  copy(other) {
    this._refObject = new InteractiveMarkerInitRefStruct(other._refObject.toObject());
    this._wrapperFields.server_id.copy(other._wrapperFields.server_id);
    this._wrapperFields.markers.copy(other._wrapperFields.markers);
  }
  static get classType() {
    return InteractiveMarkerInitWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"server_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"seq_num","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"markers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"visualization_msgs","type":"InteractiveMarker","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"visualization_msgs","type":"InteractiveMarkerInit","stringUpperBound":null,"isPrimitiveType":false},"msgName":"InteractiveMarkerInit"};
  }
  hasMember(name) {
    let memberNames = ["server_id","seq_num","markers"];
    return memberNames.indexOf(name) !== -1;
  }
}
class InteractiveMarkerInitArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof InteractiveMarkerInitWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new InteractiveMarkerInitWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InteractiveMarkerInitArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InteractiveMarkerInitArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new InteractiveMarkerInitRefArray(size);
    this._refObject = new InteractiveMarkerInitRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new InteractiveMarkerInitWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof InteractiveMarkerInitArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "InteractiveMarkerInitArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      InteractiveMarkerInitWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return InteractiveMarkerInitWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return InteractiveMarkerInitArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = InteractiveMarkerInitWrapper;
