// rclnodejs/generated/visualization_msgs/visualization_msgs__msg__MenuEntry.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MenuEntryRefStruct = StructType({
  id: primitiveTypes.uint32,
  parent_id: primitiveTypes.uint32,
  title: primitiveTypes.string,
  command: primitiveTypes.string,
  command_type: primitiveTypes.uint8,
});
const MenuEntryRefArray = ArrayType(MenuEntryRefStruct);
const MenuEntryRefStructArray = StructType({
  data: MenuEntryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MenuEntryWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MenuEntryRefStruct(other._refObject.toObject());
      this._wrapperFields.title =  new StringWrapper(other._wrapperFields.title);
      this._wrapperFields.command =  new StringWrapper(other._wrapperFields.command);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MenuEntryRefStruct();
    this._wrapperFields.title = new StringWrapper();
    this._wrapperFields.command = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MenuEntryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MenuEntryArrayWrapper();
  }
  static get ArrayType() {
    return MenuEntryArrayWrapper;
  }
  static get refObjectArrayType() {
    return MenuEntryRefStructArray;
  }
  static get refObjectType() {
    return MenuEntryRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.title.freeze(own);
    }
    this._refObject.title = this._wrapperFields.title.refObject;
    if (own) {
      this._wrapperFields.command.freeze(own);
    }
    this._refObject.command = this._wrapperFields.command.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.id = refObject.id;
    this._refObject.parent_id = refObject.parent_id;
    this._wrapperFields.title.data = refObject.title.data;
    this._wrapperFields.command.data = refObject.command.data;
    this._refObject.command_type = refObject.command_type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.title);
    StringWrapper.freeStruct(refObject.command);
  }
  static destroyRawROS(msg) {
    MenuEntryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'visualization_msgs', subFolder: 'msg', interfaceName: 'MenuEntry'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get parent_id() {
    return this._refObject.parent_id;
  }
  set parent_id(value) {
    this._refObject.parent_id = value;
  }
  get title() {
    return this._wrapperFields.title.data;
  }
  set title(value) {
    this._wrapperFields.title.data = value;
  }
  get command() {
    return this._wrapperFields.command.data;
  }
  set command(value) {
    this._wrapperFields.command.data = value;
  }
  get command_type() {
    return this._refObject.command_type;
  }
  set command_type(value) {
    this._refObject.command_type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MenuEntryRefStruct(refObject.toObject());
    this._wrapperFields.title.copyRefObject(this._refObject.title);
    this._wrapperFields.command.copyRefObject(this._refObject.command);
  }
  copy(other) {
    this._refObject = new MenuEntryRefStruct(other._refObject.toObject());
    this._wrapperFields.title.copy(other._wrapperFields.title);
    this._wrapperFields.command.copy(other._wrapperFields.command);
  }
  static get classType() {
    return MenuEntryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FEEDBACK","value":0},{"type":"uint8","name":"ROSRUN","value":1},{"type":"uint8","name":"ROSLAUNCH","value":2}],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"parent_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"title","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"visualization_msgs","type":"MenuEntry","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MenuEntry"};
  }
  hasMember(name) {
    let memberNames = ["id","parent_id","title","command","command_type"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MenuEntryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MenuEntryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MenuEntryWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MenuEntryArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MenuEntryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MenuEntryRefArray(size);
    this._refObject = new MenuEntryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MenuEntryWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MenuEntryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MenuEntryArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MenuEntryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MenuEntryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MenuEntryArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(MenuEntryWrapper, "FEEDBACK", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MenuEntryWrapper, "ROSRUN", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MenuEntryWrapper, "ROSLAUNCH", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = MenuEntryWrapper;
