// rclnodejs/generated/visualization_msgs/visualization_msgs__msg__UVCoordinate.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UVCoordinateRefStruct = StructType({
  u: primitiveTypes.float32,
  v: primitiveTypes.float32,
});
const UVCoordinateRefArray = ArrayType(UVCoordinateRefStruct);
const UVCoordinateRefStructArray = StructType({
  data: UVCoordinateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class UVCoordinateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new UVCoordinateRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new UVCoordinateRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new UVCoordinateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UVCoordinateArrayWrapper();
  }
  static get ArrayType() {
    return UVCoordinateArrayWrapper;
  }
  static get refObjectArrayType() {
    return UVCoordinateRefStructArray;
  }
  static get refObjectType() {
    return UVCoordinateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.u = refObject.u;
    this._refObject.v = refObject.v;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    UVCoordinateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'visualization_msgs', subFolder: 'msg', interfaceName: 'UVCoordinate'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get u() {
    return this._refObject.u;
  }
  set u(value) {
    this._refObject.u = value;
  }
  get v() {
    return this._refObject.v;
  }
  set v(value) {
    this._refObject.v = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UVCoordinateRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new UVCoordinateRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return UVCoordinateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"u","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"visualization_msgs","type":"UVCoordinate","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UVCoordinate"};
  }
  hasMember(name) {
    let memberNames = ["u","v"];
    return memberNames.indexOf(name) !== -1;
  }
}
class UVCoordinateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UVCoordinateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UVCoordinateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UVCoordinateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UVCoordinateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new UVCoordinateRefArray(size);
    this._refObject = new UVCoordinateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UVCoordinateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UVCoordinateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UVCoordinateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UVCoordinateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UVCoordinateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UVCoordinateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = UVCoordinateWrapper;
