// rclnodejs/generated/composition_interfaces/composition_interfaces__srv__LoadNode_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LoadNode_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  error_message: primitiveTypes.string,
  full_node_name: primitiveTypes.string,
  unique_id: primitiveTypes.uint64,
});
const LoadNode_ResponseRefArray = ArrayType(LoadNode_ResponseRefStruct);
const LoadNode_ResponseRefStructArray = StructType({
  data: LoadNode_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LoadNode_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LoadNode_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.error_message =  new StringWrapper(other._wrapperFields.error_message);
      this._wrapperFields.full_node_name =  new StringWrapper(other._wrapperFields.full_node_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LoadNode_ResponseRefStruct();
    this._wrapperFields.error_message = new StringWrapper();
    this._wrapperFields.full_node_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LoadNode_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LoadNode_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return LoadNode_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return LoadNode_ResponseRefStructArray;
  }
  static get refObjectType() {
    return LoadNode_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.error_message.freeze(own);
    }
    this._refObject.error_message = this._wrapperFields.error_message.refObject;
    if (own) {
      this._wrapperFields.full_node_name.freeze(own);
    }
    this._refObject.full_node_name = this._wrapperFields.full_node_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.success = refObject.success;
    this._wrapperFields.error_message.data = refObject.error_message.data;
    this._wrapperFields.full_node_name.data = refObject.full_node_name.data;
    this._refObject.unique_id = refObject.unique_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_message);
    StringWrapper.freeStruct(refObject.full_node_name);
  }
  static destroyRawROS(msg) {
    LoadNode_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'composition_interfaces', subFolder: 'srv', interfaceName: 'LoadNode_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  get error_message() {
    return this._wrapperFields.error_message.data;
  }
  set error_message(value) {
    this._wrapperFields.error_message.data = value;
  }
  get full_node_name() {
    return this._wrapperFields.full_node_name.data;
  }
  set full_node_name(value) {
    this._wrapperFields.full_node_name.data = value;
  }
  get unique_id() {
    return BigInt(this._refObject.unique_id);
  }
  set unique_id(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('unique_id must be type of bigint');
    }
    this._refObject.unique_id = value.toString();
  }
  copyRefObject(refObject) {
    this._refObject = new LoadNode_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.error_message.copyRefObject(this._refObject.error_message);
    this._wrapperFields.full_node_name.copyRefObject(this._refObject.full_node_name);
  }
  copy(other) {
    this._refObject = new LoadNode_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.error_message.copy(other._wrapperFields.error_message);
    this._wrapperFields.full_node_name.copy(other._wrapperFields.full_node_name);
  }
  static get classType() {
    return LoadNode_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"full_node_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"unique_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"composition_interfaces","type":"LoadNode_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LoadNode_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","error_message","full_node_name","unique_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LoadNode_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LoadNode_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LoadNode_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadNode_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadNode_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LoadNode_ResponseRefArray(size);
    this._refObject = new LoadNode_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LoadNode_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LoadNode_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LoadNode_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LoadNode_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LoadNode_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LoadNode_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = LoadNode_ResponseWrapper;
