// rclnodejs/generated/lifecycle_msgs/lifecycle_msgs__srv__GetAvailableStates_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StateWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__State.js');
const GetAvailableStates_ResponseRefStruct = StructType({
  available_states: StateWrapper.refObjectArrayType,
});
const GetAvailableStates_ResponseRefArray = ArrayType(GetAvailableStates_ResponseRefStruct);
const GetAvailableStates_ResponseRefStructArray = StructType({
  data: GetAvailableStates_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetAvailableStates_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetAvailableStates_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.available_states = StateWrapper.createArray();
      this._wrapperFields.available_states.copy(other._wrapperFields.available_states);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetAvailableStates_ResponseRefStruct();
    this._wrapperFields.available_states = StateWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetAvailableStates_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetAvailableStates_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetAvailableStates_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetAvailableStates_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetAvailableStates_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.available_states.fill([]);
    }
    this._wrapperFields.available_states.freeze(own);
    this._refObject.available_states = this._wrapperFields.available_states.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.available_states.copyRefObject(refObject.available_states);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.available_states.size != 0) {
      StateWrapper.ArrayType.freeArray(refObject.available_states);
      if (!StateWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.available_states, StateWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GetAvailableStates_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'lifecycle_msgs', subFolder: 'srv', interfaceName: 'GetAvailableStates_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get available_states() {
    return this._wrapperFields.available_states;
  }
  set available_states(value) {
    this._wrapperFields.available_states.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetAvailableStates_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.available_states.copyRefObject(this._refObject.available_states);
  }
  copy(other) {
    this._refObject = new GetAvailableStates_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.available_states.copy(other._wrapperFields.available_states);
  }
  static get classType() {
    return GetAvailableStates_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"available_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"lifecycle_msgs","type":"GetAvailableStates_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetAvailableStates_Response"};
  }
  hasMember(name) {
    let memberNames = ["available_states"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetAvailableStates_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetAvailableStates_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetAvailableStates_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetAvailableStates_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetAvailableStates_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetAvailableStates_ResponseRefArray(size);
    this._refObject = new GetAvailableStates_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetAvailableStates_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetAvailableStates_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetAvailableStates_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetAvailableStates_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetAvailableStates_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetAvailableStates_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetAvailableStates_ResponseWrapper;
