// rclnodejs/generated/lifecycle_msgs/lifecycle_msgs__srv__GetState_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StateWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__State.js');
const GetState_ResponseRefStruct = StructType({
  current_state: StateWrapper.refObjectType,
});
const GetState_ResponseRefArray = ArrayType(GetState_ResponseRefStruct);
const GetState_ResponseRefStructArray = StructType({
  data: GetState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetState_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetState_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.current_state =  new StateWrapper(other._wrapperFields.current_state);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetState_ResponseRefStruct();
    this._wrapperFields.current_state = new StateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetState_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetState_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.current_state.freeze(own);
    this._refObject.current_state = this._wrapperFields.current_state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.current_state.copyRefObject(refObject.current_state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StateWrapper.freeStruct(refObject.current_state);
  }
  static destroyRawROS(msg) {
    GetState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'lifecycle_msgs', subFolder: 'srv', interfaceName: 'GetState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get current_state() {
    return this._wrapperFields.current_state;
  }
  set current_state(value) {
    if (value instanceof StateWrapper) {
      this._wrapperFields.current_state.copy(value);
    } else {
      this._wrapperFields.current_state.copy(new StateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetState_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.current_state.copyRefObject(this._refObject.current_state);
  }
  copy(other) {
    this._refObject = new GetState_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.current_state.copy(other._wrapperFields.current_state);
  }
  static get classType() {
    return GetState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"current_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"lifecycle_msgs","type":"GetState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetState_Response"};
  }
  hasMember(name) {
    let memberNames = ["current_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetState_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetState_ResponseRefArray(size);
    this._refObject = new GetState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetState_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetState_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetState_ResponseWrapper;
