// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__action__TriggeredCalibration_Goal.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TriggeredCalibration_GoalRefStruct = StructType({
  json: primitiveTypes.string,
});
const TriggeredCalibration_GoalRefArray = ArrayType(TriggeredCalibration_GoalRefStruct);
const TriggeredCalibration_GoalRefStructArray = StructType({
  data: TriggeredCalibration_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TriggeredCalibration_GoalWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TriggeredCalibration_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.json =  new StringWrapper(other._wrapperFields.json);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
    this._refObject.json = "calib run";
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TriggeredCalibration_GoalRefStruct();
    this._wrapperFields.json = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new TriggeredCalibration_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TriggeredCalibration_GoalArrayWrapper();
  }
  static get ArrayType() {
    return TriggeredCalibration_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return TriggeredCalibration_GoalRefStructArray;
  }
  static get refObjectType() {
    return TriggeredCalibration_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.json.freeze(own);
    }
    this._refObject.json = this._wrapperFields.json.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.json.data = refObject.json.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.json);
  }
  static destroyRawROS(msg) {
    TriggeredCalibration_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'action', interfaceName: 'TriggeredCalibration_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get json() {
    return this._wrapperFields.json.data;
  }
  set json(value) {
    this._wrapperFields.json.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TriggeredCalibration_GoalRefStruct(refObject.toObject());
    this._wrapperFields.json.copyRefObject(this._refObject.json);
  }
  copy(other) {
    this._refObject = new TriggeredCalibration_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.json.copy(other._wrapperFields.json);
  }
  static get classType() {
    return TriggeredCalibration_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"json","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":"calib run"}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"TriggeredCalibration_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TriggeredCalibration_Goal"};
  }
  hasMember(name) {
    let memberNames = ["json"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TriggeredCalibration_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TriggeredCalibration_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TriggeredCalibration_GoalWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggeredCalibration_GoalArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggeredCalibration_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TriggeredCalibration_GoalRefArray(size);
    this._refObject = new TriggeredCalibration_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TriggeredCalibration_GoalWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TriggeredCalibration_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TriggeredCalibration_GoalArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TriggeredCalibration_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TriggeredCalibration_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TriggeredCalibration_GoalArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TriggeredCalibration_GoalWrapper;
