// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__DeviceInfo_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DeviceInfo_ResponseRefStruct = StructType({
  device_name: primitiveTypes.string,
  serial_number: primitiveTypes.string,
  firmware_version: primitiveTypes.string,
  usb_type_descriptor: primitiveTypes.string,
  firmware_update_id: primitiveTypes.string,
  sensors: primitiveTypes.string,
  physical_port: primitiveTypes.string,
});
const DeviceInfo_ResponseRefArray = ArrayType(DeviceInfo_ResponseRefStruct);
const DeviceInfo_ResponseRefStructArray = StructType({
  data: DeviceInfo_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class DeviceInfo_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new DeviceInfo_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.device_name =  new StringWrapper(other._wrapperFields.device_name);
      this._wrapperFields.serial_number =  new StringWrapper(other._wrapperFields.serial_number);
      this._wrapperFields.firmware_version =  new StringWrapper(other._wrapperFields.firmware_version);
      this._wrapperFields.usb_type_descriptor =  new StringWrapper(other._wrapperFields.usb_type_descriptor);
      this._wrapperFields.firmware_update_id =  new StringWrapper(other._wrapperFields.firmware_update_id);
      this._wrapperFields.sensors =  new StringWrapper(other._wrapperFields.sensors);
      this._wrapperFields.physical_port =  new StringWrapper(other._wrapperFields.physical_port);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new DeviceInfo_ResponseRefStruct();
    this._wrapperFields.device_name = new StringWrapper();
    this._wrapperFields.serial_number = new StringWrapper();
    this._wrapperFields.firmware_version = new StringWrapper();
    this._wrapperFields.usb_type_descriptor = new StringWrapper();
    this._wrapperFields.firmware_update_id = new StringWrapper();
    this._wrapperFields.sensors = new StringWrapper();
    this._wrapperFields.physical_port = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new DeviceInfo_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DeviceInfo_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return DeviceInfo_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return DeviceInfo_ResponseRefStructArray;
  }
  static get refObjectType() {
    return DeviceInfo_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.device_name.freeze(own);
    }
    this._refObject.device_name = this._wrapperFields.device_name.refObject;
    if (own) {
      this._wrapperFields.serial_number.freeze(own);
    }
    this._refObject.serial_number = this._wrapperFields.serial_number.refObject;
    if (own) {
      this._wrapperFields.firmware_version.freeze(own);
    }
    this._refObject.firmware_version = this._wrapperFields.firmware_version.refObject;
    if (own) {
      this._wrapperFields.usb_type_descriptor.freeze(own);
    }
    this._refObject.usb_type_descriptor = this._wrapperFields.usb_type_descriptor.refObject;
    if (own) {
      this._wrapperFields.firmware_update_id.freeze(own);
    }
    this._refObject.firmware_update_id = this._wrapperFields.firmware_update_id.refObject;
    if (own) {
      this._wrapperFields.sensors.freeze(own);
    }
    this._refObject.sensors = this._wrapperFields.sensors.refObject;
    if (own) {
      this._wrapperFields.physical_port.freeze(own);
    }
    this._refObject.physical_port = this._wrapperFields.physical_port.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.device_name.data = refObject.device_name.data;
    this._wrapperFields.serial_number.data = refObject.serial_number.data;
    this._wrapperFields.firmware_version.data = refObject.firmware_version.data;
    this._wrapperFields.usb_type_descriptor.data = refObject.usb_type_descriptor.data;
    this._wrapperFields.firmware_update_id.data = refObject.firmware_update_id.data;
    this._wrapperFields.sensors.data = refObject.sensors.data;
    this._wrapperFields.physical_port.data = refObject.physical_port.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.device_name);
    StringWrapper.freeStruct(refObject.serial_number);
    StringWrapper.freeStruct(refObject.firmware_version);
    StringWrapper.freeStruct(refObject.usb_type_descriptor);
    StringWrapper.freeStruct(refObject.firmware_update_id);
    StringWrapper.freeStruct(refObject.sensors);
    StringWrapper.freeStruct(refObject.physical_port);
  }
  static destroyRawROS(msg) {
    DeviceInfo_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'DeviceInfo_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get device_name() {
    return this._wrapperFields.device_name.data;
  }
  set device_name(value) {
    this._wrapperFields.device_name.data = value;
  }
  get serial_number() {
    return this._wrapperFields.serial_number.data;
  }
  set serial_number(value) {
    this._wrapperFields.serial_number.data = value;
  }
  get firmware_version() {
    return this._wrapperFields.firmware_version.data;
  }
  set firmware_version(value) {
    this._wrapperFields.firmware_version.data = value;
  }
  get usb_type_descriptor() {
    return this._wrapperFields.usb_type_descriptor.data;
  }
  set usb_type_descriptor(value) {
    this._wrapperFields.usb_type_descriptor.data = value;
  }
  get firmware_update_id() {
    return this._wrapperFields.firmware_update_id.data;
  }
  set firmware_update_id(value) {
    this._wrapperFields.firmware_update_id.data = value;
  }
  get sensors() {
    return this._wrapperFields.sensors.data;
  }
  set sensors(value) {
    this._wrapperFields.sensors.data = value;
  }
  get physical_port() {
    return this._wrapperFields.physical_port.data;
  }
  set physical_port(value) {
    this._wrapperFields.physical_port.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DeviceInfo_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.device_name.copyRefObject(this._refObject.device_name);
    this._wrapperFields.serial_number.copyRefObject(this._refObject.serial_number);
    this._wrapperFields.firmware_version.copyRefObject(this._refObject.firmware_version);
    this._wrapperFields.usb_type_descriptor.copyRefObject(this._refObject.usb_type_descriptor);
    this._wrapperFields.firmware_update_id.copyRefObject(this._refObject.firmware_update_id);
    this._wrapperFields.sensors.copyRefObject(this._refObject.sensors);
    this._wrapperFields.physical_port.copyRefObject(this._refObject.physical_port);
  }
  copy(other) {
    this._refObject = new DeviceInfo_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.device_name.copy(other._wrapperFields.device_name);
    this._wrapperFields.serial_number.copy(other._wrapperFields.serial_number);
    this._wrapperFields.firmware_version.copy(other._wrapperFields.firmware_version);
    this._wrapperFields.usb_type_descriptor.copy(other._wrapperFields.usb_type_descriptor);
    this._wrapperFields.firmware_update_id.copy(other._wrapperFields.firmware_update_id);
    this._wrapperFields.sensors.copy(other._wrapperFields.sensors);
    this._wrapperFields.physical_port.copy(other._wrapperFields.physical_port);
  }
  static get classType() {
    return DeviceInfo_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"device_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"serial_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"firmware_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"usb_type_descriptor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"firmware_update_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensors","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"physical_port","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"DeviceInfo_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DeviceInfo_Response"};
  }
  hasMember(name) {
    let memberNames = ["device_name","serial_number","firmware_version","usb_type_descriptor","firmware_update_id","sensors","physical_port"];
    return memberNames.indexOf(name) !== -1;
  }
}
class DeviceInfo_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DeviceInfo_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DeviceInfo_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DeviceInfo_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DeviceInfo_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new DeviceInfo_ResponseRefArray(size);
    this._refObject = new DeviceInfo_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DeviceInfo_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DeviceInfo_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DeviceInfo_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DeviceInfo_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DeviceInfo_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DeviceInfo_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = DeviceInfo_ResponseWrapper;
