// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__action__TriggeredCalibration_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TriggeredCalibration_ResultRefStruct = StructType({
  success: primitiveTypes.bool,
  error_msg: primitiveTypes.string,
  calibration: primitiveTypes.string,
  health: primitiveTypes.float32,
});
const TriggeredCalibration_ResultRefArray = ArrayType(TriggeredCalibration_ResultRefStruct);
const TriggeredCalibration_ResultRefStructArray = StructType({
  data: TriggeredCalibration_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TriggeredCalibration_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TriggeredCalibration_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.error_msg =  new StringWrapper(other._wrapperFields.error_msg);
      this._wrapperFields.calibration =  new StringWrapper(other._wrapperFields.calibration);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TriggeredCalibration_ResultRefStruct();
    this._wrapperFields.error_msg = new StringWrapper();
    this._wrapperFields.calibration = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new TriggeredCalibration_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TriggeredCalibration_ResultArrayWrapper();
  }
  static get ArrayType() {
    return TriggeredCalibration_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return TriggeredCalibration_ResultRefStructArray;
  }
  static get refObjectType() {
    return TriggeredCalibration_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.error_msg.freeze(own);
    }
    this._refObject.error_msg = this._wrapperFields.error_msg.refObject;
    if (own) {
      this._wrapperFields.calibration.freeze(own);
    }
    this._refObject.calibration = this._wrapperFields.calibration.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.success = refObject.success;
    this._wrapperFields.error_msg.data = refObject.error_msg.data;
    this._wrapperFields.calibration.data = refObject.calibration.data;
    this._refObject.health = refObject.health;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_msg);
    StringWrapper.freeStruct(refObject.calibration);
  }
  static destroyRawROS(msg) {
    TriggeredCalibration_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'action', interfaceName: 'TriggeredCalibration_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  get error_msg() {
    return this._wrapperFields.error_msg.data;
  }
  set error_msg(value) {
    this._wrapperFields.error_msg.data = value;
  }
  get calibration() {
    return this._wrapperFields.calibration.data;
  }
  set calibration(value) {
    this._wrapperFields.calibration.data = value;
  }
  get health() {
    return this._refObject.health;
  }
  set health(value) {
    this._refObject.health = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TriggeredCalibration_ResultRefStruct(refObject.toObject());
    this._wrapperFields.error_msg.copyRefObject(this._refObject.error_msg);
    this._wrapperFields.calibration.copyRefObject(this._refObject.calibration);
  }
  copy(other) {
    this._refObject = new TriggeredCalibration_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.error_msg.copy(other._wrapperFields.error_msg);
    this._wrapperFields.calibration.copy(other._wrapperFields.calibration);
  }
  static get classType() {
    return TriggeredCalibration_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_msg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"health","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"TriggeredCalibration_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TriggeredCalibration_Result"};
  }
  hasMember(name) {
    let memberNames = ["success","error_msg","calibration","health"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TriggeredCalibration_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TriggeredCalibration_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TriggeredCalibration_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggeredCalibration_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggeredCalibration_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TriggeredCalibration_ResultRefArray(size);
    this._refObject = new TriggeredCalibration_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TriggeredCalibration_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TriggeredCalibration_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TriggeredCalibration_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TriggeredCalibration_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TriggeredCalibration_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TriggeredCalibration_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TriggeredCalibration_ResultWrapper;
