// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__SafetyPresetRead_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SafetyPresetRead_RequestRefStruct = StructType({
  index: primitiveTypes.uint8,
});
const SafetyPresetRead_RequestRefArray = ArrayType(SafetyPresetRead_RequestRefStruct);
const SafetyPresetRead_RequestRefStructArray = StructType({
  data: SafetyPresetRead_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SafetyPresetRead_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SafetyPresetRead_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SafetyPresetRead_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new SafetyPresetRead_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SafetyPresetRead_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SafetyPresetRead_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SafetyPresetRead_RequestRefStructArray;
  }
  static get refObjectType() {
    return SafetyPresetRead_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.index = refObject.index;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    SafetyPresetRead_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'SafetyPresetRead_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get index() {
    return this._refObject.index;
  }
  set index(value) {
    this._refObject.index = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SafetyPresetRead_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new SafetyPresetRead_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return SafetyPresetRead_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"SafetyPresetRead_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SafetyPresetRead_Request"};
  }
  hasMember(name) {
    let memberNames = ["index"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SafetyPresetRead_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SafetyPresetRead_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SafetyPresetRead_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyPresetRead_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyPresetRead_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SafetyPresetRead_RequestRefArray(size);
    this._refObject = new SafetyPresetRead_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SafetyPresetRead_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SafetyPresetRead_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SafetyPresetRead_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SafetyPresetRead_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SafetyPresetRead_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SafetyPresetRead_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SafetyPresetRead_RequestWrapper;
