// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__msg__Extrinsics.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const ExtrinsicsRefStruct = StructType({
  rotation: ArrayType(primitiveTypes.float64, 9),
  translation: ArrayType(primitiveTypes.float64, 3),
});
const ExtrinsicsRefArray = ArrayType(ExtrinsicsRefStruct);
const ExtrinsicsRefStructArray = StructType({
  data: ExtrinsicsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ExtrinsicsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ExtrinsicsRefStruct(other._refObject.toObject());
      this._wrapperFields.rotation = Float64Wrapper.createArray();
      this._wrapperFields.rotation.copy(other._wrapperFields.rotation);
      this._wrapperFields.translation = Float64Wrapper.createArray();
      this._wrapperFields.translation.copy(other._wrapperFields.translation);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ExtrinsicsRefStruct();
    this._wrapperFields.rotation = Float64Wrapper.createArray();
    this._wrapperFields.translation = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ExtrinsicsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ExtrinsicsArrayWrapper();
  }
  static get ArrayType() {
    return ExtrinsicsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ExtrinsicsRefStructArray;
  }
  static get refObjectType() {
    return ExtrinsicsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._refObject.rotation = Array.from(this._wrapperFields.rotation.data);
    this._refObject.translation = Array.from(this._wrapperFields.translation.data);
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.rotation.fill(refObject.rotation.toArray());
    this._wrapperFields.translation.fill(refObject.translation.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ExtrinsicsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'msg', interfaceName: 'Extrinsics'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get rotation() {
    return this._wrapperFields['rotation'].data;
  }
  set rotation(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['rotation'].fill(value);
  }
  get translation() {
    return this._wrapperFields['translation'].data;
  }
  set translation(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['translation'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ExtrinsicsRefStruct(refObject.toObject());
    this._wrapperFields.rotation.fill(refObject.rotation.toArray());
    this._wrapperFields.translation.fill(refObject.translation.toArray());
  }
  copy(other) {
    this._refObject = new ExtrinsicsRefStruct(other._refObject.toObject());
    this._wrapperFields.rotation.copy(other._wrapperFields.rotation);
    this._wrapperFields.translation.copy(other._wrapperFields.translation);
  }
  static get classType() {
    return ExtrinsicsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"rotation","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"translation","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"Extrinsics","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Extrinsics"};
  }
  hasMember(name) {
    let memberNames = ["rotation","translation"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ExtrinsicsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ExtrinsicsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ExtrinsicsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExtrinsicsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExtrinsicsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ExtrinsicsRefArray(size);
    this._refObject = new ExtrinsicsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ExtrinsicsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ExtrinsicsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ExtrinsicsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ExtrinsicsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ExtrinsicsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ExtrinsicsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ExtrinsicsWrapper;
