// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__CalibConfigWrite_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CalibConfigWrite_RequestRefStruct = StructType({
  calib_config: primitiveTypes.string,
});
const CalibConfigWrite_RequestRefArray = ArrayType(CalibConfigWrite_RequestRefStruct);
const CalibConfigWrite_RequestRefStructArray = StructType({
  data: CalibConfigWrite_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class CalibConfigWrite_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new CalibConfigWrite_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.calib_config =  new StringWrapper(other._wrapperFields.calib_config);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new CalibConfigWrite_RequestRefStruct();
    this._wrapperFields.calib_config = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CalibConfigWrite_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CalibConfigWrite_RequestArrayWrapper();
  }
  static get ArrayType() {
    return CalibConfigWrite_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CalibConfigWrite_RequestRefStructArray;
  }
  static get refObjectType() {
    return CalibConfigWrite_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.calib_config.freeze(own);
    }
    this._refObject.calib_config = this._wrapperFields.calib_config.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.calib_config.data = refObject.calib_config.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.calib_config);
  }
  static destroyRawROS(msg) {
    CalibConfigWrite_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'CalibConfigWrite_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get calib_config() {
    return this._wrapperFields.calib_config.data;
  }
  set calib_config(value) {
    this._wrapperFields.calib_config.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CalibConfigWrite_RequestRefStruct(refObject.toObject());
    this._wrapperFields.calib_config.copyRefObject(this._refObject.calib_config);
  }
  copy(other) {
    this._refObject = new CalibConfigWrite_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.calib_config.copy(other._wrapperFields.calib_config);
  }
  static get classType() {
    return CalibConfigWrite_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"calib_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"CalibConfigWrite_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CalibConfigWrite_Request"};
  }
  hasMember(name) {
    let memberNames = ["calib_config"];
    return memberNames.indexOf(name) !== -1;
  }
}
class CalibConfigWrite_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CalibConfigWrite_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CalibConfigWrite_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CalibConfigWrite_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CalibConfigWrite_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new CalibConfigWrite_RequestRefArray(size);
    this._refObject = new CalibConfigWrite_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CalibConfigWrite_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CalibConfigWrite_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CalibConfigWrite_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CalibConfigWrite_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CalibConfigWrite_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CalibConfigWrite_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = CalibConfigWrite_RequestWrapper;
