// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__SafetyPresetWrite_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SafetyPresetWrite_RequestRefStruct = StructType({
  safety_preset: primitiveTypes.string,
  index: primitiveTypes.uint8,
});
const SafetyPresetWrite_RequestRefArray = ArrayType(SafetyPresetWrite_RequestRefStruct);
const SafetyPresetWrite_RequestRefStructArray = StructType({
  data: SafetyPresetWrite_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SafetyPresetWrite_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SafetyPresetWrite_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.safety_preset =  new StringWrapper(other._wrapperFields.safety_preset);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SafetyPresetWrite_RequestRefStruct();
    this._wrapperFields.safety_preset = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SafetyPresetWrite_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SafetyPresetWrite_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SafetyPresetWrite_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SafetyPresetWrite_RequestRefStructArray;
  }
  static get refObjectType() {
    return SafetyPresetWrite_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.safety_preset.freeze(own);
    }
    this._refObject.safety_preset = this._wrapperFields.safety_preset.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.safety_preset.data = refObject.safety_preset.data;
    this._refObject.index = refObject.index;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.safety_preset);
  }
  static destroyRawROS(msg) {
    SafetyPresetWrite_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'SafetyPresetWrite_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get safety_preset() {
    return this._wrapperFields.safety_preset.data;
  }
  set safety_preset(value) {
    this._wrapperFields.safety_preset.data = value;
  }
  get index() {
    return this._refObject.index;
  }
  set index(value) {
    this._refObject.index = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SafetyPresetWrite_RequestRefStruct(refObject.toObject());
    this._wrapperFields.safety_preset.copyRefObject(this._refObject.safety_preset);
  }
  copy(other) {
    this._refObject = new SafetyPresetWrite_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.safety_preset.copy(other._wrapperFields.safety_preset);
  }
  static get classType() {
    return SafetyPresetWrite_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"safety_preset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"SafetyPresetWrite_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SafetyPresetWrite_Request"};
  }
  hasMember(name) {
    let memberNames = ["safety_preset","index"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SafetyPresetWrite_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SafetyPresetWrite_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SafetyPresetWrite_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyPresetWrite_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyPresetWrite_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SafetyPresetWrite_RequestRefArray(size);
    this._refObject = new SafetyPresetWrite_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SafetyPresetWrite_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SafetyPresetWrite_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SafetyPresetWrite_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SafetyPresetWrite_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SafetyPresetWrite_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SafetyPresetWrite_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SafetyPresetWrite_RequestWrapper;
