// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__HardwareMonitorCommandSend_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const HardwareMonitorCommandSend_RequestRefStruct = StructType({
  opcode: primitiveTypes.uint32,
  param1: primitiveTypes.uint32,
  param2: primitiveTypes.uint32,
  param3: primitiveTypes.uint32,
  param4: primitiveTypes.uint32,
  data: UInt8Wrapper.refObjectArrayType,
});
const HardwareMonitorCommandSend_RequestRefArray = ArrayType(HardwareMonitorCommandSend_RequestRefStruct);
const HardwareMonitorCommandSend_RequestRefStructArray = StructType({
  data: HardwareMonitorCommandSend_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class HardwareMonitorCommandSend_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new HardwareMonitorCommandSend_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new HardwareMonitorCommandSend_RequestRefStruct();
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new HardwareMonitorCommandSend_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HardwareMonitorCommandSend_RequestArrayWrapper();
  }
  static get ArrayType() {
    return HardwareMonitorCommandSend_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return HardwareMonitorCommandSend_RequestRefStructArray;
  }
  static get refObjectType() {
    return HardwareMonitorCommandSend_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.data.fill(Uint8Array.from([]));
    }
    this._wrapperFields.data.freeze(own);
    this._refObject.data = this._wrapperFields.data.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.opcode = refObject.opcode;
    this._refObject.param1 = refObject.param1;
    this._refObject.param2 = refObject.param2;
    this._refObject.param3 = refObject.param3;
    this._refObject.param4 = refObject.param4;
    this._wrapperFields.data.copyRefObject(refObject.data);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.data.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.data);
      if (!UInt8Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.data, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    HardwareMonitorCommandSend_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'HardwareMonitorCommandSend_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get opcode() {
    return this._refObject.opcode;
  }
  set opcode(value) {
    this._refObject.opcode = value;
  }
  get param1() {
    return this._refObject.param1;
  }
  set param1(value) {
    this._refObject.param1 = value;
  }
  get param2() {
    return this._refObject.param2;
  }
  set param2(value) {
    this._refObject.param2 = value;
  }
  get param3() {
    return this._refObject.param3;
  }
  set param3(value) {
    this._refObject.param3 = value;
  }
  get param4() {
    return this._refObject.param4;
  }
  set param4(value) {
    this._refObject.param4 = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new HardwareMonitorCommandSend_RequestRefStruct(refObject.toObject());
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new HardwareMonitorCommandSend_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return HardwareMonitorCommandSend_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"opcode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"HardwareMonitorCommandSend_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HardwareMonitorCommandSend_Request"};
  }
  hasMember(name) {
    let memberNames = ["opcode","param1","param2","param3","param4","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
class HardwareMonitorCommandSend_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HardwareMonitorCommandSend_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HardwareMonitorCommandSend_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HardwareMonitorCommandSend_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HardwareMonitorCommandSend_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new HardwareMonitorCommandSend_RequestRefArray(size);
    this._refObject = new HardwareMonitorCommandSend_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HardwareMonitorCommandSend_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HardwareMonitorCommandSend_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HardwareMonitorCommandSend_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HardwareMonitorCommandSend_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HardwareMonitorCommandSend_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HardwareMonitorCommandSend_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = HardwareMonitorCommandSend_RequestWrapper;
