// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__SafetyInterfaceConfigRead_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SafetyInterfaceConfigRead_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  error_message: primitiveTypes.string,
  safety_interface_config: primitiveTypes.string,
});
const SafetyInterfaceConfigRead_ResponseRefArray = ArrayType(SafetyInterfaceConfigRead_ResponseRefStruct);
const SafetyInterfaceConfigRead_ResponseRefStructArray = StructType({
  data: SafetyInterfaceConfigRead_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SafetyInterfaceConfigRead_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SafetyInterfaceConfigRead_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.error_message =  new StringWrapper(other._wrapperFields.error_message);
      this._wrapperFields.safety_interface_config =  new StringWrapper(other._wrapperFields.safety_interface_config);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SafetyInterfaceConfigRead_ResponseRefStruct();
    this._wrapperFields.error_message = new StringWrapper();
    this._wrapperFields.safety_interface_config = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SafetyInterfaceConfigRead_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SafetyInterfaceConfigRead_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return SafetyInterfaceConfigRead_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SafetyInterfaceConfigRead_ResponseRefStructArray;
  }
  static get refObjectType() {
    return SafetyInterfaceConfigRead_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.error_message.freeze(own);
    }
    this._refObject.error_message = this._wrapperFields.error_message.refObject;
    if (own) {
      this._wrapperFields.safety_interface_config.freeze(own);
    }
    this._refObject.safety_interface_config = this._wrapperFields.safety_interface_config.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.success = refObject.success;
    this._wrapperFields.error_message.data = refObject.error_message.data;
    this._wrapperFields.safety_interface_config.data = refObject.safety_interface_config.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_message);
    StringWrapper.freeStruct(refObject.safety_interface_config);
  }
  static destroyRawROS(msg) {
    SafetyInterfaceConfigRead_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'SafetyInterfaceConfigRead_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  get error_message() {
    return this._wrapperFields.error_message.data;
  }
  set error_message(value) {
    this._wrapperFields.error_message.data = value;
  }
  get safety_interface_config() {
    return this._wrapperFields.safety_interface_config.data;
  }
  set safety_interface_config(value) {
    this._wrapperFields.safety_interface_config.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SafetyInterfaceConfigRead_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.error_message.copyRefObject(this._refObject.error_message);
    this._wrapperFields.safety_interface_config.copyRefObject(this._refObject.safety_interface_config);
  }
  copy(other) {
    this._refObject = new SafetyInterfaceConfigRead_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.error_message.copy(other._wrapperFields.error_message);
    this._wrapperFields.safety_interface_config.copy(other._wrapperFields.safety_interface_config);
  }
  static get classType() {
    return SafetyInterfaceConfigRead_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"safety_interface_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"SafetyInterfaceConfigRead_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SafetyInterfaceConfigRead_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","error_message","safety_interface_config"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SafetyInterfaceConfigRead_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SafetyInterfaceConfigRead_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SafetyInterfaceConfigRead_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyInterfaceConfigRead_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyInterfaceConfigRead_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SafetyInterfaceConfigRead_ResponseRefArray(size);
    this._refObject = new SafetyInterfaceConfigRead_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SafetyInterfaceConfigRead_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SafetyInterfaceConfigRead_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SafetyInterfaceConfigRead_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SafetyInterfaceConfigRead_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SafetyInterfaceConfigRead_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SafetyInterfaceConfigRead_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SafetyInterfaceConfigRead_ResponseWrapper;
