// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__srv__SafetyInterfaceConfigWrite_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SafetyInterfaceConfigWrite_RequestRefStruct = StructType({
  safety_interface_config: primitiveTypes.string,
});
const SafetyInterfaceConfigWrite_RequestRefArray = ArrayType(SafetyInterfaceConfigWrite_RequestRefStruct);
const SafetyInterfaceConfigWrite_RequestRefStructArray = StructType({
  data: SafetyInterfaceConfigWrite_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SafetyInterfaceConfigWrite_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SafetyInterfaceConfigWrite_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.safety_interface_config =  new StringWrapper(other._wrapperFields.safety_interface_config);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SafetyInterfaceConfigWrite_RequestRefStruct();
    this._wrapperFields.safety_interface_config = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SafetyInterfaceConfigWrite_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SafetyInterfaceConfigWrite_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SafetyInterfaceConfigWrite_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SafetyInterfaceConfigWrite_RequestRefStructArray;
  }
  static get refObjectType() {
    return SafetyInterfaceConfigWrite_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.safety_interface_config.freeze(own);
    }
    this._refObject.safety_interface_config = this._wrapperFields.safety_interface_config.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.safety_interface_config.data = refObject.safety_interface_config.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.safety_interface_config);
  }
  static destroyRawROS(msg) {
    SafetyInterfaceConfigWrite_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'srv', interfaceName: 'SafetyInterfaceConfigWrite_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get safety_interface_config() {
    return this._wrapperFields.safety_interface_config.data;
  }
  set safety_interface_config(value) {
    this._wrapperFields.safety_interface_config.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SafetyInterfaceConfigWrite_RequestRefStruct(refObject.toObject());
    this._wrapperFields.safety_interface_config.copyRefObject(this._refObject.safety_interface_config);
  }
  copy(other) {
    this._refObject = new SafetyInterfaceConfigWrite_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.safety_interface_config.copy(other._wrapperFields.safety_interface_config);
  }
  static get classType() {
    return SafetyInterfaceConfigWrite_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"safety_interface_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"SafetyInterfaceConfigWrite_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SafetyInterfaceConfigWrite_Request"};
  }
  hasMember(name) {
    let memberNames = ["safety_interface_config"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SafetyInterfaceConfigWrite_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SafetyInterfaceConfigWrite_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SafetyInterfaceConfigWrite_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyInterfaceConfigWrite_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyInterfaceConfigWrite_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SafetyInterfaceConfigWrite_RequestRefArray(size);
    this._refObject = new SafetyInterfaceConfigWrite_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SafetyInterfaceConfigWrite_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SafetyInterfaceConfigWrite_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SafetyInterfaceConfigWrite_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SafetyInterfaceConfigWrite_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SafetyInterfaceConfigWrite_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SafetyInterfaceConfigWrite_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SafetyInterfaceConfigWrite_RequestWrapper;
