// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__action__TriggeredCalibration.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class realsense2_camera_msgs__action__TriggeredCalibration {
  static get Goal() {
    return require('./realsense2_camera_msgs__action__TriggeredCalibration_Goal.js');
  }
  static get Result() {
    return require('./realsense2_camera_msgs__action__TriggeredCalibration_Result.js');
  }
  static get Feedback() {
    return require('./realsense2_camera_msgs__action__TriggeredCalibration_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./realsense2_camera_msgs__action__TriggeredCalibration_SendGoal.js');
      },
      get GetResultService() {
        return require('./realsense2_camera_msgs__action__TriggeredCalibration_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./realsense2_camera_msgs__action__TriggeredCalibration_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'action', interfaceName: 'TriggeredCalibration'};
  }
}
module.exports = realsense2_camera_msgs__action__TriggeredCalibration;
