// rclnodejs/generated/i2c_interfaces/i2c_interfaces__msg__MotorCmd.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MotorCmdRefStruct = StructType({
  id: primitiveTypes.string,
  direction: primitiveTypes.int32,
  intensity: primitiveTypes.int32,
});
const MotorCmdRefArray = ArrayType(MotorCmdRefStruct);
const MotorCmdRefStructArray = StructType({
  data: MotorCmdRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MotorCmdWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotorCmdRefStruct(other._refObject.toObject());
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MotorCmdRefStruct();
    this._wrapperFields.id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MotorCmdWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotorCmdArrayWrapper();
  }
  static get ArrayType() {
    return MotorCmdArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotorCmdRefStructArray;
  }
  static get refObjectType() {
    return MotorCmdRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.id.freeze(own);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.id.data = refObject.id.data;
    this._refObject.direction = refObject.direction;
    this._refObject.intensity = refObject.intensity;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.id);
  }
  static destroyRawROS(msg) {
    MotorCmdWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'i2c_interfaces', subFolder: 'msg', interfaceName: 'MotorCmd'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._wrapperFields.id.data = value;
  }
  get direction() {
    return this._refObject.direction;
  }
  set direction(value) {
    this._refObject.direction = value;
  }
  get intensity() {
    return this._refObject.intensity;
  }
  set intensity(value) {
    this._refObject.intensity = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MotorCmdRefStruct(refObject.toObject());
    this._wrapperFields.id.copyRefObject(this._refObject.id);
  }
  copy(other) {
    this._refObject = new MotorCmdRefStruct(other._refObject.toObject());
    this._wrapperFields.id.copy(other._wrapperFields.id);
  }
  static get classType() {
    return MotorCmdWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"direction","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"intensity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"i2c_interfaces","type":"MotorCmd","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotorCmd"};
  }
  hasMember(name) {
    let memberNames = ["id","direction","intensity"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MotorCmdArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotorCmdWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotorCmdWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotorCmdArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotorCmdArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MotorCmdRefArray(size);
    this._refObject = new MotorCmdRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotorCmdWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotorCmdArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotorCmdArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotorCmdWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotorCmdWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotorCmdArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MotorCmdWrapper;
