// rclnodejs/generated/i2c_interfaces/i2c_interfaces__msg__I2cRaw.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const I2cRawRefStruct = StructType({
  address: primitiveTypes.uint8,
  reg: primitiveTypes.uint8,
  data: UInt8Wrapper.refObjectArrayType,
});
const I2cRawRefArray = ArrayType(I2cRawRefStruct);
const I2cRawRefStructArray = StructType({
  data: I2cRawRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class I2cRawWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new I2cRawRefStruct(other._refObject.toObject());
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new I2cRawRefStruct();
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new I2cRawWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new I2cRawArrayWrapper();
  }
  static get ArrayType() {
    return I2cRawArrayWrapper;
  }
  static get refObjectArrayType() {
    return I2cRawRefStructArray;
  }
  static get refObjectType() {
    return I2cRawRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.data.fill(Uint8Array.from([]));
    }
    this._wrapperFields.data.freeze(own);
    this._refObject.data = this._wrapperFields.data.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.address = refObject.address;
    this._refObject.reg = refObject.reg;
    this._wrapperFields.data.copyRefObject(refObject.data);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.data.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.data);
      if (!UInt8Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.data, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    I2cRawWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'i2c_interfaces', subFolder: 'msg', interfaceName: 'I2cRaw'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get address() {
    return this._refObject.address;
  }
  set address(value) {
    this._refObject.address = value;
  }
  get reg() {
    return this._refObject.reg;
  }
  set reg(value) {
    this._refObject.reg = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new I2cRawRefStruct(refObject.toObject());
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new I2cRawRefStruct(other._refObject.toObject());
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return I2cRawWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"address","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"i2c_interfaces","type":"I2cRaw","stringUpperBound":null,"isPrimitiveType":false},"msgName":"I2cRaw"};
  }
  hasMember(name) {
    let memberNames = ["address","reg","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
class I2cRawArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof I2cRawWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new I2cRawWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to I2cRawArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to I2cRawArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new I2cRawRefArray(size);
    this._refObject = new I2cRawRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new I2cRawWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof I2cRawArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "I2cRawArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      I2cRawWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return I2cRawWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return I2cRawArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = I2cRawWrapper;
