// rclnodejs/generated/i2c_interfaces/i2c_interfaces__msg__DacCmd.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DacCmdRefStruct = StructType({
  channel: primitiveTypes.int32,
  value: primitiveTypes.uint16,
});
const DacCmdRefArray = ArrayType(DacCmdRefStruct);
const DacCmdRefStructArray = StructType({
  data: DacCmdRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class DacCmdWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new DacCmdRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new DacCmdRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new DacCmdWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DacCmdArrayWrapper();
  }
  static get ArrayType() {
    return DacCmdArrayWrapper;
  }
  static get refObjectArrayType() {
    return DacCmdRefStructArray;
  }
  static get refObjectType() {
    return DacCmdRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.channel = refObject.channel;
    this._refObject.value = refObject.value;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    DacCmdWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'i2c_interfaces', subFolder: 'msg', interfaceName: 'DacCmd'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get channel() {
    return this._refObject.channel;
  }
  set channel(value) {
    this._refObject.channel = value;
  }
  get value() {
    return this._refObject.value;
  }
  set value(value) {
    this._refObject.value = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DacCmdRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new DacCmdRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return DacCmdWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"channel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"i2c_interfaces","type":"DacCmd","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DacCmd"};
  }
  hasMember(name) {
    let memberNames = ["channel","value"];
    return memberNames.indexOf(name) !== -1;
  }
}
class DacCmdArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DacCmdWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DacCmdWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DacCmdArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DacCmdArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new DacCmdRefArray(size);
    this._refObject = new DacCmdRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DacCmdWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DacCmdArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DacCmdArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DacCmdWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DacCmdWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DacCmdArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = DacCmdWrapper;
