// rclnodejs/generated/tf2_msgs/tf2_msgs__srv__FrameGraph_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FrameGraph_ResponseRefStruct = StructType({
  frame_yaml: primitiveTypes.string,
});
const FrameGraph_ResponseRefArray = ArrayType(FrameGraph_ResponseRefStruct);
const FrameGraph_ResponseRefStructArray = StructType({
  data: FrameGraph_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class FrameGraph_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new FrameGraph_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.frame_yaml =  new StringWrapper(other._wrapperFields.frame_yaml);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new FrameGraph_ResponseRefStruct();
    this._wrapperFields.frame_yaml = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FrameGraph_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FrameGraph_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return FrameGraph_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return FrameGraph_ResponseRefStructArray;
  }
  static get refObjectType() {
    return FrameGraph_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.frame_yaml.freeze(own);
    }
    this._refObject.frame_yaml = this._wrapperFields.frame_yaml.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.frame_yaml.data = refObject.frame_yaml.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.frame_yaml);
  }
  static destroyRawROS(msg) {
    FrameGraph_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'srv', interfaceName: 'FrameGraph_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frame_yaml() {
    return this._wrapperFields.frame_yaml.data;
  }
  set frame_yaml(value) {
    this._wrapperFields.frame_yaml.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FrameGraph_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.frame_yaml.copyRefObject(this._refObject.frame_yaml);
  }
  copy(other) {
    this._refObject = new FrameGraph_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.frame_yaml.copy(other._wrapperFields.frame_yaml);
  }
  static get classType() {
    return FrameGraph_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"frame_yaml","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"tf2_msgs","type":"FrameGraph_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FrameGraph_Response"};
  }
  hasMember(name) {
    let memberNames = ["frame_yaml"];
    return memberNames.indexOf(name) !== -1;
  }
}
class FrameGraph_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FrameGraph_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FrameGraph_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrameGraph_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrameGraph_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new FrameGraph_ResponseRefArray(size);
    this._refObject = new FrameGraph_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FrameGraph_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FrameGraph_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FrameGraph_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FrameGraph_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FrameGraph_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FrameGraph_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = FrameGraph_ResponseWrapper;
