// rclnodejs/generated/pendulum_msgs/pendulum_msgs__msg__RttestResults.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const JointCommandWrapper = require('../../generated/pendulum_msgs/pendulum_msgs__msg__JointCommand.js');
const JointStateWrapper = require('../../generated/pendulum_msgs/pendulum_msgs__msg__JointState.js');
const RttestResultsRefStruct = StructType({
  stamp: TimeWrapper.refObjectType,
  command: JointCommandWrapper.refObjectType,
  state: JointStateWrapper.refObjectType,
  cur_latency: primitiveTypes.uint64,
  mean_latency: primitiveTypes.float64,
  min_latency: primitiveTypes.uint64,
  max_latency: primitiveTypes.uint64,
  minor_pagefaults: primitiveTypes.uint64,
  major_pagefaults: primitiveTypes.uint64,
});
const RttestResultsRefArray = ArrayType(RttestResultsRefStruct);
const RttestResultsRefStructArray = StructType({
  data: RttestResultsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RttestResultsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RttestResultsRefStruct(other._refObject.toObject());
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
      this._wrapperFields.command =  new JointCommandWrapper(other._wrapperFields.command);
      this._wrapperFields.state =  new JointStateWrapper(other._wrapperFields.state);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RttestResultsRefStruct();
    this._wrapperFields.stamp = new TimeWrapper();
    this._wrapperFields.command = new JointCommandWrapper();
    this._wrapperFields.state = new JointStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RttestResultsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RttestResultsArrayWrapper();
  }
  static get ArrayType() {
    return RttestResultsArrayWrapper;
  }
  static get refObjectArrayType() {
    return RttestResultsRefStructArray;
  }
  static get refObjectType() {
    return RttestResultsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.stamp.freeze(own);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
    this._wrapperFields.command.freeze(own);
    this._refObject.command = this._wrapperFields.command.refObject;
    this._wrapperFields.state.freeze(own);
    this._refObject.state = this._wrapperFields.state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
    this._wrapperFields.command.copyRefObject(refObject.command);
    this._wrapperFields.state.copyRefObject(refObject.state);
    this._refObject.cur_latency = refObject.cur_latency;
    this._refObject.mean_latency = refObject.mean_latency;
    this._refObject.min_latency = refObject.min_latency;
    this._refObject.max_latency = refObject.max_latency;
    this._refObject.minor_pagefaults = refObject.minor_pagefaults;
    this._refObject.major_pagefaults = refObject.major_pagefaults;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.stamp);
    JointCommandWrapper.freeStruct(refObject.command);
    JointStateWrapper.freeStruct(refObject.state);
  }
  static destroyRawROS(msg) {
    RttestResultsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pendulum_msgs', subFolder: 'msg', interfaceName: 'RttestResults'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  get command() {
    return this._wrapperFields.command;
  }
  set command(value) {
    if (value instanceof JointCommandWrapper) {
      this._wrapperFields.command.copy(value);
    } else {
      this._wrapperFields.command.copy(new JointCommandWrapper(value));
    }
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof JointStateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new JointStateWrapper(value));
    }
  }
  get cur_latency() {
    return BigInt(this._refObject.cur_latency);
  }
  set cur_latency(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('cur_latency must be type of bigint');
    }
    this._refObject.cur_latency = value.toString();
  }
  get mean_latency() {
    return this._refObject.mean_latency;
  }
  set mean_latency(value) {
    this._refObject.mean_latency = value;
  }
  get min_latency() {
    return BigInt(this._refObject.min_latency);
  }
  set min_latency(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('min_latency must be type of bigint');
    }
    this._refObject.min_latency = value.toString();
  }
  get max_latency() {
    return BigInt(this._refObject.max_latency);
  }
  set max_latency(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('max_latency must be type of bigint');
    }
    this._refObject.max_latency = value.toString();
  }
  get minor_pagefaults() {
    return BigInt(this._refObject.minor_pagefaults);
  }
  set minor_pagefaults(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('minor_pagefaults must be type of bigint');
    }
    this._refObject.minor_pagefaults = value.toString();
  }
  get major_pagefaults() {
    return BigInt(this._refObject.major_pagefaults);
  }
  set major_pagefaults(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('major_pagefaults must be type of bigint');
    }
    this._refObject.major_pagefaults = value.toString();
  }
  copyRefObject(refObject) {
    this._refObject = new RttestResultsRefStruct(refObject.toObject());
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
    this._wrapperFields.command.copyRefObject(this._refObject.command);
    this._wrapperFields.state.copyRefObject(this._refObject.state);
  }
  copy(other) {
    this._refObject = new RttestResultsRefStruct(other._refObject.toObject());
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
    this._wrapperFields.command.copy(other._wrapperFields.command);
    this._wrapperFields.state.copy(other._wrapperFields.state);
  }
  static get classType() {
    return RttestResultsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"pendulum_msgs","type":"JointCommand","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"pendulum_msgs","type":"JointState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cur_latency","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mean_latency","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_latency","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_latency","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"minor_pagefaults","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"major_pagefaults","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pendulum_msgs","type":"RttestResults","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RttestResults"};
  }
  hasMember(name) {
    let memberNames = ["stamp","command","state","cur_latency","mean_latency","min_latency","max_latency","minor_pagefaults","major_pagefaults"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RttestResultsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RttestResultsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RttestResultsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RttestResultsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RttestResultsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RttestResultsRefArray(size);
    this._refObject = new RttestResultsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RttestResultsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RttestResultsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RttestResultsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RttestResultsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RttestResultsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RttestResultsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RttestResultsWrapper;
