// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__RGBDImages.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RGBDImageWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__RGBDImage.js');
const RGBDImagesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  rgbd_images: RGBDImageWrapper.refObjectArrayType,
});
const RGBDImagesRefArray = ArrayType(RGBDImagesRefStruct);
const RGBDImagesRefStructArray = StructType({
  data: RGBDImagesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RGBDImagesWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RGBDImagesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.rgbd_images = RGBDImageWrapper.createArray();
      this._wrapperFields.rgbd_images.copy(other._wrapperFields.rgbd_images);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RGBDImagesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.rgbd_images = RGBDImageWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new RGBDImagesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RGBDImagesArrayWrapper();
  }
  static get ArrayType() {
    return RGBDImagesArrayWrapper;
  }
  static get refObjectArrayType() {
    return RGBDImagesRefStructArray;
  }
  static get refObjectType() {
    return RGBDImagesRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.rgbd_images.fill([]);
    }
    this._wrapperFields.rgbd_images.freeze(own);
    this._refObject.rgbd_images = this._wrapperFields.rgbd_images.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.rgbd_images.copyRefObject(refObject.rgbd_images);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.rgbd_images.size != 0) {
      RGBDImageWrapper.ArrayType.freeArray(refObject.rgbd_images);
      if (!RGBDImageWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.rgbd_images, RGBDImageWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    RGBDImagesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'RGBDImages'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get rgbd_images() {
    return this._wrapperFields.rgbd_images;
  }
  set rgbd_images(value) {
    this._wrapperFields.rgbd_images.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new RGBDImagesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.rgbd_images.copyRefObject(this._refObject.rgbd_images);
  }
  copy(other) {
    this._refObject = new RGBDImagesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.rgbd_images.copy(other._wrapperFields.rgbd_images);
  }
  static get classType() {
    return RGBDImagesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"rgbd_images","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rtabmap_msgs","type":"RGBDImage","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"RGBDImages","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RGBDImages"};
  }
  hasMember(name) {
    let memberNames = ["header","rgbd_images"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RGBDImagesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RGBDImagesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RGBDImagesWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RGBDImagesArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RGBDImagesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RGBDImagesRefArray(size);
    this._refObject = new RGBDImagesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RGBDImagesWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RGBDImagesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RGBDImagesArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RGBDImagesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RGBDImagesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RGBDImagesArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RGBDImagesWrapper;
