// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__GlobalBundleAdjustment_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GlobalBundleAdjustment_RequestRefStruct = StructType({
  type: primitiveTypes.int32,
  iterations: primitiveTypes.int32,
  pixel_variance: primitiveTypes.float32,
  voc_matches: primitiveTypes.bool,
});
const GlobalBundleAdjustment_RequestRefArray = ArrayType(GlobalBundleAdjustment_RequestRefStruct);
const GlobalBundleAdjustment_RequestRefStructArray = StructType({
  data: GlobalBundleAdjustment_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GlobalBundleAdjustment_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GlobalBundleAdjustment_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GlobalBundleAdjustment_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new GlobalBundleAdjustment_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GlobalBundleAdjustment_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GlobalBundleAdjustment_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GlobalBundleAdjustment_RequestRefStructArray;
  }
  static get refObjectType() {
    return GlobalBundleAdjustment_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.type = refObject.type;
    this._refObject.iterations = refObject.iterations;
    this._refObject.pixel_variance = refObject.pixel_variance;
    this._refObject.voc_matches = refObject.voc_matches;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    GlobalBundleAdjustment_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'GlobalBundleAdjustment_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type() {
    return this._refObject.type;
  }
  set type(value) {
    this._refObject.type = value;
  }
  get iterations() {
    return this._refObject.iterations;
  }
  set iterations(value) {
    this._refObject.iterations = value;
  }
  get pixel_variance() {
    return this._refObject.pixel_variance;
  }
  set pixel_variance(value) {
    this._refObject.pixel_variance = value;
  }
  get voc_matches() {
    return this._refObject.voc_matches;
  }
  set voc_matches(value) {
    this._refObject.voc_matches = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GlobalBundleAdjustment_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new GlobalBundleAdjustment_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return GlobalBundleAdjustment_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iterations","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pixel_variance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voc_matches","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GlobalBundleAdjustment_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GlobalBundleAdjustment_Request"};
  }
  hasMember(name) {
    let memberNames = ["type","iterations","pixel_variance","voc_matches"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GlobalBundleAdjustment_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GlobalBundleAdjustment_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GlobalBundleAdjustment_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalBundleAdjustment_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalBundleAdjustment_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GlobalBundleAdjustment_RequestRefArray(size);
    this._refObject = new GlobalBundleAdjustment_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GlobalBundleAdjustment_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GlobalBundleAdjustment_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GlobalBundleAdjustment_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GlobalBundleAdjustment_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GlobalBundleAdjustment_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GlobalBundleAdjustment_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GlobalBundleAdjustment_RequestWrapper;
