// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__DetectMoreLoopClosures_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DetectMoreLoopClosures_RequestRefStruct = StructType({
  cluster_radius_max: primitiveTypes.float32,
  cluster_radius_min: primitiveTypes.float32,
  cluster_angle: primitiveTypes.float32,
  iterations: primitiveTypes.int32,
  intra_only: primitiveTypes.bool,
  inter_only: primitiveTypes.bool,
});
const DetectMoreLoopClosures_RequestRefArray = ArrayType(DetectMoreLoopClosures_RequestRefStruct);
const DetectMoreLoopClosures_RequestRefStructArray = StructType({
  data: DetectMoreLoopClosures_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class DetectMoreLoopClosures_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new DetectMoreLoopClosures_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new DetectMoreLoopClosures_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new DetectMoreLoopClosures_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DetectMoreLoopClosures_RequestArrayWrapper();
  }
  static get ArrayType() {
    return DetectMoreLoopClosures_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return DetectMoreLoopClosures_RequestRefStructArray;
  }
  static get refObjectType() {
    return DetectMoreLoopClosures_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.cluster_radius_max = refObject.cluster_radius_max;
    this._refObject.cluster_radius_min = refObject.cluster_radius_min;
    this._refObject.cluster_angle = refObject.cluster_angle;
    this._refObject.iterations = refObject.iterations;
    this._refObject.intra_only = refObject.intra_only;
    this._refObject.inter_only = refObject.inter_only;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    DetectMoreLoopClosures_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'DetectMoreLoopClosures_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get cluster_radius_max() {
    return this._refObject.cluster_radius_max;
  }
  set cluster_radius_max(value) {
    this._refObject.cluster_radius_max = value;
  }
  get cluster_radius_min() {
    return this._refObject.cluster_radius_min;
  }
  set cluster_radius_min(value) {
    this._refObject.cluster_radius_min = value;
  }
  get cluster_angle() {
    return this._refObject.cluster_angle;
  }
  set cluster_angle(value) {
    this._refObject.cluster_angle = value;
  }
  get iterations() {
    return this._refObject.iterations;
  }
  set iterations(value) {
    this._refObject.iterations = value;
  }
  get intra_only() {
    return this._refObject.intra_only;
  }
  set intra_only(value) {
    this._refObject.intra_only = value;
  }
  get inter_only() {
    return this._refObject.inter_only;
  }
  set inter_only(value) {
    this._refObject.inter_only = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DetectMoreLoopClosures_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new DetectMoreLoopClosures_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return DetectMoreLoopClosures_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"cluster_radius_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cluster_radius_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cluster_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iterations","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"intra_only","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"inter_only","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"DetectMoreLoopClosures_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DetectMoreLoopClosures_Request"};
  }
  hasMember(name) {
    let memberNames = ["cluster_radius_max","cluster_radius_min","cluster_angle","iterations","intra_only","inter_only"];
    return memberNames.indexOf(name) !== -1;
  }
}
class DetectMoreLoopClosures_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DetectMoreLoopClosures_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DetectMoreLoopClosures_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DetectMoreLoopClosures_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DetectMoreLoopClosures_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new DetectMoreLoopClosures_RequestRefArray(size);
    this._refObject = new DetectMoreLoopClosures_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DetectMoreLoopClosures_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DetectMoreLoopClosures_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DetectMoreLoopClosures_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DetectMoreLoopClosures_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DetectMoreLoopClosures_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DetectMoreLoopClosures_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = DetectMoreLoopClosures_RequestWrapper;
