// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__AddLink_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LinkWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__Link.js');
const AddLink_RequestRefStruct = StructType({
  link: LinkWrapper.refObjectType,
});
const AddLink_RequestRefArray = ArrayType(AddLink_RequestRefStruct);
const AddLink_RequestRefStructArray = StructType({
  data: AddLink_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AddLink_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AddLink_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.link =  new LinkWrapper(other._wrapperFields.link);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AddLink_RequestRefStruct();
    this._wrapperFields.link = new LinkWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new AddLink_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AddLink_RequestArrayWrapper();
  }
  static get ArrayType() {
    return AddLink_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return AddLink_RequestRefStructArray;
  }
  static get refObjectType() {
    return AddLink_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.link.freeze(own);
    this._refObject.link = this._wrapperFields.link.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.link.copyRefObject(refObject.link);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    LinkWrapper.freeStruct(refObject.link);
  }
  static destroyRawROS(msg) {
    AddLink_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'AddLink_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get link() {
    return this._wrapperFields.link;
  }
  set link(value) {
    if (value instanceof LinkWrapper) {
      this._wrapperFields.link.copy(value);
    } else {
      this._wrapperFields.link.copy(new LinkWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new AddLink_RequestRefStruct(refObject.toObject());
    this._wrapperFields.link.copyRefObject(this._refObject.link);
  }
  copy(other) {
    this._refObject = new AddLink_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.link.copy(other._wrapperFields.link);
  }
  static get classType() {
    return AddLink_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"link","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rtabmap_msgs","type":"Link","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"AddLink_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AddLink_Request"};
  }
  hasMember(name) {
    let memberNames = ["link"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AddLink_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AddLink_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AddLink_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AddLink_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AddLink_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AddLink_RequestRefArray(size);
    this._refObject = new AddLink_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AddLink_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AddLink_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AddLink_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AddLink_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AddLink_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AddLink_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = AddLink_RequestWrapper;
