// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__LandmarkDetections.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const LandmarkDetectionWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__LandmarkDetection.js');
const LandmarkDetectionsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  landmarks: LandmarkDetectionWrapper.refObjectArrayType,
});
const LandmarkDetectionsRefArray = ArrayType(LandmarkDetectionsRefStruct);
const LandmarkDetectionsRefStructArray = StructType({
  data: LandmarkDetectionsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LandmarkDetectionsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandmarkDetectionsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.landmarks = LandmarkDetectionWrapper.createArray();
      this._wrapperFields.landmarks.copy(other._wrapperFields.landmarks);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LandmarkDetectionsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.landmarks = LandmarkDetectionWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LandmarkDetectionsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandmarkDetectionsArrayWrapper();
  }
  static get ArrayType() {
    return LandmarkDetectionsArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandmarkDetectionsRefStructArray;
  }
  static get refObjectType() {
    return LandmarkDetectionsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.landmarks.fill([]);
    }
    this._wrapperFields.landmarks.freeze(own);
    this._refObject.landmarks = this._wrapperFields.landmarks.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.landmarks.copyRefObject(refObject.landmarks);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.landmarks.size != 0) {
      LandmarkDetectionWrapper.ArrayType.freeArray(refObject.landmarks);
      if (!LandmarkDetectionWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.landmarks, LandmarkDetectionWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    LandmarkDetectionsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'LandmarkDetections'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get landmarks() {
    return this._wrapperFields.landmarks;
  }
  set landmarks(value) {
    this._wrapperFields.landmarks.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LandmarkDetectionsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.landmarks.copyRefObject(this._refObject.landmarks);
  }
  copy(other) {
    this._refObject = new LandmarkDetectionsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.landmarks.copy(other._wrapperFields.landmarks);
  }
  static get classType() {
    return LandmarkDetectionsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"landmarks","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rtabmap_msgs","type":"LandmarkDetection","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"LandmarkDetections","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandmarkDetections"};
  }
  hasMember(name) {
    let memberNames = ["header","landmarks"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LandmarkDetectionsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandmarkDetectionsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandmarkDetectionsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandmarkDetectionsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandmarkDetectionsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LandmarkDetectionsRefArray(size);
    this._refObject = new LandmarkDetectionsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandmarkDetectionsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandmarkDetectionsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandmarkDetectionsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandmarkDetectionsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandmarkDetectionsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandmarkDetectionsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = LandmarkDetectionsWrapper;
