// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__MapData.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const MapGraphWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__MapGraph.js');
const NodeWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__Node.js');
const MapDataRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  graph: MapGraphWrapper.refObjectType,
  nodes: NodeWrapper.refObjectArrayType,
});
const MapDataRefArray = ArrayType(MapDataRefStruct);
const MapDataRefStructArray = StructType({
  data: MapDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MapDataWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MapDataRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.graph =  new MapGraphWrapper(other._wrapperFields.graph);
      this._wrapperFields.nodes = NodeWrapper.createArray();
      this._wrapperFields.nodes.copy(other._wrapperFields.nodes);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MapDataRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.graph = new MapGraphWrapper();
    this._wrapperFields.nodes = NodeWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MapDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MapDataArrayWrapper();
  }
  static get ArrayType() {
    return MapDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return MapDataRefStructArray;
  }
  static get refObjectType() {
    return MapDataRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.graph.freeze(own);
    this._refObject.graph = this._wrapperFields.graph.refObject;
    if (own) {
      this._wrapperFields.nodes.fill([]);
    }
    this._wrapperFields.nodes.freeze(own);
    this._refObject.nodes = this._wrapperFields.nodes.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.graph.copyRefObject(refObject.graph);
    this._wrapperFields.nodes.copyRefObject(refObject.nodes);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    MapGraphWrapper.freeStruct(refObject.graph);
    if (refObject.nodes.size != 0) {
      NodeWrapper.ArrayType.freeArray(refObject.nodes);
      if (!NodeWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.nodes, NodeWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MapDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'MapData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get graph() {
    return this._wrapperFields.graph;
  }
  set graph(value) {
    if (value instanceof MapGraphWrapper) {
      this._wrapperFields.graph.copy(value);
    } else {
      this._wrapperFields.graph.copy(new MapGraphWrapper(value));
    }
  }
  get nodes() {
    return this._wrapperFields.nodes;
  }
  set nodes(value) {
    this._wrapperFields.nodes.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MapDataRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.graph.copyRefObject(this._refObject.graph);
    this._wrapperFields.nodes.copyRefObject(this._refObject.nodes);
  }
  copy(other) {
    this._refObject = new MapDataRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.graph.copy(other._wrapperFields.graph);
    this._wrapperFields.nodes.copy(other._wrapperFields.nodes);
  }
  static get classType() {
    return MapDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"graph","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rtabmap_msgs","type":"MapGraph","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"nodes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rtabmap_msgs","type":"Node","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"MapData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MapData"};
  }
  hasMember(name) {
    let memberNames = ["header","graph","nodes"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MapDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MapDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MapDataWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapDataArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MapDataRefArray(size);
    this._refObject = new MapDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MapDataWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MapDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MapDataArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MapDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MapDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MapDataArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MapDataWrapper;
