// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__GetMap2_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GetMap2_RequestRefStruct = StructType({
  global_map: primitiveTypes.bool,
  optimized: primitiveTypes.bool,
  with_images: primitiveTypes.bool,
  with_scans: primitiveTypes.bool,
  with_user_data: primitiveTypes.bool,
  with_grids: primitiveTypes.bool,
  with_words: primitiveTypes.bool,
  with_global_descriptors: primitiveTypes.bool,
});
const GetMap2_RequestRefArray = ArrayType(GetMap2_RequestRefStruct);
const GetMap2_RequestRefStructArray = StructType({
  data: GetMap2_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetMap2_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetMap2_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetMap2_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new GetMap2_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetMap2_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetMap2_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetMap2_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetMap2_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.global_map = refObject.global_map;
    this._refObject.optimized = refObject.optimized;
    this._refObject.with_images = refObject.with_images;
    this._refObject.with_scans = refObject.with_scans;
    this._refObject.with_user_data = refObject.with_user_data;
    this._refObject.with_grids = refObject.with_grids;
    this._refObject.with_words = refObject.with_words;
    this._refObject.with_global_descriptors = refObject.with_global_descriptors;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    GetMap2_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'GetMap2_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get global_map() {
    return this._refObject.global_map;
  }
  set global_map(value) {
    this._refObject.global_map = value;
  }
  get optimized() {
    return this._refObject.optimized;
  }
  set optimized(value) {
    this._refObject.optimized = value;
  }
  get with_images() {
    return this._refObject.with_images;
  }
  set with_images(value) {
    this._refObject.with_images = value;
  }
  get with_scans() {
    return this._refObject.with_scans;
  }
  set with_scans(value) {
    this._refObject.with_scans = value;
  }
  get with_user_data() {
    return this._refObject.with_user_data;
  }
  set with_user_data(value) {
    this._refObject.with_user_data = value;
  }
  get with_grids() {
    return this._refObject.with_grids;
  }
  set with_grids(value) {
    this._refObject.with_grids = value;
  }
  get with_words() {
    return this._refObject.with_words;
  }
  set with_words(value) {
    this._refObject.with_words = value;
  }
  get with_global_descriptors() {
    return this._refObject.with_global_descriptors;
  }
  set with_global_descriptors(value) {
    this._refObject.with_global_descriptors = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetMap2_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new GetMap2_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return GetMap2_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"global_map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"optimized","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"with_images","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"with_scans","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"with_user_data","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"with_grids","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"with_words","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"with_global_descriptors","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GetMap2_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetMap2_Request"};
  }
  hasMember(name) {
    let memberNames = ["global_map","optimized","with_images","with_scans","with_user_data","with_grids","with_words","with_global_descriptors"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetMap2_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetMap2_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetMap2_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetMap2_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetMap2_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetMap2_RequestRefArray(size);
    this._refObject = new GetMap2_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetMap2_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetMap2_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetMap2_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetMap2_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetMap2_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetMap2_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetMap2_RequestWrapper;
