// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__CameraModels.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CameraModelWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__CameraModel.js');
const CameraModelsRefStruct = StructType({
  models: CameraModelWrapper.refObjectArrayType,
});
const CameraModelsRefArray = ArrayType(CameraModelsRefStruct);
const CameraModelsRefStructArray = StructType({
  data: CameraModelsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class CameraModelsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new CameraModelsRefStruct(other._refObject.toObject());
      this._wrapperFields.models = CameraModelWrapper.createArray();
      this._wrapperFields.models.copy(other._wrapperFields.models);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new CameraModelsRefStruct();
    this._wrapperFields.models = CameraModelWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new CameraModelsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CameraModelsArrayWrapper();
  }
  static get ArrayType() {
    return CameraModelsArrayWrapper;
  }
  static get refObjectArrayType() {
    return CameraModelsRefStructArray;
  }
  static get refObjectType() {
    return CameraModelsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.models.fill([]);
    }
    this._wrapperFields.models.freeze(own);
    this._refObject.models = this._wrapperFields.models.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.models.copyRefObject(refObject.models);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.models.size != 0) {
      CameraModelWrapper.ArrayType.freeArray(refObject.models);
      if (!CameraModelWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.models, CameraModelWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    CameraModelsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'CameraModels'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get models() {
    return this._wrapperFields.models;
  }
  set models(value) {
    this._wrapperFields.models.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new CameraModelsRefStruct(refObject.toObject());
    this._wrapperFields.models.copyRefObject(this._refObject.models);
  }
  copy(other) {
    this._refObject = new CameraModelsRefStruct(other._refObject.toObject());
    this._wrapperFields.models.copy(other._wrapperFields.models);
  }
  static get classType() {
    return CameraModelsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"models","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rtabmap_msgs","type":"CameraModel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"CameraModels","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CameraModels"};
  }
  hasMember(name) {
    let memberNames = ["models"];
    return memberNames.indexOf(name) !== -1;
  }
}
class CameraModelsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CameraModelsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CameraModelsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraModelsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraModelsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new CameraModelsRefArray(size);
    this._refObject = new CameraModelsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CameraModelsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CameraModelsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CameraModelsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CameraModelsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CameraModelsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CameraModelsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = CameraModelsWrapper;
