// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__LandmarkDetection.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWithCovarianceWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseWithCovariance.js');
const LandmarkDetectionRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  landmark_frame_id: primitiveTypes.string,
  id: primitiveTypes.int32,
  size: primitiveTypes.float32,
  pose: PoseWithCovarianceWrapper.refObjectType,
});
const LandmarkDetectionRefArray = ArrayType(LandmarkDetectionRefStruct);
const LandmarkDetectionRefStructArray = StructType({
  data: LandmarkDetectionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LandmarkDetectionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandmarkDetectionRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.landmark_frame_id =  new StringWrapper(other._wrapperFields.landmark_frame_id);
      this._wrapperFields.pose =  new PoseWithCovarianceWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LandmarkDetectionRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.landmark_frame_id = new StringWrapper();
    this._wrapperFields.pose = new PoseWithCovarianceWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LandmarkDetectionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandmarkDetectionArrayWrapper();
  }
  static get ArrayType() {
    return LandmarkDetectionArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandmarkDetectionRefStructArray;
  }
  static get refObjectType() {
    return LandmarkDetectionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.landmark_frame_id.freeze(own);
    }
    this._refObject.landmark_frame_id = this._wrapperFields.landmark_frame_id.refObject;
    this._wrapperFields.pose.freeze(own);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.landmark_frame_id.data = refObject.landmark_frame_id.data;
    this._refObject.id = refObject.id;
    this._refObject.size = refObject.size;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.landmark_frame_id);
    PoseWithCovarianceWrapper.freeStruct(refObject.pose);
  }
  static destroyRawROS(msg) {
    LandmarkDetectionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'LandmarkDetection'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get landmark_frame_id() {
    return this._wrapperFields.landmark_frame_id.data;
  }
  set landmark_frame_id(value) {
    this._wrapperFields.landmark_frame_id.data = value;
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get size() {
    return this._refObject.size;
  }
  set size(value) {
    this._refObject.size = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWithCovarianceWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWithCovarianceWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new LandmarkDetectionRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.landmark_frame_id.copyRefObject(this._refObject.landmark_frame_id);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new LandmarkDetectionRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.landmark_frame_id.copy(other._wrapperFields.landmark_frame_id);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return LandmarkDetectionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"landmark_frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseWithCovariance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"LandmarkDetection","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandmarkDetection"};
  }
  hasMember(name) {
    let memberNames = ["header","landmark_frame_id","id","size","pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LandmarkDetectionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandmarkDetectionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandmarkDetectionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandmarkDetectionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandmarkDetectionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LandmarkDetectionRefArray(size);
    this._refObject = new LandmarkDetectionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandmarkDetectionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandmarkDetectionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandmarkDetectionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandmarkDetectionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandmarkDetectionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandmarkDetectionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = LandmarkDetectionWrapper;
