// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__CameraModel.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CameraInfoWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__CameraInfo.js');
const TransformWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Transform.js');
const CameraModelRefStruct = StructType({
  camera_info: CameraInfoWrapper.refObjectType,
  local_transform: TransformWrapper.refObjectType,
});
const CameraModelRefArray = ArrayType(CameraModelRefStruct);
const CameraModelRefStructArray = StructType({
  data: CameraModelRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class CameraModelWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new CameraModelRefStruct(other._refObject.toObject());
      this._wrapperFields.camera_info =  new CameraInfoWrapper(other._wrapperFields.camera_info);
      this._wrapperFields.local_transform =  new TransformWrapper(other._wrapperFields.local_transform);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new CameraModelRefStruct();
    this._wrapperFields.camera_info = new CameraInfoWrapper();
    this._wrapperFields.local_transform = new TransformWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CameraModelWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CameraModelArrayWrapper();
  }
  static get ArrayType() {
    return CameraModelArrayWrapper;
  }
  static get refObjectArrayType() {
    return CameraModelRefStructArray;
  }
  static get refObjectType() {
    return CameraModelRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.camera_info.freeze(own);
    this._refObject.camera_info = this._wrapperFields.camera_info.refObject;
    this._wrapperFields.local_transform.freeze(own);
    this._refObject.local_transform = this._wrapperFields.local_transform.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.camera_info.copyRefObject(refObject.camera_info);
    this._wrapperFields.local_transform.copyRefObject(refObject.local_transform);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    CameraInfoWrapper.freeStruct(refObject.camera_info);
    TransformWrapper.freeStruct(refObject.local_transform);
  }
  static destroyRawROS(msg) {
    CameraModelWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'CameraModel'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get camera_info() {
    return this._wrapperFields.camera_info;
  }
  set camera_info(value) {
    if (value instanceof CameraInfoWrapper) {
      this._wrapperFields.camera_info.copy(value);
    } else {
      this._wrapperFields.camera_info.copy(new CameraInfoWrapper(value));
    }
  }
  get local_transform() {
    return this._wrapperFields.local_transform;
  }
  set local_transform(value) {
    if (value instanceof TransformWrapper) {
      this._wrapperFields.local_transform.copy(value);
    } else {
      this._wrapperFields.local_transform.copy(new TransformWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CameraModelRefStruct(refObject.toObject());
    this._wrapperFields.camera_info.copyRefObject(this._refObject.camera_info);
    this._wrapperFields.local_transform.copyRefObject(this._refObject.local_transform);
  }
  copy(other) {
    this._refObject = new CameraModelRefStruct(other._refObject.toObject());
    this._wrapperFields.camera_info.copy(other._wrapperFields.camera_info);
    this._wrapperFields.local_transform.copy(other._wrapperFields.local_transform);
  }
  static get classType() {
    return CameraModelWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"camera_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"CameraInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"local_transform","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Transform","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"CameraModel","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CameraModel"};
  }
  hasMember(name) {
    let memberNames = ["camera_info","local_transform"];
    return memberNames.indexOf(name) !== -1;
  }
}
class CameraModelArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CameraModelWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CameraModelWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraModelArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraModelArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new CameraModelRefArray(size);
    this._refObject = new CameraModelRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CameraModelWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CameraModelArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CameraModelArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CameraModelWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CameraModelWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CameraModelArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = CameraModelWrapper;
