// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__GPS.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GPSRefStruct = StructType({
  stamp: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  latitude: primitiveTypes.float64,
  altitude: primitiveTypes.float64,
  error: primitiveTypes.float64,
  bearing: primitiveTypes.float64,
});
const GPSRefArray = ArrayType(GPSRefStruct);
const GPSRefStructArray = StructType({
  data: GPSRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GPSWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPSRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GPSRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new GPSWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPSArrayWrapper();
  }
  static get ArrayType() {
    return GPSArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPSRefStructArray;
  }
  static get refObjectType() {
    return GPSRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.stamp = refObject.stamp;
    this._refObject.longitude = refObject.longitude;
    this._refObject.latitude = refObject.latitude;
    this._refObject.altitude = refObject.altitude;
    this._refObject.error = refObject.error;
    this._refObject.bearing = refObject.bearing;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    GPSWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'GPS'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stamp() {
    return this._refObject.stamp;
  }
  set stamp(value) {
    this._refObject.stamp = value;
  }
  get longitude() {
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._refObject.longitude = value;
  }
  get latitude() {
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._refObject.latitude = value;
  }
  get altitude() {
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._refObject.altitude = value;
  }
  get error() {
    return this._refObject.error;
  }
  set error(value) {
    this._refObject.error = value;
  }
  get bearing() {
    return this._refObject.bearing;
  }
  set bearing(value) {
    this._refObject.bearing = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPSRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new GPSRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return GPSWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bearing","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GPS","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPS"};
  }
  hasMember(name) {
    let memberNames = ["stamp","longitude","latitude","altitude","error","bearing"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GPSArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPSWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPSWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GPSRefArray(size);
    this._refObject = new GPSRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPSWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPSArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPSArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPSWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPSWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPSArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GPSWrapper;
