// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__GlobalDescriptor.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const GlobalDescriptorRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  type: primitiveTypes.int32,
  info: UInt8Wrapper.refObjectArrayType,
  data: UInt8Wrapper.refObjectArrayType,
});
const GlobalDescriptorRefArray = ArrayType(GlobalDescriptorRefStruct);
const GlobalDescriptorRefStructArray = StructType({
  data: GlobalDescriptorRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GlobalDescriptorWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GlobalDescriptorRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.info = UInt8Wrapper.createArray();
      this._wrapperFields.info.copy(other._wrapperFields.info);
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GlobalDescriptorRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.info = UInt8Wrapper.createArray();
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GlobalDescriptorWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GlobalDescriptorArrayWrapper();
  }
  static get ArrayType() {
    return GlobalDescriptorArrayWrapper;
  }
  static get refObjectArrayType() {
    return GlobalDescriptorRefStructArray;
  }
  static get refObjectType() {
    return GlobalDescriptorRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.info.fill(Uint8Array.from([]));
    }
    this._wrapperFields.info.freeze(own);
    this._refObject.info = this._wrapperFields.info.refObject;
    if (own) {
      this._wrapperFields.data.fill(Uint8Array.from([]));
    }
    this._wrapperFields.data.freeze(own);
    this._refObject.data = this._wrapperFields.data.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.type = refObject.type;
    this._wrapperFields.info.copyRefObject(refObject.info);
    this._wrapperFields.data.copyRefObject(refObject.data);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.info.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.info);
      if (!UInt8Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.info, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.data.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.data);
      if (!UInt8Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.data, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GlobalDescriptorWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'GlobalDescriptor'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get type() {
    return this._refObject.type;
  }
  set type(value) {
    this._refObject.type = value;
  }
  get info() {
    return this._wrapperFields['info'].data;
  }
  set info(value) {
    this._wrapperFields['info'].fill(value);
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GlobalDescriptorRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.info.copyRefObject(this._refObject.info);
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new GlobalDescriptorRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.info.copy(other._wrapperFields.info);
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return GlobalDescriptorWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"info","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GlobalDescriptor","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GlobalDescriptor"};
  }
  hasMember(name) {
    let memberNames = ["header","type","info","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GlobalDescriptorArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GlobalDescriptorWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GlobalDescriptorWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalDescriptorArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalDescriptorArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GlobalDescriptorRefArray(size);
    this._refObject = new GlobalDescriptorRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GlobalDescriptorWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GlobalDescriptorArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GlobalDescriptorArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GlobalDescriptorWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GlobalDescriptorWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GlobalDescriptorArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GlobalDescriptorWrapper;
