// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__ScanDescriptor.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const LaserScanWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__LaserScan.js');
const PointCloud2Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__PointCloud2.js');
const GlobalDescriptorWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__GlobalDescriptor.js');
const ScanDescriptorRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  scan: LaserScanWrapper.refObjectType,
  scan_cloud: PointCloud2Wrapper.refObjectType,
  global_descriptor: GlobalDescriptorWrapper.refObjectType,
});
const ScanDescriptorRefArray = ArrayType(ScanDescriptorRefStruct);
const ScanDescriptorRefStructArray = StructType({
  data: ScanDescriptorRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ScanDescriptorWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ScanDescriptorRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.scan =  new LaserScanWrapper(other._wrapperFields.scan);
      this._wrapperFields.scan_cloud =  new PointCloud2Wrapper(other._wrapperFields.scan_cloud);
      this._wrapperFields.global_descriptor =  new GlobalDescriptorWrapper(other._wrapperFields.global_descriptor);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ScanDescriptorRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.scan = new LaserScanWrapper();
    this._wrapperFields.scan_cloud = new PointCloud2Wrapper();
    this._wrapperFields.global_descriptor = new GlobalDescriptorWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ScanDescriptorWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ScanDescriptorArrayWrapper();
  }
  static get ArrayType() {
    return ScanDescriptorArrayWrapper;
  }
  static get refObjectArrayType() {
    return ScanDescriptorRefStructArray;
  }
  static get refObjectType() {
    return ScanDescriptorRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.scan.freeze(own);
    this._refObject.scan = this._wrapperFields.scan.refObject;
    this._wrapperFields.scan_cloud.freeze(own);
    this._refObject.scan_cloud = this._wrapperFields.scan_cloud.refObject;
    this._wrapperFields.global_descriptor.freeze(own);
    this._refObject.global_descriptor = this._wrapperFields.global_descriptor.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.scan.copyRefObject(refObject.scan);
    this._wrapperFields.scan_cloud.copyRefObject(refObject.scan_cloud);
    this._wrapperFields.global_descriptor.copyRefObject(refObject.global_descriptor);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    LaserScanWrapper.freeStruct(refObject.scan);
    PointCloud2Wrapper.freeStruct(refObject.scan_cloud);
    GlobalDescriptorWrapper.freeStruct(refObject.global_descriptor);
  }
  static destroyRawROS(msg) {
    ScanDescriptorWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'ScanDescriptor'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get scan() {
    return this._wrapperFields.scan;
  }
  set scan(value) {
    if (value instanceof LaserScanWrapper) {
      this._wrapperFields.scan.copy(value);
    } else {
      this._wrapperFields.scan.copy(new LaserScanWrapper(value));
    }
  }
  get scan_cloud() {
    return this._wrapperFields.scan_cloud;
  }
  set scan_cloud(value) {
    if (value instanceof PointCloud2Wrapper) {
      this._wrapperFields.scan_cloud.copy(value);
    } else {
      this._wrapperFields.scan_cloud.copy(new PointCloud2Wrapper(value));
    }
  }
  get global_descriptor() {
    return this._wrapperFields.global_descriptor;
  }
  set global_descriptor(value) {
    if (value instanceof GlobalDescriptorWrapper) {
      this._wrapperFields.global_descriptor.copy(value);
    } else {
      this._wrapperFields.global_descriptor.copy(new GlobalDescriptorWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ScanDescriptorRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.scan.copyRefObject(this._refObject.scan);
    this._wrapperFields.scan_cloud.copyRefObject(this._refObject.scan_cloud);
    this._wrapperFields.global_descriptor.copyRefObject(this._refObject.global_descriptor);
  }
  copy(other) {
    this._refObject = new ScanDescriptorRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.scan.copy(other._wrapperFields.scan);
    this._wrapperFields.scan_cloud.copy(other._wrapperFields.scan_cloud);
    this._wrapperFields.global_descriptor.copy(other._wrapperFields.global_descriptor);
  }
  static get classType() {
    return ScanDescriptorWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"scan","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"LaserScan","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"scan_cloud","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"PointCloud2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"global_descriptor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rtabmap_msgs","type":"GlobalDescriptor","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"ScanDescriptor","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ScanDescriptor"};
  }
  hasMember(name) {
    let memberNames = ["header","scan","scan_cloud","global_descriptor"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ScanDescriptorArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ScanDescriptorWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ScanDescriptorWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ScanDescriptorArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ScanDescriptorArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ScanDescriptorRefArray(size);
    this._refObject = new ScanDescriptorRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ScanDescriptorWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ScanDescriptorArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ScanDescriptorArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ScanDescriptorWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ScanDescriptorWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ScanDescriptorArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ScanDescriptorWrapper;
