// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__MapGraph.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TransformWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Transform.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const LinkWrapper = require('../../generated/rtabmap_msgs/rtabmap_msgs__msg__Link.js');
const MapGraphRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  map_to_odom: TransformWrapper.refObjectType,
  poses_id: Int32Wrapper.refObjectArrayType,
  poses: PoseWrapper.refObjectArrayType,
  links: LinkWrapper.refObjectArrayType,
});
const MapGraphRefArray = ArrayType(MapGraphRefStruct);
const MapGraphRefStructArray = StructType({
  data: MapGraphRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MapGraphWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MapGraphRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.map_to_odom =  new TransformWrapper(other._wrapperFields.map_to_odom);
      this._wrapperFields.poses_id = Int32Wrapper.createArray();
      this._wrapperFields.poses_id.copy(other._wrapperFields.poses_id);
      this._wrapperFields.poses = PoseWrapper.createArray();
      this._wrapperFields.poses.copy(other._wrapperFields.poses);
      this._wrapperFields.links = LinkWrapper.createArray();
      this._wrapperFields.links.copy(other._wrapperFields.links);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MapGraphRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.map_to_odom = new TransformWrapper();
    this._wrapperFields.poses_id = Int32Wrapper.createArray();
    this._wrapperFields.poses = PoseWrapper.createArray();
    this._wrapperFields.links = LinkWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MapGraphWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MapGraphArrayWrapper();
  }
  static get ArrayType() {
    return MapGraphArrayWrapper;
  }
  static get refObjectArrayType() {
    return MapGraphRefStructArray;
  }
  static get refObjectType() {
    return MapGraphRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.map_to_odom.freeze(own);
    this._refObject.map_to_odom = this._wrapperFields.map_to_odom.refObject;
    if (own) {
      this._wrapperFields.poses_id.fill(Int32Array.from([]));
    }
    this._wrapperFields.poses_id.freeze(own);
    this._refObject.poses_id = this._wrapperFields.poses_id.refObject;
    if (own) {
      this._wrapperFields.poses.fill([]);
    }
    this._wrapperFields.poses.freeze(own);
    this._refObject.poses = this._wrapperFields.poses.refObject;
    if (own) {
      this._wrapperFields.links.fill([]);
    }
    this._wrapperFields.links.freeze(own);
    this._refObject.links = this._wrapperFields.links.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.map_to_odom.copyRefObject(refObject.map_to_odom);
    this._wrapperFields.poses_id.copyRefObject(refObject.poses_id);
    this._wrapperFields.poses.copyRefObject(refObject.poses);
    this._wrapperFields.links.copyRefObject(refObject.links);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    TransformWrapper.freeStruct(refObject.map_to_odom);
    if (refObject.poses_id.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.poses_id);
      if (!Int32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.poses_id, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.poses.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.poses);
      if (!PoseWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.poses, PoseWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.links.size != 0) {
      LinkWrapper.ArrayType.freeArray(refObject.links);
      if (!LinkWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.links, LinkWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MapGraphWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'MapGraph'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get map_to_odom() {
    return this._wrapperFields.map_to_odom;
  }
  set map_to_odom(value) {
    if (value instanceof TransformWrapper) {
      this._wrapperFields.map_to_odom.copy(value);
    } else {
      this._wrapperFields.map_to_odom.copy(new TransformWrapper(value));
    }
  }
  get poses_id() {
    return this._wrapperFields['poses_id'].data;
  }
  set poses_id(value) {
    this._wrapperFields['poses_id'].fill(value);
  }
  get poses() {
    return this._wrapperFields.poses;
  }
  set poses(value) {
    this._wrapperFields.poses.fill(value);
  }
  get links() {
    return this._wrapperFields.links;
  }
  set links(value) {
    this._wrapperFields.links.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MapGraphRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.map_to_odom.copyRefObject(this._refObject.map_to_odom);
    this._wrapperFields.poses_id.copyRefObject(this._refObject.poses_id);
    this._wrapperFields.poses.copyRefObject(this._refObject.poses);
    this._wrapperFields.links.copyRefObject(this._refObject.links);
  }
  copy(other) {
    this._refObject = new MapGraphRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.map_to_odom.copy(other._wrapperFields.map_to_odom);
    this._wrapperFields.poses_id.copy(other._wrapperFields.poses_id);
    this._wrapperFields.poses.copy(other._wrapperFields.poses);
    this._wrapperFields.links.copy(other._wrapperFields.links);
  }
  static get classType() {
    return MapGraphWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"map_to_odom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Transform","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"poses_id","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"links","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rtabmap_msgs","type":"Link","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"MapGraph","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MapGraph"};
  }
  hasMember(name) {
    let memberNames = ["header","map_to_odom","poses_id","poses","links"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MapGraphArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MapGraphWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MapGraphWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapGraphArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapGraphArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MapGraphRefArray(size);
    this._refObject = new MapGraphRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MapGraphWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MapGraphArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MapGraphArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MapGraphWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MapGraphWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MapGraphArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MapGraphWrapper;
