// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__GetNodeData_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const GetNodeData_RequestRefStruct = StructType({
  ids: Int32Wrapper.refObjectArrayType,
  images: primitiveTypes.bool,
  scan: primitiveTypes.bool,
  grid: primitiveTypes.bool,
  user_data: primitiveTypes.bool,
});
const GetNodeData_RequestRefArray = ArrayType(GetNodeData_RequestRefStruct);
const GetNodeData_RequestRefStructArray = StructType({
  data: GetNodeData_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetNodeData_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetNodeData_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.ids = Int32Wrapper.createArray();
      this._wrapperFields.ids.copy(other._wrapperFields.ids);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetNodeData_RequestRefStruct();
    this._wrapperFields.ids = Int32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetNodeData_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetNodeData_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetNodeData_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetNodeData_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetNodeData_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.ids.fill(Int32Array.from([]));
    }
    this._wrapperFields.ids.freeze(own);
    this._refObject.ids = this._wrapperFields.ids.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.ids.copyRefObject(refObject.ids);
    this._refObject.images = refObject.images;
    this._refObject.scan = refObject.scan;
    this._refObject.grid = refObject.grid;
    this._refObject.user_data = refObject.user_data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.ids.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.ids);
      if (!Int32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.ids, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GetNodeData_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'GetNodeData_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ids() {
    return this._wrapperFields['ids'].data;
  }
  set ids(value) {
    this._wrapperFields['ids'].fill(value);
  }
  get images() {
    return this._refObject.images;
  }
  set images(value) {
    this._refObject.images = value;
  }
  get scan() {
    return this._refObject.scan;
  }
  set scan(value) {
    this._refObject.scan = value;
  }
  get grid() {
    return this._refObject.grid;
  }
  set grid(value) {
    this._refObject.grid = value;
  }
  get user_data() {
    return this._refObject.user_data;
  }
  set user_data(value) {
    this._refObject.user_data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetNodeData_RequestRefStruct(refObject.toObject());
    this._wrapperFields.ids.copyRefObject(this._refObject.ids);
  }
  copy(other) {
    this._refObject = new GetNodeData_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.ids.copy(other._wrapperFields.ids);
  }
  static get classType() {
    return GetNodeData_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"images","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scan","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"grid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"user_data","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GetNodeData_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetNodeData_Request"};
  }
  hasMember(name) {
    let memberNames = ["ids","images","scan","grid","user_data"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetNodeData_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetNodeData_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetNodeData_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodeData_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodeData_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetNodeData_RequestRefArray(size);
    this._refObject = new GetNodeData_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetNodeData_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetNodeData_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetNodeData_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetNodeData_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetNodeData_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetNodeData_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetNodeData_RequestWrapper;
