// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__GetNodesInRadius_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GetNodesInRadius_RequestRefStruct = StructType({
  node_id: primitiveTypes.int32,
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  z: primitiveTypes.float32,
  radius: primitiveTypes.float32,
  k: primitiveTypes.int32,
});
const GetNodesInRadius_RequestRefArray = ArrayType(GetNodesInRadius_RequestRefStruct);
const GetNodesInRadius_RequestRefStructArray = StructType({
  data: GetNodesInRadius_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetNodesInRadius_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetNodesInRadius_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetNodesInRadius_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new GetNodesInRadius_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetNodesInRadius_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetNodesInRadius_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetNodesInRadius_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetNodesInRadius_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.node_id = refObject.node_id;
    this._refObject.x = refObject.x;
    this._refObject.y = refObject.y;
    this._refObject.z = refObject.z;
    this._refObject.radius = refObject.radius;
    this._refObject.k = refObject.k;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    GetNodesInRadius_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'GetNodesInRadius_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node_id() {
    return this._refObject.node_id;
  }
  set node_id(value) {
    this._refObject.node_id = value;
  }
  get x() {
    return this._refObject.x;
  }
  set x(value) {
    this._refObject.x = value;
  }
  get y() {
    return this._refObject.y;
  }
  set y(value) {
    this._refObject.y = value;
  }
  get z() {
    return this._refObject.z;
  }
  set z(value) {
    this._refObject.z = value;
  }
  get radius() {
    return this._refObject.radius;
  }
  set radius(value) {
    this._refObject.radius = value;
  }
  get k() {
    return this._refObject.k;
  }
  set k(value) {
    this._refObject.k = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetNodesInRadius_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new GetNodesInRadius_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return GetNodesInRadius_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"k","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GetNodesInRadius_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetNodesInRadius_Request"};
  }
  hasMember(name) {
    let memberNames = ["node_id","x","y","z","radius","k"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetNodesInRadius_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetNodesInRadius_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetNodesInRadius_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodesInRadius_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodesInRadius_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetNodesInRadius_RequestRefArray(size);
    this._refObject = new GetNodesInRadius_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetNodesInRadius_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetNodesInRadius_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetNodesInRadius_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetNodesInRadius_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetNodesInRadius_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetNodesInRadius_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetNodesInRadius_RequestWrapper;
