// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__Goal.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GoalRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  node_id: primitiveTypes.int32,
  node_label: primitiveTypes.string,
  frame_id: primitiveTypes.string,
});
const GoalRefArray = ArrayType(GoalRefStruct);
const GoalRefStructArray = StructType({
  data: GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GoalWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.node_label =  new StringWrapper(other._wrapperFields.node_label);
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GoalRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.node_label = new StringWrapper();
    this._wrapperFields.frame_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GoalArrayWrapper();
  }
  static get ArrayType() {
    return GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return GoalRefStructArray;
  }
  static get refObjectType() {
    return GoalRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.node_label.freeze(own);
    }
    this._refObject.node_label = this._wrapperFields.node_label.refObject;
    if (own) {
      this._wrapperFields.frame_id.freeze(own);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.node_id = refObject.node_id;
    this._wrapperFields.node_label.data = refObject.node_label.data;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.node_label);
    StringWrapper.freeStruct(refObject.frame_id);
  }
  static destroyRawROS(msg) {
    GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get node_id() {
    return this._refObject.node_id;
  }
  set node_id(value) {
    this._refObject.node_id = value;
  }
  get node_label() {
    return this._wrapperFields.node_label.data;
  }
  set node_label(value) {
    this._wrapperFields.node_label.data = value;
  }
  get frame_id() {
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._wrapperFields.frame_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GoalRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.node_label.copyRefObject(this._refObject.node_label);
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
  }
  copy(other) {
    this._refObject = new GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.node_label.copy(other._wrapperFields.node_label);
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
  }
  static get classType() {
    return GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"node_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"node_label","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Goal"};
  }
  hasMember(name) {
    let memberNames = ["header","node_id","node_label","frame_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GoalWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GoalRefArray(size);
    this._refObject = new GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GoalWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GoalArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GoalArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GoalWrapper;
