// rclnodejs/generated/aruco_opencv_msgs/aruco_opencv_msgs__msg__ArucoDetection.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const MarkerPoseWrapper = require('../../generated/aruco_opencv_msgs/aruco_opencv_msgs__msg__MarkerPose.js');
const BoardPoseWrapper = require('../../generated/aruco_opencv_msgs/aruco_opencv_msgs__msg__BoardPose.js');
const ArucoDetectionRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  markers: MarkerPoseWrapper.refObjectArrayType,
  boards: BoardPoseWrapper.refObjectArrayType,
});
const ArucoDetectionRefArray = ArrayType(ArucoDetectionRefStruct);
const ArucoDetectionRefStructArray = StructType({
  data: ArucoDetectionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ArucoDetectionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ArucoDetectionRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.markers = MarkerPoseWrapper.createArray();
      this._wrapperFields.markers.copy(other._wrapperFields.markers);
      this._wrapperFields.boards = BoardPoseWrapper.createArray();
      this._wrapperFields.boards.copy(other._wrapperFields.boards);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ArucoDetectionRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.markers = MarkerPoseWrapper.createArray();
    this._wrapperFields.boards = BoardPoseWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ArucoDetectionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ArucoDetectionArrayWrapper();
  }
  static get ArrayType() {
    return ArucoDetectionArrayWrapper;
  }
  static get refObjectArrayType() {
    return ArucoDetectionRefStructArray;
  }
  static get refObjectType() {
    return ArucoDetectionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.markers.fill([]);
    }
    this._wrapperFields.markers.freeze(own);
    this._refObject.markers = this._wrapperFields.markers.refObject;
    if (own) {
      this._wrapperFields.boards.fill([]);
    }
    this._wrapperFields.boards.freeze(own);
    this._refObject.boards = this._wrapperFields.boards.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.markers.copyRefObject(refObject.markers);
    this._wrapperFields.boards.copyRefObject(refObject.boards);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.markers.size != 0) {
      MarkerPoseWrapper.ArrayType.freeArray(refObject.markers);
      if (!MarkerPoseWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.markers, MarkerPoseWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.boards.size != 0) {
      BoardPoseWrapper.ArrayType.freeArray(refObject.boards);
      if (!BoardPoseWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.boards, BoardPoseWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ArucoDetectionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'aruco_opencv_msgs', subFolder: 'msg', interfaceName: 'ArucoDetection'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get markers() {
    return this._wrapperFields.markers;
  }
  set markers(value) {
    this._wrapperFields.markers.fill(value);
  }
  get boards() {
    return this._wrapperFields.boards;
  }
  set boards(value) {
    this._wrapperFields.boards.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ArucoDetectionRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.markers.copyRefObject(this._refObject.markers);
    this._wrapperFields.boards.copyRefObject(this._refObject.boards);
  }
  copy(other) {
    this._refObject = new ArucoDetectionRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.markers.copy(other._wrapperFields.markers);
    this._wrapperFields.boards.copy(other._wrapperFields.boards);
  }
  static get classType() {
    return ArucoDetectionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"markers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"aruco_opencv_msgs","type":"MarkerPose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"boards","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"aruco_opencv_msgs","type":"BoardPose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"aruco_opencv_msgs","type":"ArucoDetection","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ArucoDetection"};
  }
  hasMember(name) {
    let memberNames = ["header","markers","boards"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ArucoDetectionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ArucoDetectionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ArucoDetectionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ArucoDetectionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ArucoDetectionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ArucoDetectionRefArray(size);
    this._refObject = new ArucoDetectionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ArucoDetectionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ArucoDetectionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ArucoDetectionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ArucoDetectionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ArucoDetectionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ArucoDetectionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ArucoDetectionWrapper;
