// rclnodejs/generated/aruco_opencv_msgs/aruco_opencv_msgs__msg__MarkerPose.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const MarkerPoseRefStruct = StructType({
  marker_id: primitiveTypes.uint16,
  pose: PoseWrapper.refObjectType,
});
const MarkerPoseRefArray = ArrayType(MarkerPoseRefStruct);
const MarkerPoseRefStructArray = StructType({
  data: MarkerPoseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MarkerPoseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MarkerPoseRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MarkerPoseRefStruct();
    this._wrapperFields.pose = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MarkerPoseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MarkerPoseArrayWrapper();
  }
  static get ArrayType() {
    return MarkerPoseArrayWrapper;
  }
  static get refObjectArrayType() {
    return MarkerPoseRefStructArray;
  }
  static get refObjectType() {
    return MarkerPoseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.pose.freeze(own);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.marker_id = refObject.marker_id;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.pose);
  }
  static destroyRawROS(msg) {
    MarkerPoseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'aruco_opencv_msgs', subFolder: 'msg', interfaceName: 'MarkerPose'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get marker_id() {
    return this._refObject.marker_id;
  }
  set marker_id(value) {
    this._refObject.marker_id = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MarkerPoseRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new MarkerPoseRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return MarkerPoseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"marker_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"aruco_opencv_msgs","type":"MarkerPose","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MarkerPose"};
  }
  hasMember(name) {
    let memberNames = ["marker_id","pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MarkerPoseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MarkerPoseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MarkerPoseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MarkerPoseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MarkerPoseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MarkerPoseRefArray(size);
    this._refObject = new MarkerPoseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MarkerPoseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MarkerPoseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MarkerPoseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MarkerPoseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MarkerPoseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MarkerPoseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MarkerPoseWrapper;
