// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__Inertia.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const InertiaRefStruct = StructType({
  m: primitiveTypes.float64,
  com: Vector3Wrapper.refObjectType,
  ixx: primitiveTypes.float64,
  ixy: primitiveTypes.float64,
  ixz: primitiveTypes.float64,
  iyy: primitiveTypes.float64,
  iyz: primitiveTypes.float64,
  izz: primitiveTypes.float64,
});
const InertiaRefArray = ArrayType(InertiaRefStruct);
const InertiaRefStructArray = StructType({
  data: InertiaRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class InertiaWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new InertiaRefStruct(other._refObject.toObject());
      this._wrapperFields.com =  new Vector3Wrapper(other._wrapperFields.com);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new InertiaRefStruct();
    this._wrapperFields.com = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new InertiaWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new InertiaArrayWrapper();
  }
  static get ArrayType() {
    return InertiaArrayWrapper;
  }
  static get refObjectArrayType() {
    return InertiaRefStructArray;
  }
  static get refObjectType() {
    return InertiaRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.com.freeze(own);
    this._refObject.com = this._wrapperFields.com.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.m = refObject.m;
    this._wrapperFields.com.copyRefObject(refObject.com);
    this._refObject.ixx = refObject.ixx;
    this._refObject.ixy = refObject.ixy;
    this._refObject.ixz = refObject.ixz;
    this._refObject.iyy = refObject.iyy;
    this._refObject.iyz = refObject.iyz;
    this._refObject.izz = refObject.izz;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.com);
  }
  static destroyRawROS(msg) {
    InertiaWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'Inertia'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get m() {
    return this._refObject.m;
  }
  set m(value) {
    this._refObject.m = value;
  }
  get com() {
    return this._wrapperFields.com;
  }
  set com(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.com.copy(value);
    } else {
      this._wrapperFields.com.copy(new Vector3Wrapper(value));
    }
  }
  get ixx() {
    return this._refObject.ixx;
  }
  set ixx(value) {
    this._refObject.ixx = value;
  }
  get ixy() {
    return this._refObject.ixy;
  }
  set ixy(value) {
    this._refObject.ixy = value;
  }
  get ixz() {
    return this._refObject.ixz;
  }
  set ixz(value) {
    this._refObject.ixz = value;
  }
  get iyy() {
    return this._refObject.iyy;
  }
  set iyy(value) {
    this._refObject.iyy = value;
  }
  get iyz() {
    return this._refObject.iyz;
  }
  set iyz(value) {
    this._refObject.iyz = value;
  }
  get izz() {
    return this._refObject.izz;
  }
  set izz(value) {
    this._refObject.izz = value;
  }
  copyRefObject(refObject) {
    this._refObject = new InertiaRefStruct(refObject.toObject());
    this._wrapperFields.com.copyRefObject(this._refObject.com);
  }
  copy(other) {
    this._refObject = new InertiaRefStruct(other._refObject.toObject());
    this._wrapperFields.com.copy(other._wrapperFields.com);
  }
  static get classType() {
    return InertiaWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"m","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"com","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"ixx","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iyy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iyz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"izz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"Inertia","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Inertia"};
  }
  hasMember(name) {
    let memberNames = ["m","com","ixx","ixy","ixz","iyy","iyz","izz"];
    return memberNames.indexOf(name) !== -1;
  }
}
class InertiaArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof InertiaWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new InertiaWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InertiaArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InertiaArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new InertiaRefArray(size);
    this._refObject = new InertiaRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new InertiaWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof InertiaArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "InertiaArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      InertiaWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return InertiaWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return InertiaArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = InertiaWrapper;
