// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__PolygonStamped.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PolygonWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Polygon.js');
const PolygonStampedRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  polygon: PolygonWrapper.refObjectType,
});
const PolygonStampedRefArray = ArrayType(PolygonStampedRefStruct);
const PolygonStampedRefStructArray = StructType({
  data: PolygonStampedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class PolygonStampedWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new PolygonStampedRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.polygon =  new PolygonWrapper(other._wrapperFields.polygon);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new PolygonStampedRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.polygon = new PolygonWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PolygonStampedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PolygonStampedArrayWrapper();
  }
  static get ArrayType() {
    return PolygonStampedArrayWrapper;
  }
  static get refObjectArrayType() {
    return PolygonStampedRefStructArray;
  }
  static get refObjectType() {
    return PolygonStampedRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.polygon.freeze(own);
    this._refObject.polygon = this._wrapperFields.polygon.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.polygon.copyRefObject(refObject.polygon);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PolygonWrapper.freeStruct(refObject.polygon);
  }
  static destroyRawROS(msg) {
    PolygonStampedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'PolygonStamped'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get polygon() {
    return this._wrapperFields.polygon;
  }
  set polygon(value) {
    if (value instanceof PolygonWrapper) {
      this._wrapperFields.polygon.copy(value);
    } else {
      this._wrapperFields.polygon.copy(new PolygonWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new PolygonStampedRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.polygon.copyRefObject(this._refObject.polygon);
  }
  copy(other) {
    this._refObject = new PolygonStampedRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.polygon.copy(other._wrapperFields.polygon);
  }
  static get classType() {
    return PolygonStampedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"polygon","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Polygon","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"PolygonStamped","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PolygonStamped"};
  }
  hasMember(name) {
    let memberNames = ["header","polygon"];
    return memberNames.indexOf(name) !== -1;
  }
}
class PolygonStampedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PolygonStampedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PolygonStampedWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PolygonStampedArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PolygonStampedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new PolygonStampedRefArray(size);
    this._refObject = new PolygonStampedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PolygonStampedWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PolygonStampedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PolygonStampedArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PolygonStampedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PolygonStampedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PolygonStampedArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = PolygonStampedWrapper;
