// rclnodejs/generated/action_msgs/action_msgs__srv__CancelGoal_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GoalInfoWrapper = require('../../generated/action_msgs/action_msgs__msg__GoalInfo.js');
const CancelGoal_ResponseRefStruct = StructType({
  return_code: primitiveTypes.int8,
  goals_canceling: GoalInfoWrapper.refObjectArrayType,
});
const CancelGoal_ResponseRefArray = ArrayType(CancelGoal_ResponseRefStruct);
const CancelGoal_ResponseRefStructArray = StructType({
  data: CancelGoal_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class CancelGoal_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new CancelGoal_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.goals_canceling = GoalInfoWrapper.createArray();
      this._wrapperFields.goals_canceling.copy(other._wrapperFields.goals_canceling);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new CancelGoal_ResponseRefStruct();
    this._wrapperFields.goals_canceling = GoalInfoWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new CancelGoal_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CancelGoal_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return CancelGoal_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return CancelGoal_ResponseRefStructArray;
  }
  static get refObjectType() {
    return CancelGoal_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.goals_canceling.fill([]);
    }
    this._wrapperFields.goals_canceling.freeze(own);
    this._refObject.goals_canceling = this._wrapperFields.goals_canceling.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.return_code = refObject.return_code;
    this._wrapperFields.goals_canceling.copyRefObject(refObject.goals_canceling);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.goals_canceling.size != 0) {
      GoalInfoWrapper.ArrayType.freeArray(refObject.goals_canceling);
      if (!GoalInfoWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.goals_canceling, GoalInfoWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    CancelGoal_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_msgs', subFolder: 'srv', interfaceName: 'CancelGoal_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get return_code() {
    return this._refObject.return_code;
  }
  set return_code(value) {
    this._refObject.return_code = value;
  }
  get goals_canceling() {
    return this._wrapperFields.goals_canceling;
  }
  set goals_canceling(value) {
    this._wrapperFields.goals_canceling.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new CancelGoal_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.goals_canceling.copyRefObject(this._refObject.goals_canceling);
  }
  copy(other) {
    this._refObject = new CancelGoal_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.goals_canceling.copy(other._wrapperFields.goals_canceling);
  }
  static get classType() {
    return CancelGoal_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"ERROR_NONE","value":0},{"type":"int8","name":"ERROR_REJECTED","value":1},{"type":"int8","name":"ERROR_UNKNOWN_GOAL_ID","value":2},{"type":"int8","name":"ERROR_GOAL_TERMINATED","value":3}],"fields":[{"name":"return_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"goals_canceling","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"action_msgs","type":"GoalInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"action_msgs","type":"CancelGoal_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CancelGoal_Response"};
  }
  hasMember(name) {
    let memberNames = ["return_code","goals_canceling"];
    return memberNames.indexOf(name) !== -1;
  }
}
class CancelGoal_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CancelGoal_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CancelGoal_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CancelGoal_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CancelGoal_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new CancelGoal_ResponseRefArray(size);
    this._refObject = new CancelGoal_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CancelGoal_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CancelGoal_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CancelGoal_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CancelGoal_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CancelGoal_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CancelGoal_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(CancelGoal_ResponseWrapper, "ERROR_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CancelGoal_ResponseWrapper, "ERROR_REJECTED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CancelGoal_ResponseWrapper, "ERROR_UNKNOWN_GOAL_ID", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CancelGoal_ResponseWrapper, "ERROR_GOAL_TERMINATED", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = CancelGoal_ResponseWrapper;
