// rclnodejs/generated/map_msgs/map_msgs__msg__ProjectedMapInfo.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ProjectedMapInfoRefStruct = StructType({
  frame_id: primitiveTypes.string,
  x: primitiveTypes.float64,
  y: primitiveTypes.float64,
  width: primitiveTypes.float64,
  height: primitiveTypes.float64,
  min_z: primitiveTypes.float64,
  max_z: primitiveTypes.float64,
});
const ProjectedMapInfoRefArray = ArrayType(ProjectedMapInfoRefStruct);
const ProjectedMapInfoRefStructArray = StructType({
  data: ProjectedMapInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ProjectedMapInfoWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ProjectedMapInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ProjectedMapInfoRefStruct();
    this._wrapperFields.frame_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ProjectedMapInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ProjectedMapInfoArrayWrapper();
  }
  static get ArrayType() {
    return ProjectedMapInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return ProjectedMapInfoRefStructArray;
  }
  static get refObjectType() {
    return ProjectedMapInfoRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.frame_id.freeze(own);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
    this._refObject.x = refObject.x;
    this._refObject.y = refObject.y;
    this._refObject.width = refObject.width;
    this._refObject.height = refObject.height;
    this._refObject.min_z = refObject.min_z;
    this._refObject.max_z = refObject.max_z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.frame_id);
  }
  static destroyRawROS(msg) {
    ProjectedMapInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'msg', interfaceName: 'ProjectedMapInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frame_id() {
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._wrapperFields.frame_id.data = value;
  }
  get x() {
    return this._refObject.x;
  }
  set x(value) {
    this._refObject.x = value;
  }
  get y() {
    return this._refObject.y;
  }
  set y(value) {
    this._refObject.y = value;
  }
  get width() {
    return this._refObject.width;
  }
  set width(value) {
    this._refObject.width = value;
  }
  get height() {
    return this._refObject.height;
  }
  set height(value) {
    this._refObject.height = value;
  }
  get min_z() {
    return this._refObject.min_z;
  }
  set min_z(value) {
    this._refObject.min_z = value;
  }
  get max_z() {
    return this._refObject.max_z;
  }
  set max_z(value) {
    this._refObject.max_z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ProjectedMapInfoRefStruct(refObject.toObject());
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
  }
  copy(other) {
    this._refObject = new ProjectedMapInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
  }
  static get classType() {
    return ProjectedMapInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"ProjectedMapInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ProjectedMapInfo"};
  }
  hasMember(name) {
    let memberNames = ["frame_id","x","y","width","height","min_z","max_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ProjectedMapInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ProjectedMapInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ProjectedMapInfoWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProjectedMapInfoArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProjectedMapInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ProjectedMapInfoRefArray(size);
    this._refObject = new ProjectedMapInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ProjectedMapInfoWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ProjectedMapInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ProjectedMapInfoArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ProjectedMapInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ProjectedMapInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ProjectedMapInfoArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ProjectedMapInfoWrapper;
