// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__MultiEchoLaserScan.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const LaserEchoWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__LaserEcho.js');
const MultiEchoLaserScanRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  angle_min: primitiveTypes.float32,
  angle_max: primitiveTypes.float32,
  angle_increment: primitiveTypes.float32,
  time_increment: primitiveTypes.float32,
  scan_time: primitiveTypes.float32,
  range_min: primitiveTypes.float32,
  range_max: primitiveTypes.float32,
  ranges: LaserEchoWrapper.refObjectArrayType,
  intensities: LaserEchoWrapper.refObjectArrayType,
});
const MultiEchoLaserScanRefArray = ArrayType(MultiEchoLaserScanRefStruct);
const MultiEchoLaserScanRefStructArray = StructType({
  data: MultiEchoLaserScanRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MultiEchoLaserScanWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MultiEchoLaserScanRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.ranges = LaserEchoWrapper.createArray();
      this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
      this._wrapperFields.intensities = LaserEchoWrapper.createArray();
      this._wrapperFields.intensities.copy(other._wrapperFields.intensities);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MultiEchoLaserScanRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.ranges = LaserEchoWrapper.createArray();
    this._wrapperFields.intensities = LaserEchoWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MultiEchoLaserScanWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MultiEchoLaserScanArrayWrapper();
  }
  static get ArrayType() {
    return MultiEchoLaserScanArrayWrapper;
  }
  static get refObjectArrayType() {
    return MultiEchoLaserScanRefStructArray;
  }
  static get refObjectType() {
    return MultiEchoLaserScanRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.ranges.fill([]);
    }
    this._wrapperFields.ranges.freeze(own);
    this._refObject.ranges = this._wrapperFields.ranges.refObject;
    if (own) {
      this._wrapperFields.intensities.fill([]);
    }
    this._wrapperFields.intensities.freeze(own);
    this._refObject.intensities = this._wrapperFields.intensities.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.angle_min = refObject.angle_min;
    this._refObject.angle_max = refObject.angle_max;
    this._refObject.angle_increment = refObject.angle_increment;
    this._refObject.time_increment = refObject.time_increment;
    this._refObject.scan_time = refObject.scan_time;
    this._refObject.range_min = refObject.range_min;
    this._refObject.range_max = refObject.range_max;
    this._wrapperFields.ranges.copyRefObject(refObject.ranges);
    this._wrapperFields.intensities.copyRefObject(refObject.intensities);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.ranges.size != 0) {
      LaserEchoWrapper.ArrayType.freeArray(refObject.ranges);
      if (!LaserEchoWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.ranges, LaserEchoWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.intensities.size != 0) {
      LaserEchoWrapper.ArrayType.freeArray(refObject.intensities);
      if (!LaserEchoWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.intensities, LaserEchoWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MultiEchoLaserScanWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'MultiEchoLaserScan'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get angle_min() {
    return this._refObject.angle_min;
  }
  set angle_min(value) {
    this._refObject.angle_min = value;
  }
  get angle_max() {
    return this._refObject.angle_max;
  }
  set angle_max(value) {
    this._refObject.angle_max = value;
  }
  get angle_increment() {
    return this._refObject.angle_increment;
  }
  set angle_increment(value) {
    this._refObject.angle_increment = value;
  }
  get time_increment() {
    return this._refObject.time_increment;
  }
  set time_increment(value) {
    this._refObject.time_increment = value;
  }
  get scan_time() {
    return this._refObject.scan_time;
  }
  set scan_time(value) {
    this._refObject.scan_time = value;
  }
  get range_min() {
    return this._refObject.range_min;
  }
  set range_min(value) {
    this._refObject.range_min = value;
  }
  get range_max() {
    return this._refObject.range_max;
  }
  set range_max(value) {
    this._refObject.range_max = value;
  }
  get ranges() {
    return this._wrapperFields.ranges;
  }
  set ranges(value) {
    this._wrapperFields.ranges.fill(value);
  }
  get intensities() {
    return this._wrapperFields.intensities;
  }
  set intensities(value) {
    this._wrapperFields.intensities.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MultiEchoLaserScanRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.ranges.copyRefObject(this._refObject.ranges);
    this._wrapperFields.intensities.copyRefObject(this._refObject.intensities);
  }
  copy(other) {
    this._refObject = new MultiEchoLaserScanRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
    this._wrapperFields.intensities.copy(other._wrapperFields.intensities);
  }
  static get classType() {
    return MultiEchoLaserScanWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"angle_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle_increment","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_increment","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scan_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"range_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"range_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ranges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"sensor_msgs","type":"LaserEcho","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"intensities","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"sensor_msgs","type":"LaserEcho","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"MultiEchoLaserScan","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MultiEchoLaserScan"};
  }
  hasMember(name) {
    let memberNames = ["header","angle_min","angle_max","angle_increment","time_increment","scan_time","range_min","range_max","ranges","intensities"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MultiEchoLaserScanArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MultiEchoLaserScanWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MultiEchoLaserScanWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiEchoLaserScanArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiEchoLaserScanArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MultiEchoLaserScanRefArray(size);
    this._refObject = new MultiEchoLaserScanRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MultiEchoLaserScanWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MultiEchoLaserScanArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MultiEchoLaserScanArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MultiEchoLaserScanWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MultiEchoLaserScanWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MultiEchoLaserScanArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MultiEchoLaserScanWrapper;
