// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__Range.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RangeRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  radiation_type: primitiveTypes.uint8,
  field_of_view: primitiveTypes.float32,
  min_range: primitiveTypes.float32,
  max_range: primitiveTypes.float32,
  range: primitiveTypes.float32,
});
const RangeRefArray = ArrayType(RangeRefStruct);
const RangeRefStructArray = StructType({
  data: RangeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RangeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RangeRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RangeRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RangeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RangeArrayWrapper();
  }
  static get ArrayType() {
    return RangeArrayWrapper;
  }
  static get refObjectArrayType() {
    return RangeRefStructArray;
  }
  static get refObjectType() {
    return RangeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.radiation_type = refObject.radiation_type;
    this._refObject.field_of_view = refObject.field_of_view;
    this._refObject.min_range = refObject.min_range;
    this._refObject.max_range = refObject.max_range;
    this._refObject.range = refObject.range;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destroyRawROS(msg) {
    RangeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'Range'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get radiation_type() {
    return this._refObject.radiation_type;
  }
  set radiation_type(value) {
    this._refObject.radiation_type = value;
  }
  get field_of_view() {
    return this._refObject.field_of_view;
  }
  set field_of_view(value) {
    this._refObject.field_of_view = value;
  }
  get min_range() {
    return this._refObject.min_range;
  }
  set min_range(value) {
    this._refObject.min_range = value;
  }
  get max_range() {
    return this._refObject.max_range;
  }
  set max_range(value) {
    this._refObject.max_range = value;
  }
  get range() {
    return this._refObject.range;
  }
  set range(value) {
    this._refObject.range = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RangeRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
  }
  copy(other) {
    this._refObject = new RangeRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
  }
  static get classType() {
    return RangeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ULTRASOUND","value":0},{"type":"uint8","name":"INFRARED","value":1}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"radiation_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"field_of_view","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_range","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_range","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"range","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"Range","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Range"};
  }
  hasMember(name) {
    let memberNames = ["header","radiation_type","field_of_view","min_range","max_range","range"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RangeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RangeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RangeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RangeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RangeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RangeRefArray(size);
    this._refObject = new RangeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RangeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RangeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RangeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RangeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RangeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RangeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(RangeWrapper, "ULTRASOUND", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RangeWrapper, "INFRARED", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = RangeWrapper;
