// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__Joy.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const JoyRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  axes: Float32Wrapper.refObjectArrayType,
  buttons: Int32Wrapper.refObjectArrayType,
});
const JoyRefArray = ArrayType(JoyRefStruct);
const JoyRefStructArray = StructType({
  data: JoyRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class JoyWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new JoyRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.axes = Float32Wrapper.createArray();
      this._wrapperFields.axes.copy(other._wrapperFields.axes);
      this._wrapperFields.buttons = Int32Wrapper.createArray();
      this._wrapperFields.buttons.copy(other._wrapperFields.buttons);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new JoyRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.axes = Float32Wrapper.createArray();
    this._wrapperFields.buttons = Int32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new JoyWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JoyArrayWrapper();
  }
  static get ArrayType() {
    return JoyArrayWrapper;
  }
  static get refObjectArrayType() {
    return JoyRefStructArray;
  }
  static get refObjectType() {
    return JoyRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.axes.fill(Float32Array.from([]));
    }
    this._wrapperFields.axes.freeze(own);
    this._refObject.axes = this._wrapperFields.axes.refObject;
    if (own) {
      this._wrapperFields.buttons.fill(Int32Array.from([]));
    }
    this._wrapperFields.buttons.freeze(own);
    this._refObject.buttons = this._wrapperFields.buttons.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.axes.copyRefObject(refObject.axes);
    this._wrapperFields.buttons.copyRefObject(refObject.buttons);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.axes.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.axes);
      if (!Float32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.axes, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.buttons.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.buttons);
      if (!Int32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.buttons, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    JoyWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'Joy'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get axes() {
    return this._wrapperFields['axes'].data;
  }
  set axes(value) {
    this._wrapperFields['axes'].fill(value);
  }
  get buttons() {
    return this._wrapperFields['buttons'].data;
  }
  set buttons(value) {
    this._wrapperFields['buttons'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new JoyRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.axes.copyRefObject(this._refObject.axes);
    this._wrapperFields.buttons.copyRefObject(this._refObject.buttons);
  }
  copy(other) {
    this._refObject = new JoyRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.axes.copy(other._wrapperFields.axes);
    this._wrapperFields.buttons.copy(other._wrapperFields.buttons);
  }
  static get classType() {
    return JoyWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"axes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"buttons","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"Joy","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Joy"};
  }
  hasMember(name) {
    let memberNames = ["header","axes","buttons"];
    return memberNames.indexOf(name) !== -1;
  }
}
class JoyArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JoyWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JoyWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JoyArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JoyArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new JoyRefArray(size);
    this._refObject = new JoyRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JoyWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JoyArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JoyArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JoyWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JoyWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JoyArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = JoyWrapper;
