// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__ChannelFloat32.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const ChannelFloat32RefStruct = StructType({
  name: primitiveTypes.string,
  values: Float32Wrapper.refObjectArrayType,
});
const ChannelFloat32RefArray = ArrayType(ChannelFloat32RefStruct);
const ChannelFloat32RefStructArray = StructType({
  data: ChannelFloat32RefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ChannelFloat32Wrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ChannelFloat32RefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.values = Float32Wrapper.createArray();
      this._wrapperFields.values.copy(other._wrapperFields.values);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ChannelFloat32RefStruct();
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.values = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ChannelFloat32Wrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ChannelFloat32ArrayWrapper();
  }
  static get ArrayType() {
    return ChannelFloat32ArrayWrapper;
  }
  static get refObjectArrayType() {
    return ChannelFloat32RefStructArray;
  }
  static get refObjectType() {
    return ChannelFloat32RefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.values.fill(Float32Array.from([]));
    }
    this._wrapperFields.values.freeze(own);
    this._refObject.values = this._wrapperFields.values.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._wrapperFields.values.copyRefObject(refObject.values);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    if (refObject.values.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.values);
      if (!Float32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.values, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ChannelFloat32Wrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'ChannelFloat32'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get values() {
    return this._wrapperFields['values'].data;
  }
  set values(value) {
    this._wrapperFields['values'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ChannelFloat32RefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._wrapperFields.values.copyRefObject(this._refObject.values);
  }
  copy(other) {
    this._refObject = new ChannelFloat32RefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._wrapperFields.values.copy(other._wrapperFields.values);
  }
  static get classType() {
    return ChannelFloat32Wrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"ChannelFloat32","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ChannelFloat32"};
  }
  hasMember(name) {
    let memberNames = ["name","values"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ChannelFloat32ArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ChannelFloat32Wrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ChannelFloat32Wrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChannelFloat32ArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChannelFloat32ArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ChannelFloat32RefArray(size);
    this._refObject = new ChannelFloat32RefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ChannelFloat32Wrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ChannelFloat32ArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ChannelFloat32ArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ChannelFloat32Wrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ChannelFloat32Wrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ChannelFloat32ArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ChannelFloat32Wrapper;
