// rclnodejs/generated/turtlesim/turtlesim__srv__TeleportRelative_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TeleportRelative_RequestRefStruct = StructType({
  linear: primitiveTypes.float32,
  angular: primitiveTypes.float32,
});
const TeleportRelative_RequestRefArray = ArrayType(TeleportRelative_RequestRefStruct);
const TeleportRelative_RequestRefStructArray = StructType({
  data: TeleportRelative_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TeleportRelative_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TeleportRelative_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TeleportRelative_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new TeleportRelative_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TeleportRelative_RequestArrayWrapper();
  }
  static get ArrayType() {
    return TeleportRelative_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return TeleportRelative_RequestRefStructArray;
  }
  static get refObjectType() {
    return TeleportRelative_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.linear = refObject.linear;
    this._refObject.angular = refObject.angular;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    TeleportRelative_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'srv', interfaceName: 'TeleportRelative_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get linear() {
    return this._refObject.linear;
  }
  set linear(value) {
    this._refObject.linear = value;
  }
  get angular() {
    return this._refObject.angular;
  }
  set angular(value) {
    this._refObject.angular = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TeleportRelative_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new TeleportRelative_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return TeleportRelative_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"linear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"TeleportRelative_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TeleportRelative_Request"};
  }
  hasMember(name) {
    let memberNames = ["linear","angular"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TeleportRelative_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TeleportRelative_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TeleportRelative_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportRelative_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportRelative_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TeleportRelative_RequestRefArray(size);
    this._refObject = new TeleportRelative_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TeleportRelative_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TeleportRelative_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TeleportRelative_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TeleportRelative_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TeleportRelative_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TeleportRelative_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TeleportRelative_RequestWrapper;
