// rclnodejs/generated/turtlesim/turtlesim__action__RotateAbsolute_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RotateAbsolute_ResultRefStruct = StructType({
  delta: primitiveTypes.float32,
});
const RotateAbsolute_ResultRefArray = ArrayType(RotateAbsolute_ResultRefStruct);
const RotateAbsolute_ResultRefStructArray = StructType({
  data: RotateAbsolute_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RotateAbsolute_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RotateAbsolute_ResultRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RotateAbsolute_ResultRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new RotateAbsolute_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RotateAbsolute_ResultArrayWrapper();
  }
  static get ArrayType() {
    return RotateAbsolute_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return RotateAbsolute_ResultRefStructArray;
  }
  static get refObjectType() {
    return RotateAbsolute_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.delta = refObject.delta;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    RotateAbsolute_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'action', interfaceName: 'RotateAbsolute_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get delta() {
    return this._refObject.delta;
  }
  set delta(value) {
    this._refObject.delta = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RotateAbsolute_ResultRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new RotateAbsolute_ResultRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return RotateAbsolute_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"delta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"RotateAbsolute_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RotateAbsolute_Result"};
  }
  hasMember(name) {
    let memberNames = ["delta"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RotateAbsolute_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RotateAbsolute_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RotateAbsolute_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAbsolute_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAbsolute_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RotateAbsolute_ResultRefArray(size);
    this._refObject = new RotateAbsolute_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RotateAbsolute_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RotateAbsolute_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RotateAbsolute_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RotateAbsolute_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RotateAbsolute_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RotateAbsolute_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RotateAbsolute_ResultWrapper;
