// rclnodejs/generated/turtlesim/turtlesim__srv__Spawn_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Spawn_RequestRefStruct = StructType({
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  theta: primitiveTypes.float32,
  name: primitiveTypes.string,
});
const Spawn_RequestRefArray = ArrayType(Spawn_RequestRefStruct);
const Spawn_RequestRefStructArray = StructType({
  data: Spawn_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Spawn_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Spawn_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Spawn_RequestRefStruct();
    this._wrapperFields.name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Spawn_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Spawn_RequestArrayWrapper();
  }
  static get ArrayType() {
    return Spawn_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return Spawn_RequestRefStructArray;
  }
  static get refObjectType() {
    return Spawn_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.x = refObject.x;
    this._refObject.y = refObject.y;
    this._refObject.theta = refObject.theta;
    this._wrapperFields.name.data = refObject.name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destroyRawROS(msg) {
    Spawn_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'srv', interfaceName: 'Spawn_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get x() {
    return this._refObject.x;
  }
  set x(value) {
    this._refObject.x = value;
  }
  get y() {
    return this._refObject.y;
  }
  set y(value) {
    this._refObject.y = value;
  }
  get theta() {
    return this._refObject.theta;
  }
  set theta(value) {
    this._refObject.theta = value;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Spawn_RequestRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
  }
  copy(other) {
    this._refObject = new Spawn_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
  }
  static get classType() {
    return Spawn_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"theta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"Spawn_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Spawn_Request"};
  }
  hasMember(name) {
    let memberNames = ["x","y","theta","name"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Spawn_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Spawn_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Spawn_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Spawn_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Spawn_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Spawn_RequestRefArray(size);
    this._refObject = new Spawn_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Spawn_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Spawn_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Spawn_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Spawn_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Spawn_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Spawn_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Spawn_RequestWrapper;
