// rclnodejs/generated/turtlesim/turtlesim__srv__TeleportAbsolute_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TeleportAbsolute_RequestRefStruct = StructType({
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  theta: primitiveTypes.float32,
});
const TeleportAbsolute_RequestRefArray = ArrayType(TeleportAbsolute_RequestRefStruct);
const TeleportAbsolute_RequestRefStructArray = StructType({
  data: TeleportAbsolute_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TeleportAbsolute_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TeleportAbsolute_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TeleportAbsolute_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new TeleportAbsolute_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TeleportAbsolute_RequestArrayWrapper();
  }
  static get ArrayType() {
    return TeleportAbsolute_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return TeleportAbsolute_RequestRefStructArray;
  }
  static get refObjectType() {
    return TeleportAbsolute_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.x = refObject.x;
    this._refObject.y = refObject.y;
    this._refObject.theta = refObject.theta;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    TeleportAbsolute_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'srv', interfaceName: 'TeleportAbsolute_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get x() {
    return this._refObject.x;
  }
  set x(value) {
    this._refObject.x = value;
  }
  get y() {
    return this._refObject.y;
  }
  set y(value) {
    this._refObject.y = value;
  }
  get theta() {
    return this._refObject.theta;
  }
  set theta(value) {
    this._refObject.theta = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TeleportAbsolute_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new TeleportAbsolute_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return TeleportAbsolute_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"theta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"TeleportAbsolute_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TeleportAbsolute_Request"};
  }
  hasMember(name) {
    let memberNames = ["x","y","theta"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TeleportAbsolute_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TeleportAbsolute_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TeleportAbsolute_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportAbsolute_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportAbsolute_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TeleportAbsolute_RequestRefArray(size);
    this._refObject = new TeleportAbsolute_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TeleportAbsolute_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TeleportAbsolute_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TeleportAbsolute_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TeleportAbsolute_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TeleportAbsolute_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TeleportAbsolute_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TeleportAbsolute_RequestWrapper;
