// rclnodejs/generated/dwb_msgs/dwb_msgs__msg__CriticScore.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CriticScoreRefStruct = StructType({
  name: primitiveTypes.string,
  raw_score: primitiveTypes.float32,
  scale: primitiveTypes.float32,
});
const CriticScoreRefArray = ArrayType(CriticScoreRefStruct);
const CriticScoreRefStructArray = StructType({
  data: CriticScoreRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class CriticScoreWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new CriticScoreRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new CriticScoreRefStruct();
    this._wrapperFields.name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CriticScoreWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CriticScoreArrayWrapper();
  }
  static get ArrayType() {
    return CriticScoreArrayWrapper;
  }
  static get refObjectArrayType() {
    return CriticScoreRefStructArray;
  }
  static get refObjectType() {
    return CriticScoreRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._refObject.raw_score = refObject.raw_score;
    this._refObject.scale = refObject.scale;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destroyRawROS(msg) {
    CriticScoreWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'msg', interfaceName: 'CriticScore'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get raw_score() {
    return this._refObject.raw_score;
  }
  set raw_score(value) {
    this._refObject.raw_score = value;
  }
  get scale() {
    return this._refObject.scale;
  }
  set scale(value) {
    this._refObject.scale = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CriticScoreRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
  }
  copy(other) {
    this._refObject = new CriticScoreRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
  }
  static get classType() {
    return CriticScoreWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"raw_score","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scale","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"CriticScore","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CriticScore"};
  }
  hasMember(name) {
    let memberNames = ["name","raw_score","scale"];
    return memberNames.indexOf(name) !== -1;
  }
}
class CriticScoreArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CriticScoreWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CriticScoreWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CriticScoreArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CriticScoreArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new CriticScoreRefArray(size);
    this._refObject = new CriticScoreRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CriticScoreWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CriticScoreArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CriticScoreArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CriticScoreWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CriticScoreWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CriticScoreArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = CriticScoreWrapper;
