// rclnodejs/generated/dwb_msgs/dwb_msgs__srv__GenerateTwists_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Twist2DWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Twist2D.js');
const GenerateTwists_ResponseRefStruct = StructType({
  twists: Twist2DWrapper.refObjectArrayType,
});
const GenerateTwists_ResponseRefArray = ArrayType(GenerateTwists_ResponseRefStruct);
const GenerateTwists_ResponseRefStructArray = StructType({
  data: GenerateTwists_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GenerateTwists_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GenerateTwists_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.twists = Twist2DWrapper.createArray();
      this._wrapperFields.twists.copy(other._wrapperFields.twists);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GenerateTwists_ResponseRefStruct();
    this._wrapperFields.twists = Twist2DWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GenerateTwists_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GenerateTwists_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GenerateTwists_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GenerateTwists_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GenerateTwists_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.twists.fill([]);
    }
    this._wrapperFields.twists.freeze(own);
    this._refObject.twists = this._wrapperFields.twists.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.twists.copyRefObject(refObject.twists);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.twists.size != 0) {
      Twist2DWrapper.ArrayType.freeArray(refObject.twists);
      if (!Twist2DWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.twists, Twist2DWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GenerateTwists_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'srv', interfaceName: 'GenerateTwists_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get twists() {
    return this._wrapperFields.twists;
  }
  set twists(value) {
    this._wrapperFields.twists.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GenerateTwists_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.twists.copyRefObject(this._refObject.twists);
  }
  copy(other) {
    this._refObject = new GenerateTwists_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.twists.copy(other._wrapperFields.twists);
  }
  static get classType() {
    return GenerateTwists_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"twists","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"nav_2d_msgs","type":"Twist2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"GenerateTwists_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GenerateTwists_Response"};
  }
  hasMember(name) {
    let memberNames = ["twists"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GenerateTwists_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GenerateTwists_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GenerateTwists_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTwists_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTwists_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GenerateTwists_ResponseRefArray(size);
    this._refObject = new GenerateTwists_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GenerateTwists_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GenerateTwists_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GenerateTwists_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GenerateTwists_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GenerateTwists_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GenerateTwists_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GenerateTwists_ResponseWrapper;
