// rclnodejs/generated/dwb_msgs/dwb_msgs__srv__GenerateTrajectory_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Trajectory2DWrapper = require('../../generated/dwb_msgs/dwb_msgs__msg__Trajectory2D.js');
const GenerateTrajectory_ResponseRefStruct = StructType({
  traj: Trajectory2DWrapper.refObjectType,
});
const GenerateTrajectory_ResponseRefArray = ArrayType(GenerateTrajectory_ResponseRefStruct);
const GenerateTrajectory_ResponseRefStructArray = StructType({
  data: GenerateTrajectory_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GenerateTrajectory_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GenerateTrajectory_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.traj =  new Trajectory2DWrapper(other._wrapperFields.traj);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GenerateTrajectory_ResponseRefStruct();
    this._wrapperFields.traj = new Trajectory2DWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GenerateTrajectory_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GenerateTrajectory_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GenerateTrajectory_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GenerateTrajectory_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GenerateTrajectory_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.traj.freeze(own);
    this._refObject.traj = this._wrapperFields.traj.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.traj.copyRefObject(refObject.traj);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Trajectory2DWrapper.freeStruct(refObject.traj);
  }
  static destroyRawROS(msg) {
    GenerateTrajectory_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'srv', interfaceName: 'GenerateTrajectory_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get traj() {
    return this._wrapperFields.traj;
  }
  set traj(value) {
    if (value instanceof Trajectory2DWrapper) {
      this._wrapperFields.traj.copy(value);
    } else {
      this._wrapperFields.traj.copy(new Trajectory2DWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GenerateTrajectory_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.traj.copyRefObject(this._refObject.traj);
  }
  copy(other) {
    this._refObject = new GenerateTrajectory_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.traj.copy(other._wrapperFields.traj);
  }
  static get classType() {
    return GenerateTrajectory_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"traj","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"dwb_msgs","type":"Trajectory2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"GenerateTrajectory_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GenerateTrajectory_Response"};
  }
  hasMember(name) {
    let memberNames = ["traj"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GenerateTrajectory_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GenerateTrajectory_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GenerateTrajectory_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTrajectory_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTrajectory_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GenerateTrajectory_ResponseRefArray(size);
    this._refObject = new GenerateTrajectory_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GenerateTrajectory_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GenerateTrajectory_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GenerateTrajectory_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GenerateTrajectory_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GenerateTrajectory_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GenerateTrajectory_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GenerateTrajectory_ResponseWrapper;
