// rclnodejs/generated/apriltag_msgs/apriltag_msgs__msg__AprilTagDetectionArray.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const AprilTagDetectionWrapper = require('../../generated/apriltag_msgs/apriltag_msgs__msg__AprilTagDetection.js');
const AprilTagDetectionArrayRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  detections: AprilTagDetectionWrapper.refObjectArrayType,
});
const AprilTagDetectionArrayRefArray = ArrayType(AprilTagDetectionArrayRefStruct);
const AprilTagDetectionArrayRefStructArray = StructType({
  data: AprilTagDetectionArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AprilTagDetectionArrayWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AprilTagDetectionArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.detections = AprilTagDetectionWrapper.createArray();
      this._wrapperFields.detections.copy(other._wrapperFields.detections);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AprilTagDetectionArrayRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.detections = AprilTagDetectionWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new AprilTagDetectionArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AprilTagDetectionArrayArrayWrapper();
  }
  static get ArrayType() {
    return AprilTagDetectionArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return AprilTagDetectionArrayRefStructArray;
  }
  static get refObjectType() {
    return AprilTagDetectionArrayRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.detections.fill([]);
    }
    this._wrapperFields.detections.freeze(own);
    this._refObject.detections = this._wrapperFields.detections.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.detections.copyRefObject(refObject.detections);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.detections.size != 0) {
      AprilTagDetectionWrapper.ArrayType.freeArray(refObject.detections);
      if (!AprilTagDetectionWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.detections, AprilTagDetectionWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    AprilTagDetectionArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'apriltag_msgs', subFolder: 'msg', interfaceName: 'AprilTagDetectionArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get detections() {
    return this._wrapperFields.detections;
  }
  set detections(value) {
    this._wrapperFields.detections.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new AprilTagDetectionArrayRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.detections.copyRefObject(this._refObject.detections);
  }
  copy(other) {
    this._refObject = new AprilTagDetectionArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.detections.copy(other._wrapperFields.detections);
  }
  static get classType() {
    return AprilTagDetectionArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"detections","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"apriltag_msgs","type":"AprilTagDetection","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"apriltag_msgs","type":"AprilTagDetectionArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AprilTagDetectionArray"};
  }
  hasMember(name) {
    let memberNames = ["header","detections"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AprilTagDetectionArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AprilTagDetectionArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AprilTagDetectionArrayWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AprilTagDetectionArrayArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AprilTagDetectionArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AprilTagDetectionArrayRefArray(size);
    this._refObject = new AprilTagDetectionArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AprilTagDetectionArrayWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AprilTagDetectionArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AprilTagDetectionArrayArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AprilTagDetectionArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AprilTagDetectionArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AprilTagDetectionArrayArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = AprilTagDetectionArrayWrapper;
