// rclnodejs/generated/apriltag_msgs/apriltag_msgs__msg__AprilTagDetection.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PointWrapper = require('../../generated/apriltag_msgs/apriltag_msgs__msg__Point.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const AprilTagDetectionRefStruct = StructType({
  family: primitiveTypes.string,
  id: primitiveTypes.int32,
  hamming: primitiveTypes.int32,
  goodness: primitiveTypes.float32,
  decision_margin: primitiveTypes.float32,
  centre: PointWrapper.refObjectType,
  corners: ArrayType(PointWrapper.refObjectType, 4),
  homography: ArrayType(primitiveTypes.float64, 9),
});
const AprilTagDetectionRefArray = ArrayType(AprilTagDetectionRefStruct);
const AprilTagDetectionRefStructArray = StructType({
  data: AprilTagDetectionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AprilTagDetectionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AprilTagDetectionRefStruct(other._refObject.toObject());
      this._wrapperFields.family =  new StringWrapper(other._wrapperFields.family);
      this._wrapperFields.centre =  new PointWrapper(other._wrapperFields.centre);
      this._wrapperFields.corners = PointWrapper.createArray();
      this._wrapperFields.corners.copy(other._wrapperFields.corners);
      this._wrapperFields.homography = Float64Wrapper.createArray();
      this._wrapperFields.homography.copy(other._wrapperFields.homography);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AprilTagDetectionRefStruct();
    this._wrapperFields.family = new StringWrapper();
    this._wrapperFields.centre = new PointWrapper();
    this._wrapperFields.corners = PointWrapper.createArray();
    this._wrapperFields.homography = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new AprilTagDetectionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AprilTagDetectionArrayWrapper();
  }
  static get ArrayType() {
    return AprilTagDetectionArrayWrapper;
  }
  static get refObjectArrayType() {
    return AprilTagDetectionRefStructArray;
  }
  static get refObjectType() {
    return AprilTagDetectionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.family.freeze(own);
    }
    this._refObject.family = this._wrapperFields.family.refObject;
    this._wrapperFields.centre.freeze(own);
    this._refObject.centre = this._wrapperFields.centre.refObject;
    for (let i = 0; i < 4; i++) {
      if (this._wrapperFields.corners.data[i]) {
        this._wrapperFields.corners.data[i].freeze(own);
        this._refObject.corners[i] = this._wrapperFields.corners.data[i].refObject;
      }
    }
    this._refObject.homography = Array.from(this._wrapperFields.homography.data);
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.family.data = refObject.family.data;
    this._refObject.id = refObject.id;
    this._refObject.hamming = refObject.hamming;
    this._refObject.goodness = refObject.goodness;
    this._refObject.decision_margin = refObject.decision_margin;
    this._wrapperFields.centre.copyRefObject(refObject.centre);
    this._refObject.corners = refObject.corners;
    this._wrapperFields.corners.size = 4
    for (let i = 0; i < 4; i++) {
      this._wrapperFields.corners.data[i].copyRefObject(refObject.corners[i]);
    }
    this._wrapperFields.homography.fill(refObject.homography.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.family);
    PointWrapper.freeStruct(refObject.centre);
    for (let i = 0; i < 4; i++) {
      PointWrapper.freeStruct(refObject.corners[i]);
    }
  }
  static destroyRawROS(msg) {
    AprilTagDetectionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'apriltag_msgs', subFolder: 'msg', interfaceName: 'AprilTagDetection'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get family() {
    return this._wrapperFields.family.data;
  }
  set family(value) {
    this._wrapperFields.family.data = value;
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get hamming() {
    return this._refObject.hamming;
  }
  set hamming(value) {
    this._refObject.hamming = value;
  }
  get goodness() {
    return this._refObject.goodness;
  }
  set goodness(value) {
    this._refObject.goodness = value;
  }
  get decision_margin() {
    return this._refObject.decision_margin;
  }
  set decision_margin(value) {
    this._refObject.decision_margin = value;
  }
  get centre() {
    return this._wrapperFields.centre;
  }
  set centre(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.centre.copy(value);
    } else {
      this._wrapperFields.centre.copy(new PointWrapper(value));
    }
  }
  get corners() {
    return this._wrapperFields.corners;
  }
  set corners(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields.corners.fill(value);
  }
  get homography() {
    return this._wrapperFields['homography'].data;
  }
  set homography(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['homography'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new AprilTagDetectionRefStruct(refObject.toObject());
    this._wrapperFields.family.copyRefObject(this._refObject.family);
    this._wrapperFields.centre.copyRefObject(this._refObject.centre);
    this._refObject.corners = refObject.corners;
    this._wrapperFields.corners.size = 4
    for (let i = 0; i < 4; i++) {
      this._wrapperFields.corners.data[i].copyRefObject(refObject.corners[i]);
    }
    this._wrapperFields.homography.fill(refObject.homography.toArray());
  }
  copy(other) {
    this._refObject = new AprilTagDetectionRefStruct(other._refObject.toObject());
    this._wrapperFields.family.copy(other._wrapperFields.family);
    this._wrapperFields.centre.copy(other._wrapperFields.centre);
    this._wrapperFields.corners.copy(other._wrapperFields.corners);
    this._wrapperFields.homography.copy(other._wrapperFields.homography);
  }
  static get classType() {
    return AprilTagDetectionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"family","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hamming","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"goodness","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"decision_margin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"centre","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"apriltag_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"corners","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"apriltag_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"homography","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"apriltag_msgs","type":"AprilTagDetection","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AprilTagDetection"};
  }
  hasMember(name) {
    let memberNames = ["family","id","hamming","goodness","decision_margin","centre","corners","homography"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AprilTagDetectionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AprilTagDetectionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AprilTagDetectionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AprilTagDetectionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AprilTagDetectionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AprilTagDetectionRefArray(size);
    this._refObject = new AprilTagDetectionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AprilTagDetectionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AprilTagDetectionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AprilTagDetectionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AprilTagDetectionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AprilTagDetectionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AprilTagDetectionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = AprilTagDetectionWrapper;
