// rclnodejs/generated/std_msgs/std_msgs__msg__Int8.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const rclnodejs = require('../../lib/native_loader.js');
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int8RefStruct = StructType({
  data: primitiveTypes.int8,
});
const Int8RefArray = ArrayType(Int8RefStruct);
const Int8RefStructArray = StructType({
  data: ref.refType(ref.types.int8),
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Int8Wrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Int8RefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Int8RefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new Int8Wrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Int8ArrayWrapper();
  }
  static get ArrayType() {
    return Int8ArrayWrapper;
  }
  static get refObjectArrayType() {
    return Int8RefStructArray;
  }
  static get refObjectType() {
    return Int8RefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.data = refObject.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    Int8Wrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'std_msgs', subFolder: 'msg', interfaceName: 'Int8'};
  }
  static isPrimitive() {
    return true;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._refObject.data;
  }
  set data(value) {
    this._refObject.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Int8RefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new Int8RefStruct(other._refObject.toObject());
  }
  static get classType() {
    return Int8Wrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"data","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"std_msgs","type":"Int8","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Int8"};
  }
  hasMember(name) {
    let memberNames = ["data"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Int8ArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    if (Array.isArray(values)) {
      this._wrappers = new Int8Array(values);
    } else {
      this._wrappers = values;
    }
  }
  freeze(own) {
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      const buffer = Buffer.from(new Uint8Array(this._wrappers.buffer));
      this._refObject.data = buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Int8ArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Int8ArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = undefined;
    this._refObject = new Int8RefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Int8Array(size);
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    const byteLen = refObject.size * ref.types.int8.size;
    const arrayBuffer = refObject.data.length !== 0 ?
        rclnodejs.createArrayBufferFromAddress(refObject.data.hexAddress(), byteLen) :
        Buffer.alloc(0);
    this._wrappers = new Int8Array(arrayBuffer);
  }
  copy(other) {
    if (! (other instanceof Int8ArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Int8ArrayWrapper".');
    }
    this._resize(other.size);
    this._wrappers = other._wrappers.slice();
  }
  static freeArray(refObject) {
  }
  static get elementType() {
    return Int8Wrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return true;
  }
  get classType() {
    return Int8ArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Int8Wrapper;
