// rclnodejs/generated/std_msgs/std_msgs__msg__MultiArrayDimension.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MultiArrayDimensionRefStruct = StructType({
  label: primitiveTypes.string,
  size: primitiveTypes.uint32,
  stride: primitiveTypes.uint32,
});
const MultiArrayDimensionRefArray = ArrayType(MultiArrayDimensionRefStruct);
const MultiArrayDimensionRefStructArray = StructType({
  data: MultiArrayDimensionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MultiArrayDimensionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MultiArrayDimensionRefStruct(other._refObject.toObject());
      this._wrapperFields.label =  new StringWrapper(other._wrapperFields.label);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MultiArrayDimensionRefStruct();
    this._wrapperFields.label = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MultiArrayDimensionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MultiArrayDimensionArrayWrapper();
  }
  static get ArrayType() {
    return MultiArrayDimensionArrayWrapper;
  }
  static get refObjectArrayType() {
    return MultiArrayDimensionRefStructArray;
  }
  static get refObjectType() {
    return MultiArrayDimensionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.label.freeze(own);
    }
    this._refObject.label = this._wrapperFields.label.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.label.data = refObject.label.data;
    this._refObject.size = refObject.size;
    this._refObject.stride = refObject.stride;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.label);
  }
  static destroyRawROS(msg) {
    MultiArrayDimensionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'std_msgs', subFolder: 'msg', interfaceName: 'MultiArrayDimension'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get label() {
    return this._wrapperFields.label.data;
  }
  set label(value) {
    this._wrapperFields.label.data = value;
  }
  get size() {
    return this._refObject.size;
  }
  set size(value) {
    this._refObject.size = value;
  }
  get stride() {
    return this._refObject.stride;
  }
  set stride(value) {
    this._refObject.stride = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MultiArrayDimensionRefStruct(refObject.toObject());
    this._wrapperFields.label.copyRefObject(this._refObject.label);
  }
  copy(other) {
    this._refObject = new MultiArrayDimensionRefStruct(other._refObject.toObject());
    this._wrapperFields.label.copy(other._wrapperFields.label);
  }
  static get classType() {
    return MultiArrayDimensionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"label","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stride","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"std_msgs","type":"MultiArrayDimension","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MultiArrayDimension"};
  }
  hasMember(name) {
    let memberNames = ["label","size","stride"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MultiArrayDimensionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MultiArrayDimensionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MultiArrayDimensionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiArrayDimensionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiArrayDimensionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MultiArrayDimensionRefArray(size);
    this._refObject = new MultiArrayDimensionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MultiArrayDimensionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MultiArrayDimensionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MultiArrayDimensionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MultiArrayDimensionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MultiArrayDimensionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MultiArrayDimensionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MultiArrayDimensionWrapper;
