// rclnodejs/generated/nav_msgs/nav_msgs__msg__Path.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const PathRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  poses: PoseStampedWrapper.refObjectArrayType,
});
const PathRefArray = ArrayType(PathRefStruct);
const PathRefStructArray = StructType({
  data: PathRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class PathWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new PathRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.poses = PoseStampedWrapper.createArray();
      this._wrapperFields.poses.copy(other._wrapperFields.poses);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new PathRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.poses = PoseStampedWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PathWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PathArrayWrapper();
  }
  static get ArrayType() {
    return PathArrayWrapper;
  }
  static get refObjectArrayType() {
    return PathRefStructArray;
  }
  static get refObjectType() {
    return PathRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.poses.fill([]);
    }
    this._wrapperFields.poses.freeze(own);
    this._refObject.poses = this._wrapperFields.poses.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.poses.copyRefObject(refObject.poses);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.poses.size != 0) {
      PoseStampedWrapper.ArrayType.freeArray(refObject.poses);
      if (!PoseStampedWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.poses, PoseStampedWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    PathWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'msg', interfaceName: 'Path'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get poses() {
    return this._wrapperFields.poses;
  }
  set poses(value) {
    this._wrapperFields.poses.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new PathRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.poses.copyRefObject(this._refObject.poses);
  }
  copy(other) {
    this._refObject = new PathRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.poses.copy(other._wrapperFields.poses);
  }
  static get classType() {
    return PathWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Path"};
  }
  hasMember(name) {
    let memberNames = ["header","poses"];
    return memberNames.indexOf(name) !== -1;
  }
}
class PathArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PathWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PathWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PathArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PathArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new PathRefArray(size);
    this._refObject = new PathRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PathWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PathArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PathArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PathWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PathWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PathArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = PathWrapper;
