// rclnodejs/generated/nav_msgs/nav_msgs__srv__GetPlan_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const GetPlan_RequestRefStruct = StructType({
  start: PoseStampedWrapper.refObjectType,
  goal: PoseStampedWrapper.refObjectType,
  tolerance: primitiveTypes.float32,
});
const GetPlan_RequestRefArray = ArrayType(GetPlan_RequestRefStruct);
const GetPlan_RequestRefStructArray = StructType({
  data: GetPlan_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetPlan_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPlan_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.start =  new PoseStampedWrapper(other._wrapperFields.start);
      this._wrapperFields.goal =  new PoseStampedWrapper(other._wrapperFields.goal);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetPlan_RequestRefStruct();
    this._wrapperFields.start = new PoseStampedWrapper();
    this._wrapperFields.goal = new PoseStampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPlan_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPlan_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetPlan_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPlan_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetPlan_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.start.freeze(own);
    this._refObject.start = this._wrapperFields.start.refObject;
    this._wrapperFields.goal.freeze(own);
    this._refObject.goal = this._wrapperFields.goal.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.start.copyRefObject(refObject.start);
    this._wrapperFields.goal.copyRefObject(refObject.goal);
    this._refObject.tolerance = refObject.tolerance;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.start);
    PoseStampedWrapper.freeStruct(refObject.goal);
  }
  static destroyRawROS(msg) {
    GetPlan_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'srv', interfaceName: 'GetPlan_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start() {
    return this._wrapperFields.start;
  }
  set start(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.start.copy(value);
    } else {
      this._wrapperFields.start.copy(new PoseStampedWrapper(value));
    }
  }
  get goal() {
    return this._wrapperFields.goal;
  }
  set goal(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.goal.copy(value);
    } else {
      this._wrapperFields.goal.copy(new PoseStampedWrapper(value));
    }
  }
  get tolerance() {
    return this._refObject.tolerance;
  }
  set tolerance(value) {
    this._refObject.tolerance = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetPlan_RequestRefStruct(refObject.toObject());
    this._wrapperFields.start.copyRefObject(this._refObject.start);
    this._wrapperFields.goal.copyRefObject(this._refObject.goal);
  }
  copy(other) {
    this._refObject = new GetPlan_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.start.copy(other._wrapperFields.start);
    this._wrapperFields.goal.copy(other._wrapperFields.goal);
  }
  static get classType() {
    return GetPlan_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"tolerance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"GetPlan_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPlan_Request"};
  }
  hasMember(name) {
    let memberNames = ["start","goal","tolerance"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetPlan_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPlan_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPlan_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPlan_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPlan_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetPlan_RequestRefArray(size);
    this._refObject = new GetPlan_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPlan_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPlan_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPlan_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPlan_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPlan_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPlan_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetPlan_RequestWrapper;
