// rclnodejs/generated/nav_msgs/nav_msgs__srv__LoadMap_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const OccupancyGridWrapper = require('../../generated/nav_msgs/nav_msgs__msg__OccupancyGrid.js');
const LoadMap_ResponseRefStruct = StructType({
  map: OccupancyGridWrapper.refObjectType,
  result: primitiveTypes.uint8,
});
const LoadMap_ResponseRefArray = ArrayType(LoadMap_ResponseRefStruct);
const LoadMap_ResponseRefStructArray = StructType({
  data: LoadMap_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LoadMap_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LoadMap_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.map =  new OccupancyGridWrapper(other._wrapperFields.map);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LoadMap_ResponseRefStruct();
    this._wrapperFields.map = new OccupancyGridWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LoadMap_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LoadMap_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return LoadMap_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return LoadMap_ResponseRefStructArray;
  }
  static get refObjectType() {
    return LoadMap_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.map.freeze(own);
    this._refObject.map = this._wrapperFields.map.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map.copyRefObject(refObject.map);
    this._refObject.result = refObject.result;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    OccupancyGridWrapper.freeStruct(refObject.map);
  }
  static destroyRawROS(msg) {
    LoadMap_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'srv', interfaceName: 'LoadMap_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map() {
    return this._wrapperFields.map;
  }
  set map(value) {
    if (value instanceof OccupancyGridWrapper) {
      this._wrapperFields.map.copy(value);
    } else {
      this._wrapperFields.map.copy(new OccupancyGridWrapper(value));
    }
  }
  get result() {
    return this._refObject.result;
  }
  set result(value) {
    this._refObject.result = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LoadMap_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.map.copyRefObject(this._refObject.map);
  }
  copy(other) {
    this._refObject = new LoadMap_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.map.copy(other._wrapperFields.map);
  }
  static get classType() {
    return LoadMap_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"RESULT_SUCCESS","value":0},{"type":"uint8","name":"RESULT_MAP_DOES_NOT_EXIST","value":1},{"type":"uint8","name":"RESULT_INVALID_MAP_DATA","value":2},{"type":"uint8","name":"RESULT_INVALID_MAP_METADATA","value":3},{"type":"uint8","name":"RESULT_UNDEFINED_FAILURE","value":255}],"fields":[{"name":"map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"OccupancyGrid","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"LoadMap_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LoadMap_Response"};
  }
  hasMember(name) {
    let memberNames = ["map","result"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LoadMap_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LoadMap_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LoadMap_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadMap_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadMap_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LoadMap_ResponseRefArray(size);
    this._refObject = new LoadMap_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LoadMap_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LoadMap_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LoadMap_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LoadMap_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LoadMap_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LoadMap_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(LoadMap_ResponseWrapper, "RESULT_SUCCESS", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoadMap_ResponseWrapper, "RESULT_MAP_DOES_NOT_EXIST", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoadMap_ResponseWrapper, "RESULT_INVALID_MAP_DATA", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoadMap_ResponseWrapper, "RESULT_INVALID_MAP_METADATA", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoadMap_ResponseWrapper, "RESULT_UNDEFINED_FAILURE", {value: 255, writable: false, enumerable: true, configurable: true});
module.exports = LoadMap_ResponseWrapper;
