// rclnodejs/generated/nav_msgs/nav_msgs__msg__Goals.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const GoalsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  goals: PoseStampedWrapper.refObjectArrayType,
});
const GoalsRefArray = ArrayType(GoalsRefStruct);
const GoalsRefStructArray = StructType({
  data: GoalsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GoalsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GoalsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.goals = PoseStampedWrapper.createArray();
      this._wrapperFields.goals.copy(other._wrapperFields.goals);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GoalsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.goals = PoseStampedWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GoalsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GoalsArrayWrapper();
  }
  static get ArrayType() {
    return GoalsArrayWrapper;
  }
  static get refObjectArrayType() {
    return GoalsRefStructArray;
  }
  static get refObjectType() {
    return GoalsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.goals.fill([]);
    }
    this._wrapperFields.goals.freeze(own);
    this._refObject.goals = this._wrapperFields.goals.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.goals.copyRefObject(refObject.goals);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.goals.size != 0) {
      PoseStampedWrapper.ArrayType.freeArray(refObject.goals);
      if (!PoseStampedWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.goals, PoseStampedWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GoalsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'msg', interfaceName: 'Goals'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get goals() {
    return this._wrapperFields.goals;
  }
  set goals(value) {
    this._wrapperFields.goals.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GoalsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.goals.copyRefObject(this._refObject.goals);
  }
  copy(other) {
    this._refObject = new GoalsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.goals.copy(other._wrapperFields.goals);
  }
  static get classType() {
    return GoalsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goals","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"Goals","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Goals"};
  }
  hasMember(name) {
    let memberNames = ["header","goals"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GoalsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GoalsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GoalsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GoalsRefArray(size);
    this._refObject = new GoalsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GoalsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GoalsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GoalsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GoalsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GoalsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GoalsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GoalsWrapper;
