// rclnodejs/generated/rosbag2_interfaces/rosbag2_interfaces__srv__Seek_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const Seek_RequestRefStruct = StructType({
  time: TimeWrapper.refObjectType,
});
const Seek_RequestRefArray = ArrayType(Seek_RequestRefStruct);
const Seek_RequestRefStructArray = StructType({
  data: Seek_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Seek_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Seek_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.time =  new TimeWrapper(other._wrapperFields.time);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Seek_RequestRefStruct();
    this._wrapperFields.time = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Seek_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Seek_RequestArrayWrapper();
  }
  static get ArrayType() {
    return Seek_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return Seek_RequestRefStructArray;
  }
  static get refObjectType() {
    return Seek_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.time.freeze(own);
    this._refObject.time = this._wrapperFields.time.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.time.copyRefObject(refObject.time);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.time);
  }
  static destroyRawROS(msg) {
    Seek_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbag2_interfaces', subFolder: 'srv', interfaceName: 'Seek_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get time() {
    return this._wrapperFields.time;
  }
  set time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.time.copy(value);
    } else {
      this._wrapperFields.time.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new Seek_RequestRefStruct(refObject.toObject());
    this._wrapperFields.time.copyRefObject(this._refObject.time);
  }
  copy(other) {
    this._refObject = new Seek_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.time.copy(other._wrapperFields.time);
  }
  static get classType() {
    return Seek_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rosbag2_interfaces","type":"Seek_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Seek_Request"};
  }
  hasMember(name) {
    let memberNames = ["time"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Seek_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Seek_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Seek_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Seek_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Seek_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Seek_RequestRefArray(size);
    this._refObject = new Seek_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Seek_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Seek_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Seek_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Seek_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Seek_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Seek_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Seek_RequestWrapper;
