// rclnodejs/generated/rosbag2_interfaces/rosbag2_interfaces__msg__ReadSplitEvent.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ReadSplitEventRefStruct = StructType({
  closed_file: primitiveTypes.string,
  opened_file: primitiveTypes.string,
});
const ReadSplitEventRefArray = ArrayType(ReadSplitEventRefStruct);
const ReadSplitEventRefStructArray = StructType({
  data: ReadSplitEventRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ReadSplitEventWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ReadSplitEventRefStruct(other._refObject.toObject());
      this._wrapperFields.closed_file =  new StringWrapper(other._wrapperFields.closed_file);
      this._wrapperFields.opened_file =  new StringWrapper(other._wrapperFields.opened_file);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ReadSplitEventRefStruct();
    this._wrapperFields.closed_file = new StringWrapper();
    this._wrapperFields.opened_file = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ReadSplitEventWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ReadSplitEventArrayWrapper();
  }
  static get ArrayType() {
    return ReadSplitEventArrayWrapper;
  }
  static get refObjectArrayType() {
    return ReadSplitEventRefStructArray;
  }
  static get refObjectType() {
    return ReadSplitEventRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.closed_file.freeze(own);
    }
    this._refObject.closed_file = this._wrapperFields.closed_file.refObject;
    if (own) {
      this._wrapperFields.opened_file.freeze(own);
    }
    this._refObject.opened_file = this._wrapperFields.opened_file.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.closed_file.data = refObject.closed_file.data;
    this._wrapperFields.opened_file.data = refObject.opened_file.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.closed_file);
    StringWrapper.freeStruct(refObject.opened_file);
  }
  static destroyRawROS(msg) {
    ReadSplitEventWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbag2_interfaces', subFolder: 'msg', interfaceName: 'ReadSplitEvent'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get closed_file() {
    return this._wrapperFields.closed_file.data;
  }
  set closed_file(value) {
    this._wrapperFields.closed_file.data = value;
  }
  get opened_file() {
    return this._wrapperFields.opened_file.data;
  }
  set opened_file(value) {
    this._wrapperFields.opened_file.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ReadSplitEventRefStruct(refObject.toObject());
    this._wrapperFields.closed_file.copyRefObject(this._refObject.closed_file);
    this._wrapperFields.opened_file.copyRefObject(this._refObject.opened_file);
  }
  copy(other) {
    this._refObject = new ReadSplitEventRefStruct(other._refObject.toObject());
    this._wrapperFields.closed_file.copy(other._wrapperFields.closed_file);
    this._wrapperFields.opened_file.copy(other._wrapperFields.opened_file);
  }
  static get classType() {
    return ReadSplitEventWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"closed_file","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"opened_file","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosbag2_interfaces","type":"ReadSplitEvent","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ReadSplitEvent"};
  }
  hasMember(name) {
    let memberNames = ["closed_file","opened_file"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ReadSplitEventArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ReadSplitEventWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ReadSplitEventWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ReadSplitEventArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ReadSplitEventArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ReadSplitEventRefArray(size);
    this._refObject = new ReadSplitEventRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ReadSplitEventWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ReadSplitEventArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ReadSplitEventArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ReadSplitEventWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ReadSplitEventWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ReadSplitEventArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ReadSplitEventWrapper;
