// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__FloatingPointRange.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const FloatingPointRangeRefStruct = StructType({
  from_value: primitiveTypes.float64,
  to_value: primitiveTypes.float64,
  step: primitiveTypes.float64,
});
const FloatingPointRangeRefArray = ArrayType(FloatingPointRangeRefStruct);
const FloatingPointRangeRefStructArray = StructType({
  data: FloatingPointRangeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class FloatingPointRangeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new FloatingPointRangeRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new FloatingPointRangeRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new FloatingPointRangeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FloatingPointRangeArrayWrapper();
  }
  static get ArrayType() {
    return FloatingPointRangeArrayWrapper;
  }
  static get refObjectArrayType() {
    return FloatingPointRangeRefStructArray;
  }
  static get refObjectType() {
    return FloatingPointRangeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.from_value = refObject.from_value;
    this._refObject.to_value = refObject.to_value;
    this._refObject.step = refObject.step;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    FloatingPointRangeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'FloatingPointRange'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get from_value() {
    return this._refObject.from_value;
  }
  set from_value(value) {
    this._refObject.from_value = value;
  }
  get to_value() {
    return this._refObject.to_value;
  }
  set to_value(value) {
    this._refObject.to_value = value;
  }
  get step() {
    return this._refObject.step;
  }
  set step(value) {
    this._refObject.step = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FloatingPointRangeRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new FloatingPointRangeRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return FloatingPointRangeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"from_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"to_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"FloatingPointRange","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FloatingPointRange"};
  }
  hasMember(name) {
    let memberNames = ["from_value","to_value","step"];
    return memberNames.indexOf(name) !== -1;
  }
}
class FloatingPointRangeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FloatingPointRangeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FloatingPointRangeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FloatingPointRangeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FloatingPointRangeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new FloatingPointRangeRefArray(size);
    this._refObject = new FloatingPointRangeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FloatingPointRangeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FloatingPointRangeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FloatingPointRangeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FloatingPointRangeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FloatingPointRangeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FloatingPointRangeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = FloatingPointRangeWrapper;
