// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__ParameterType.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ParameterTypeRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const ParameterTypeRefArray = ArrayType(ParameterTypeRefStruct);
const ParameterTypeRefStructArray = StructType({
  data: ParameterTypeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ParameterTypeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParameterTypeRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ParameterTypeRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new ParameterTypeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParameterTypeArrayWrapper();
  }
  static get ArrayType() {
    return ParameterTypeArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParameterTypeRefStructArray;
  }
  static get refObjectType() {
    return ParameterTypeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ParameterTypeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'ParameterType'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ParameterTypeRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new ParameterTypeRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return ParameterTypeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"PARAMETER_NOT_SET","value":0},{"type":"uint8","name":"PARAMETER_BOOL","value":1},{"type":"uint8","name":"PARAMETER_INTEGER","value":2},{"type":"uint8","name":"PARAMETER_DOUBLE","value":3},{"type":"uint8","name":"PARAMETER_STRING","value":4},{"type":"uint8","name":"PARAMETER_BYTE_ARRAY","value":5},{"type":"uint8","name":"PARAMETER_BOOL_ARRAY","value":6},{"type":"uint8","name":"PARAMETER_INTEGER_ARRAY","value":7},{"type":"uint8","name":"PARAMETER_DOUBLE_ARRAY","value":8},{"type":"uint8","name":"PARAMETER_STRING_ARRAY","value":9}],"fields":[],"baseType":{"pkgName":"rcl_interfaces","type":"ParameterType","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParameterType"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ParameterTypeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParameterTypeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParameterTypeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParameterTypeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParameterTypeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ParameterTypeRefArray(size);
    this._refObject = new ParameterTypeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParameterTypeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParameterTypeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParameterTypeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParameterTypeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParameterTypeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParameterTypeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_NOT_SET", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_BOOL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_INTEGER", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_DOUBLE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_STRING", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_BYTE_ARRAY", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_BOOL_ARRAY", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_INTEGER_ARRAY", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_DOUBLE_ARRAY", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ParameterTypeWrapper, "PARAMETER_STRING_ARRAY", {value: 9, writable: false, enumerable: true, configurable: true});
module.exports = ParameterTypeWrapper;
