// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__ListParameters_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ListParameters_RequestRefStruct = StructType({
  prefixes: StringWrapper.refObjectArrayType,
  depth: primitiveTypes.uint64,
});
const ListParameters_RequestRefArray = ArrayType(ListParameters_RequestRefStruct);
const ListParameters_RequestRefStructArray = StructType({
  data: ListParameters_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ListParameters_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListParameters_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.prefixes = StringWrapper.createArray();
      this._wrapperFields.prefixes.copy(other._wrapperFields.prefixes);
      this.prefixes = other.prefixes;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ListParameters_RequestRefStruct();
    this._wrapperFields.prefixes = StringWrapper.createArray();
    this._prefixesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new ListParameters_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListParameters_RequestArrayWrapper();
  }
  static get ArrayType() {
    return ListParameters_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListParameters_RequestRefStructArray;
  }
  static get refObjectType() {
    return ListParameters_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.prefixes.fill([]);
    } else {
      this._wrapperFields.prefixes.fill(this._prefixesArray);
    }
    this._wrapperFields.prefixes.freeze(own);
    this._refObject.prefixes = this._wrapperFields.prefixes.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.prefixes.data.length = refObject.prefixes.size;
    for (let index = 0; index < refObject.prefixes.size; index++) {
      this._prefixesArray[index] = refObject.prefixes.data[index].data;
    }
    this._refObject.depth = refObject.depth;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.prefixes.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.prefixes);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.prefixes, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ListParameters_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'ListParameters_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get prefixes() {
    return this._prefixesArray;
  }
  set prefixes(value) {
    this._prefixesArray = value;
  }
  get depth() {
    return BigInt(this._refObject.depth);
  }
  set depth(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('depth must be type of bigint');
    }
    this._refObject.depth = value.toString();
  }
  copyRefObject(refObject) {
    this._refObject = new ListParameters_RequestRefStruct(refObject.toObject());
    refObject.prefixes.data.length = refObject.prefixes.size;
    for (let index = 0; index < refObject.prefixes.size; index++) {
      this._prefixesArray[index] = refObject.prefixes.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ListParameters_RequestRefStruct(other._refObject.toObject());
    this._prefixesArray = other._prefixesArray.slice();
  }
  static get classType() {
    return ListParameters_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint64","name":"DEPTH_RECURSIVE","value":0}],"fields":[{"name":"prefixes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"depth","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"ListParameters_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListParameters_Request"};
  }
  hasMember(name) {
    let memberNames = ["prefixes","depth"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ListParameters_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListParameters_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListParameters_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListParameters_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListParameters_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ListParameters_RequestRefArray(size);
    this._refObject = new ListParameters_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListParameters_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListParameters_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListParameters_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListParameters_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListParameters_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListParameters_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ListParameters_RequestWrapper, "DEPTH_RECURSIVE", {value: 0, writable: false, enumerable: true, configurable: true});
module.exports = ListParameters_RequestWrapper;
