// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__ListParametersResult.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ListParametersResultRefStruct = StructType({
  names: StringWrapper.refObjectArrayType,
  prefixes: StringWrapper.refObjectArrayType,
});
const ListParametersResultRefArray = ArrayType(ListParametersResultRefStruct);
const ListParametersResultRefStructArray = StructType({
  data: ListParametersResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ListParametersResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListParametersResultRefStruct(other._refObject.toObject());
      this._wrapperFields.names = StringWrapper.createArray();
      this._wrapperFields.names.copy(other._wrapperFields.names);
      this.names = other.names;
      this._wrapperFields.prefixes = StringWrapper.createArray();
      this._wrapperFields.prefixes.copy(other._wrapperFields.prefixes);
      this.prefixes = other.prefixes;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ListParametersResultRefStruct();
    this._wrapperFields.names = StringWrapper.createArray();
    this._namesArray = [];
    this._wrapperFields.prefixes = StringWrapper.createArray();
    this._prefixesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new ListParametersResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListParametersResultArrayWrapper();
  }
  static get ArrayType() {
    return ListParametersResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListParametersResultRefStructArray;
  }
  static get refObjectType() {
    return ListParametersResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.names.fill([]);
    } else {
      this._wrapperFields.names.fill(this._namesArray);
    }
    this._wrapperFields.names.freeze(own);
    this._refObject.names = this._wrapperFields.names.refObject;
    if (own) {
      this._wrapperFields.prefixes.fill([]);
    } else {
      this._wrapperFields.prefixes.fill(this._prefixesArray);
    }
    this._wrapperFields.prefixes.freeze(own);
    this._refObject.prefixes = this._wrapperFields.prefixes.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
    refObject.prefixes.data.length = refObject.prefixes.size;
    for (let index = 0; index < refObject.prefixes.size; index++) {
      this._prefixesArray[index] = refObject.prefixes.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.names);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.prefixes.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.prefixes);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.prefixes, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ListParametersResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'ListParametersResult'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get names() {
    return this._namesArray;
  }
  set names(value) {
    this._namesArray = value;
  }
  get prefixes() {
    return this._prefixesArray;
  }
  set prefixes(value) {
    this._prefixesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ListParametersResultRefStruct(refObject.toObject());
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
    refObject.prefixes.data.length = refObject.prefixes.size;
    for (let index = 0; index < refObject.prefixes.size; index++) {
      this._prefixesArray[index] = refObject.prefixes.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ListParametersResultRefStruct(other._refObject.toObject());
    this._namesArray = other._namesArray.slice();
    this._prefixesArray = other._prefixesArray.slice();
  }
  static get classType() {
    return ListParametersResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"prefixes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"ListParametersResult","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListParametersResult"};
  }
  hasMember(name) {
    let memberNames = ["names","prefixes"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ListParametersResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListParametersResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListParametersResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListParametersResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListParametersResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ListParametersResultRefArray(size);
    this._refObject = new ListParametersResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListParametersResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListParametersResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListParametersResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListParametersResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListParametersResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListParametersResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ListParametersResultWrapper;
