// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__Log.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LogRefStruct = StructType({
  stamp: TimeWrapper.refObjectType,
  level: primitiveTypes.uint8,
  name: primitiveTypes.string,
  msg: primitiveTypes.string,
  file: primitiveTypes.string,
  function: primitiveTypes.string,
  line: primitiveTypes.uint32,
});
const LogRefArray = ArrayType(LogRefStruct);
const LogRefStructArray = StructType({
  data: LogRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LogWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LogRefStruct(other._refObject.toObject());
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.msg =  new StringWrapper(other._wrapperFields.msg);
      this._wrapperFields.file =  new StringWrapper(other._wrapperFields.file);
      this._wrapperFields.function =  new StringWrapper(other._wrapperFields.function);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LogRefStruct();
    this._wrapperFields.stamp = new TimeWrapper();
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.msg = new StringWrapper();
    this._wrapperFields.file = new StringWrapper();
    this._wrapperFields.function = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LogWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LogArrayWrapper();
  }
  static get ArrayType() {
    return LogArrayWrapper;
  }
  static get refObjectArrayType() {
    return LogRefStructArray;
  }
  static get refObjectType() {
    return LogRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.stamp.freeze(own);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.msg.freeze(own);
    }
    this._refObject.msg = this._wrapperFields.msg.refObject;
    if (own) {
      this._wrapperFields.file.freeze(own);
    }
    this._refObject.file = this._wrapperFields.file.refObject;
    if (own) {
      this._wrapperFields.function.freeze(own);
    }
    this._refObject.function = this._wrapperFields.function.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
    this._refObject.level = refObject.level;
    this._wrapperFields.name.data = refObject.name.data;
    this._wrapperFields.msg.data = refObject.msg.data;
    this._wrapperFields.file.data = refObject.file.data;
    this._wrapperFields.function.data = refObject.function.data;
    this._refObject.line = refObject.line;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.stamp);
    StringWrapper.freeStruct(refObject.name);
    StringWrapper.freeStruct(refObject.msg);
    StringWrapper.freeStruct(refObject.file);
    StringWrapper.freeStruct(refObject.function);
  }
  static destroyRawROS(msg) {
    LogWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'Log'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  get level() {
    return this._refObject.level;
  }
  set level(value) {
    this._refObject.level = value;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get msg() {
    return this._wrapperFields.msg.data;
  }
  set msg(value) {
    this._wrapperFields.msg.data = value;
  }
  get file() {
    return this._wrapperFields.file.data;
  }
  set file(value) {
    this._wrapperFields.file.data = value;
  }
  get function() {
    return this._wrapperFields.function.data;
  }
  set function(value) {
    this._wrapperFields.function.data = value;
  }
  get line() {
    return this._refObject.line;
  }
  set line(value) {
    this._refObject.line = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LogRefStruct(refObject.toObject());
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._wrapperFields.msg.copyRefObject(this._refObject.msg);
    this._wrapperFields.file.copyRefObject(this._refObject.file);
    this._wrapperFields.function.copyRefObject(this._refObject.function);
  }
  copy(other) {
    this._refObject = new LogRefStruct(other._refObject.toObject());
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._wrapperFields.msg.copy(other._wrapperFields.msg);
    this._wrapperFields.file.copy(other._wrapperFields.file);
    this._wrapperFields.function.copy(other._wrapperFields.function);
  }
  static get classType() {
    return LogWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"byte","name":"DEBUG","value":10},{"type":"byte","name":"INFO","value":20},{"type":"byte","name":"WARN","value":30},{"type":"byte","name":"ERROR","value":40},{"type":"byte","name":"FATAL","value":50}],"fields":[{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"level","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"msg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"file","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"function","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"line","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"Log","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Log"};
  }
  hasMember(name) {
    let memberNames = ["stamp","level","name","msg","file","function","line"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LogArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LogWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LogWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LogRefArray(size);
    this._refObject = new LogRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LogWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LogArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LogArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LogWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LogWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LogArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(LogWrapper, "DEBUG", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LogWrapper, "INFO", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LogWrapper, "WARN", {value: 30, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LogWrapper, "ERROR", {value: 40, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LogWrapper, "FATAL", {value: 50, writable: false, enumerable: true, configurable: true});
module.exports = LogWrapper;
