// rclnodejs/generated/builtin_interfaces/builtin_interfaces__msg__Time.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeRefStruct = StructType({
  sec: primitiveTypes.int32,
  nanosec: primitiveTypes.uint32,
});
const TimeRefArray = ArrayType(TimeRefStruct);
const TimeRefStructArray = StructType({
  data: TimeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TimeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TimeRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TimeRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new TimeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TimeArrayWrapper();
  }
  static get ArrayType() {
    return TimeArrayWrapper;
  }
  static get refObjectArrayType() {
    return TimeRefStructArray;
  }
  static get refObjectType() {
    return TimeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.sec = refObject.sec;
    this._refObject.nanosec = refObject.nanosec;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    TimeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'builtin_interfaces', subFolder: 'msg', interfaceName: 'Time'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sec() {
    return this._refObject.sec;
  }
  set sec(value) {
    this._refObject.sec = value;
  }
  get nanosec() {
    return this._refObject.nanosec;
  }
  set nanosec(value) {
    this._refObject.nanosec = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TimeRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new TimeRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return TimeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"sec","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nanosec","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Time"};
  }
  hasMember(name) {
    let memberNames = ["sec","nanosec"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TimeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TimeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TimeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TimeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TimeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TimeRefArray(size);
    this._refObject = new TimeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TimeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TimeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TimeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TimeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TimeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TimeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TimeWrapper;
